---
title: "Known Pathways in DEGs of Each Cel Types"
author: "Altair Wei"
date: "2023-03-30"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source("../scripts/LoadUtils.R", chdir = TRUE)
```

## Load Data

### Metadata

```{r}
cell_types <- c(
  "Stomata",
  "Epidermis",
  "Chlorenchyma",
  "Parenchyma",
  "Outer sheath",
  "Inner sheath",
  "Phloem",
  "Procambium"
)
```

### Gene Lists

```{r}
filelist <- Sys.glob("../data/pathways/*.txt")
names(filelist) <- basename(filelist) |> stringr::str_remove(stringr::fixed(".txt"))

pathways <- lapply(filelist, function(filename) {
  genes <- stringr::str_trim(readLines(filename))
  genes[genes != ""]
})
```

```{r}
summary(pathways)[, 1, drop = FALSE]
```

```{r}
pathways |>
  tibble::enframe(name = "pathway", value = "gene") |>
  tidyr::unnest(cols = c(gene)) |>
  dplyr::group_by(gene) |>
  dplyr::summarise(pathways = paste(pathway, collapse = " | ")) |>
  download_table(
    table_id = "gene_pathways",
    filterable = TRUE,
    searchable = TRUE,
    showPageSizeOptions = TRUE)
```


### DEGs of Cell Types

```{r}
ds_res <- readRDS(Sys.glob(
  "../results/ObjectCache/DifferentialState/ds_res_*.rds"))
```

```{r}
logfc_data <- xfun::cache_rds(
  file = "ds_deg_logfc_data.rds",
  dir = "../results/ObjectCache/DifferentialState/",
  expr = local({
    logfc_mtx <- ds_res$DESeq2 |>
      rhapsodykit::diff_state_significant(fdr_limit = 0.05, logfc_limit = 0) |>
      rhapsodykit::diff_state_format() |>
      dplyr::filter(!stringr::str_detect(contrast, "X\\dDPI\\.PNR2-X\\dDPI\\.TR4")) |>
      dplyr::select(gene, cluster_id, contrast, logFC) |>
      dplyr::mutate(
        time = stringr::str_extract(contrast, "\\dDPI"),
        contrast = stringr::str_replace(contrast,
          "X(\\dDPI)\\.(.*)-X(\\dDPI)\\.(.*)", "\\2-\\4")) |>
      tidyr::pivot_wider(names_from = "time", values_from = "logFC", values_fill = NA) |>
      tidyr::pivot_longer(cols = dplyr::ends_with("DPI"), names_to = "time", values_to = "logFC") |>
      tidyr::pivot_wider(names_from = c(cluster_id, contrast, time),
                         names_sep = "#",
                         values_from = logFC) |>
      tibble::column_to_rownames("gene") |>
      as.matrix()

    str_coldata <- strsplit(colnames(logfc_mtx), "#")

    mtx_coldata <- data.frame(
      colidx = seq_along(colnames(logfc_mtx)),
      cell_types = factor(sapply(str_coldata, function(x) x[[1]]),
                          levels = cell_types),
      contrast = factor(sapply(str_coldata, function(x) x[[2]]),
                        levels = c("PNR2-MOCK", "TR4-MOCK")),
      times = factor(sapply(str_coldata, function(x) x[[3]]),
                     levels = c("1DPI", "2DPI", "3DPI"))
    )

    mtx_coldata <- mtx_coldata[with(mtx_coldata, order(cell_types, contrast, times)),]

    logfc_mtx <- logfc_mtx[, mtx_coldata$colidx]

    ha = ComplexHeatmap::HeatmapAnnotation(
      which = "column",
      df = mtx_coldata[-1],
      col = list(
        cell_types = structure(
          ggthemes::tableau_color_pal()(
            length(levels(mtx_coldata$cell_types))),
          names = levels(mtx_coldata$cell_types)
        ),
        contrast = structure(
          c("#E41A1C", "#377EB8"),
          names = levels(mtx_coldata$contrast)
        ),
        times = structure(
          RColorBrewer::brewer.pal(3, "YlGn"),
          names = levels(mtx_coldata$times)
        )
      ))

    list(mtx = logfc_mtx, coldata = mtx_coldata, heatanno = ha)
  })
)
```

### Calculate GSVA

```{r}
pseudobulk_cpm_mean <- readRDS(Sys.glob(
  "../results/ObjectCache/Pseudobulk/pseudobulk_cpm_mean_*.rds"))
```

```{r}
gsva_df <- xfun::cache_rds(
  file = "pseudobulk_8celltypes_gsva_df.rds",
  dir = "../results/ObjectCache/Pseudobulk/",
  rerun = TRUE,
  expr = local({
    expr_gsva <- GSVA::gsva(
      pseudobulk_cpm_mean, pathways,
      method = "gsva", kcdf = "Gaussian",
      verbose = FALSE)

    gsva_df <- as.data.frame(expr_gsva) |>
        tibble::rownames_to_column("Pathway") |>
        tidyr::pivot_longer(-Pathway, names_to = "bulk", values_to = "ES") |>
        tidyr::separate(bulk, sep = "\\.", into = c("treatment", "time", "cellType"))

    gsva_df$cellType <- factor(gsva_df$cellType, levels = cell_types)

    gsva_df
  })
)
```

### Functions

```{r}
printDEGNum <- function(mtx, coldata, genes) {
  detected <- intersect(rownames(mtx), genes)
  mtx <- mtx[detected,]
  coldata$UP <- colSums(mtx > 0, na.rm = TRUE)
  coldata$DOWN <- colSums(mtx < 0, na.rm = TRUE)
  dplyr::select(coldata, cell_types, contrast, times, UP, DOWN) |>
    tibble::remove_rownames()
}
```


## Analysis Based on Known Pathways

### Phytosynthesis related {.tabset}

#### PhotosynAnt

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "PhotosynAnt")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "PhotosynAnt")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["PhotosynAnt"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=6, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["PhotosynAnt"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### PhotosynCfix

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "PhotosynCfix")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "PhotosynCfix")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["PhotosynCfix"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=14, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["PhotosynCfix"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### PhotosynPSEA

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "PhotosynPSEA")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "PhotosynPSEA")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["PhotosynPSEA"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=14, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["PhotosynPSEA"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

### Cell wall related {.tabset}

#### Aromatic amino acid

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "Aromatic amino acid")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "Aromatic amino acid")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["Aromatic amino acid"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=6, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["Aromatic amino acid"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### General phenylpropanoid

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "General phenylpropanoid")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "General phenylpropanoid")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["General phenylpropanoid"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=8, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["General phenylpropanoid"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### Lignin biosynthesis

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "Lignin biosynthesis")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "Lignin biosynthesis")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["Lignin biosynthesis"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=6, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["Lignin biosynthesis"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### Flavanones

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "Flavanones")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "Flavanones")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["Flavanones"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["Flavanones"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### HCAAs

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "HCAAs")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "HCAAs")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["HCAAs"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["HCAAs"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### Suberin

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "Suberin")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "Suberin")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["Suberin"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["Suberin"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### CalloseSyn

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "CalloseSyn")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "CalloseSyn")
```

```{r fig.height=4, fig.width=14}
detected <- intersect(rownames(logfc_data$mtx), pathways[["CalloseSyn"]])
print(detected)
logfc_data$mtx[detected,][!is.na(logfc_data$mtx[detected,])]
```

#### Polyamine

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "Polyamine")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "Polyamine")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["Polyamine"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=6, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["Polyamine"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

### Immune related {.tabset}

#### CCNBLRR

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "CCNBLRR")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "CCNBLRR")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["CCNBLRR"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["CCNBLRR"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### NBLRR

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "NBLRR")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "NBLRR")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["NBLRR"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=10, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["NBLRR"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### TIR

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "TIR")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "TIR")
```

```{r fig.height=4, fig.width=14}
detected <- intersect(rownames(logfc_data$mtx), pathways[["TIR"]])
print(detected)
logfc_data$mtx[detected,][!is.na(logfc_data$mtx[detected,])]
```

#### PR

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "PR")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "PR")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["PR"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["PR"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### RLK

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "RLK")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "RLK")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["RLK"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=20, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["RLK"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### RBOH

Gene list got from:

> Reference: Hu, C.-H., Wei, X.-Y., Yuan, B., Yao, L.-B., Ma, T.-T., Zhang, P.-P., Wang, X., Wang, P.-Q., Liu, W.-T., Li, W.-Q., Meng, L.-S., & Chen, K.-M. (2018). Genome-Wide Identification and Functional Analysis of NADPH Oxidase Family Genes in Wheat During Development and Environmental Stress Responses. Frontiers in Plant Science, 9. https://doi.org/10.3389/fpls.2018.00906

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "RBOH")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "RBOH")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["RBOH"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["RBOH"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### TaNPR

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "TaNPR")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "TaNPR")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["TaNPR"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["TaNPR"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### RxNonly

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "RxNonly")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "RxNonly")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["RxNonly"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["RxNonly"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### GOSAR

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "GOSAR")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "GOSAR")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["GOSAR"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["GOSAR"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### NHP

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "NHP")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "NHP")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["NHP"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["NHP"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

### Metabolism related {.tabset}

#### gluconeogenesis

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "gluconeogenesis")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "gluconeogenesis")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["gluconeogenesis"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["gluconeogenesis"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### glycolysis

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "glycolysis")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "glycolysis")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["glycolysis"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["glycolysis"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### OxiPhos

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "OxiPhos")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "OxiPhos")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["OxiPhos"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=6, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["OxiPhos"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### Sweet

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "Sweet")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "Sweet")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["Sweet"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["Sweet"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### TaSUT

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "TaSUT")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "TaSUT")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["TaSUT"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["TaSUT"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### TaGST

Gene list got from:

> Wang, R., Ma, J., Zhang, Q., Wu, C., Zhao, H., Wu, Y., Yang, G., & He, G. (2019). Genome-wide identification and expression profiling of glutathione transferase gene family under multiple stresses and hormone treatments in wheat (Triticum aestivum L.). BMC Genomics, 20(1), 986. https://doi.org/10.1186/s12864-019-6374-x

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "TaGST")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "TaGST")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["TaGST"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r}
GSTs <- readxl::read_xlsx(
    path = "../data/12864_2019_6374_MOESM2_ESM.xlsx",
    skip = 3, col_names = c("Name", "ID"), range = "B4:C333") |>
  dplyr::mutate(ID = stringr::str_sub(ID, end = -3L))
```

```{r fig.height=20, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, GSTs$ID,
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types,
  row_labels = with(GSTs,
    Name[match(intersect(rownames(logfc_data$mtx), ID), ID)]
  ))
```

### Signaling {.tabset}

#### GOCaSignaling

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "GOCaSignaling")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "GOCaSignaling")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["GOCaSignaling"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["GOCaSignaling"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### MAPK

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "MAPK")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "MAPK")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["MAPK"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["MAPK"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

### Transcription Factors {.tabset}

#### Myb

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "Myb")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "Myb")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["Myb"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=20, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["Myb"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### NAC

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "NAC")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "NAC")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["NAC"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=14, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["NAC"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### WRKY

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "WRKY")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "WRKY")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["WRKY"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=14, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["WRKY"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

## Analysis of Target Genes

### FHB related genes

```{r fig.height=7, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["FHBGenes"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

## Analysis of Other RNA-seq results

### fusaoctaxin A {.tabset}

#### FA3FB3up

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "FA3FB3up")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "FA3FB3up")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["FA3FB3up"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=14, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["FA3FB3up"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### FA3hUp

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "FA3hUp")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "FA3hUp")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["FA3hUp"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=26, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["FA3hUp"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### FAFB-3hUp

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "FAFB-3hUp")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "FAFB-3hUp")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["FAFB-3hUp"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=6, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["FAFB-3hUp"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

### TIRH2Y

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "TIRH2Y")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "TIRH2Y")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["TIRH2Y"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["TIRH2Y"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```
