---
title: "Known Pathways in DEGs of Each Cel Types"
author: "Altair Wei"
date: "2023-03-30"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source("../scripts/LoadUtils.R", chdir = TRUE)
```

## Load Data

### Seurat Object

```{r}
obj <- readRDS(Sys.glob("../results/ObjectCache/IntegrationAndAnnotation/obj_annotated_*.rds"))
```

### Metadata

```{r}
cell_types <- c(
  "Stomata",
  "Epidermis",
  "Chlorenchyma",
  "Parenchyma",
  "Outer sheath",
  "Inner sheath",
  "Phloem",
  "Procambium"
)
```

### Gene Lists

```{r}
filelist <- Sys.glob("../data/pathways/*.txt")
names(filelist) <- basename(filelist) |> stringr::str_remove(stringr::fixed(".txt"))

pathways <- lapply(filelist, function(filename) {
  genes <- stringr::str_trim(readLines(filename))
  genes[genes != ""]
})
```

```{r}
summary(pathways)[, 1, drop = FALSE]
```

```{r}
pathways |>
  tibble::enframe(name = "pathway", value = "gene") |>
  tidyr::unnest(cols = c(gene)) |>
  dplyr::group_by(gene) |>
  dplyr::summarise(pathways = paste(pathway, collapse = " | ")) |>
  download_table(
    table_id = "gene_pathways",
    filterable = TRUE,
    searchable = TRUE,
    showPageSizeOptions = TRUE)
```


### DEGs of Cell Types

```{r}
ds_res <- readRDS(Sys.glob(
  "../results/ObjectCache/DifferentialState/ds_res_*.rds"))
```

```{r}
logfc_data <- xfun::cache_rds(
  file = "ds_deg_logfc_data.rds",
  dir = "../results/ObjectCache/DifferentialState/",
  expr = local({
    logfc_mtx <- ds_res$DESeq2 |>
      rhapsodykit::diff_state_significant(fdr_limit = 0.05, logfc_limit = 0) |>
      rhapsodykit::diff_state_format() |>
      dplyr::filter(!stringr::str_detect(contrast, "X\\dDPI\\.PNR2-X\\dDPI\\.TR4")) |>
      dplyr::select(gene, cluster_id, contrast, logFC) |>
      dplyr::mutate(
        time = stringr::str_extract(contrast, "\\dDPI"),
        contrast = stringr::str_replace(contrast,
          "X(\\dDPI)\\.(.*)-X(\\dDPI)\\.(.*)", "\\2-\\4")) |>
      tidyr::pivot_wider(names_from = "time", values_from = "logFC", values_fill = NA) |>
      tidyr::pivot_longer(cols = dplyr::ends_with("DPI"), names_to = "time", values_to = "logFC") |>
      tidyr::pivot_wider(names_from = c(cluster_id, contrast, time),
                         names_sep = "#",
                         values_from = logFC) |>
      tibble::column_to_rownames("gene") |>
      as.matrix()

    str_coldata <- strsplit(colnames(logfc_mtx), "#")

    mtx_coldata <- data.frame(
      colidx = seq_along(colnames(logfc_mtx)),
      cell_types = factor(sapply(str_coldata, function(x) x[[1]]),
                          levels = cell_types),
      contrast = factor(sapply(str_coldata, function(x) x[[2]]),
                        levels = c("PNR2-MOCK", "TR4-MOCK")),
      times = factor(sapply(str_coldata, function(x) x[[3]]),
                     levels = c("1DPI", "2DPI", "3DPI"))
    )

    mtx_coldata <- mtx_coldata[with(mtx_coldata, order(cell_types, contrast, times)),]

    logfc_mtx <- logfc_mtx[, mtx_coldata$colidx]

    ha = ComplexHeatmap::HeatmapAnnotation(
      which = "column",
      df = mtx_coldata[-1],
      col = list(
        cell_types = structure(
          ggthemes::tableau_color_pal()(
            length(levels(mtx_coldata$cell_types))),
          names = levels(mtx_coldata$cell_types)
        ),
        contrast = structure(
          c("#E41A1C", "#377EB8"),
          names = levels(mtx_coldata$contrast)
        ),
        times = structure(
          RColorBrewer::brewer.pal(3, "YlGn"),
          names = levels(mtx_coldata$times)
        )
      ))

    list(mtx = logfc_mtx, coldata = mtx_coldata, heatanno = ha)
  })
)
```

### Pseudobulk CPM

```{r}
pseudobulk_cpm_mean <- readRDS(Sys.glob(
  "../results/ObjectCache/Pseudobulk/pseudobulk_cpm_mean_*.rds"))
```

```{r}
cpm_se <- local({
  cpm_df <- as.data.frame(pseudobulk_cpm_mean) |>
      tibble::rownames_to_column("Gene") |>
      tidyr::pivot_longer(-Gene, names_to = "bulk", values_to = "cpm") |>
      tidyr::separate(bulk, sep = "\\.", into = c("treatment", "time", "cellType"))
  
  cpm_df_0 <- dplyr::filter(cpm_df, time == "0DPI")
  
  cpm_df <- cpm_df |>
    dplyr::add_row(dplyr::mutate(cpm_df_0, treatment = "PNR2")) |>
    dplyr::add_row(dplyr::mutate(cpm_df_0, treatment = "TR4"))
  
  cpm_df$cellType <- factor(cpm_df$cellType, levels = TISSUE_TYPES)
  cpm_df$treatment <- factor(cpm_df$treatment, levels = c("MOCK", "PNR2", "TR4"))
  cpm_df$time <- factor(cpm_df$time, levels = c("0DPI", "1DPI", "2DPI", "3DPI"))
  
  cpm_df <- cpm_df |>
    dplyr::select(cellType, treatment, time, Gene, cpm) |>
    dplyr::arrange(cellType, treatment, time)
  
  cpm_coldata <- cpm_df |>
    dplyr::group_by(cellType, treatment, time) |>
    dplyr::summarise(.groups = "drop")
  
  cpm_mtx <- cpm_df |>
    tidyr::pivot_wider(names_from = c(cellType, treatment, time),
                       values_from = cpm, names_sep = "#") |>
    tibble::column_to_rownames("Gene") |>
    as.matrix()
  
  SummarizedExperiment(
    assays = list(
      raw = cpm_mtx,
      scaled = t(scale(t(cpm_mtx)))),
    colData = cpm_coldata)
})
```

### Calculate GSVA

```{r}
gsva_df <- xfun::cache_rds(
  file = "pseudobulk_8celltypes_gsva_df.rds",
  dir = "../results/ObjectCache/Pseudobulk/",
  rerun = TRUE,
  expr = local({
    expr_gsva <- GSVA::gsva(
      pseudobulk_cpm_mean, pathways,
      method = "gsva", kcdf = "Gaussian",
      verbose = FALSE)

    gsva_df <- as.data.frame(expr_gsva) |>
        tibble::rownames_to_column("Pathway") |>
        tidyr::pivot_longer(-Pathway, names_to = "bulk", values_to = "ES") |>
        tidyr::separate(bulk, sep = "\\.", into = c("treatment", "time", "cellType"))

    gsva_df$cellType <- factor(gsva_df$cellType, levels = cell_types)

    gsva_df
  })
)
```

### Functions

```{r}
printDEGNum <- function(mtx, coldata, genes) {
  detected <- intersect(rownames(mtx), genes)
  mtx <- mtx[detected,]
  coldata$UP <- colSums(mtx > 0, na.rm = TRUE)
  coldata$DOWN <- colSums(mtx < 0, na.rm = TRUE)
  dplyr::select(coldata, cell_types, contrast, times, UP, DOWN) |>
    tibble::remove_rownames()
}
```

## Analysis Based on Known Pathways

### Phytosynthesis related {.tabset}

```{r}
with(pathways, {
  eulerr::venn(list(
    GOCC_Chloroplast = GOCC_Chloroplast,
    PhotosynAnt = PhotosynAnt,
    PhotosynCfix = PhotosynCfix,
    PhotosynPSEA = PhotosynPSEA)) |>
  plot()
})
```


#### PhotosynAnt

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "PhotosynAnt")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "PhotosynAnt")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["PhotosynAnt"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=6, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["PhotosynAnt"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### PhotosynCfix

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "PhotosynCfix")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "PhotosynCfix")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["PhotosynCfix"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=14, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["PhotosynCfix"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### PhotosynPSEA

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "PhotosynPSEA")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "PhotosynPSEA")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["PhotosynPSEA"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=14, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["PhotosynPSEA"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### GOCC_Chloroplast

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "GOCC_Chloroplast")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "GOCC_Chloroplast")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["GOCC_Chloroplast"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=7, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["GOCC_Chloroplast"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types,
  show_row_names = FALSE, asterisk = FALSE)
```

### Cell wall related {.tabset}

#### Aromatic amino acid

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "Aromatic amino acid")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "Aromatic amino acid")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["Aromatic amino acid"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=6, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["Aromatic amino acid"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### General phenylpropanoid

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "General phenylpropanoid")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "General phenylpropanoid")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["General phenylpropanoid"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=8, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["General phenylpropanoid"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### Lignin biosynthesis

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "Lignin biosynthesis")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "Lignin biosynthesis")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["Lignin biosynthesis"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=6, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["Lignin biosynthesis"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### Flavanones

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "Flavanones")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "Flavanones")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["Flavanones"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["Flavanones"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### HCAAs

> Kage, U., Karre, S., Kushalappa, A. C., & McCartney, C. (2017). Identification and characterization of a fusarium head blight resistance gene TaACT in wheat QTL-2DL. Plant Biotechnology Journal, 15(4), 447–457.

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "HCAAs")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "HCAAs")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["HCAAs"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["HCAAs"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

```{r}
Seurat::FeaturePlot(
    obj, features = "TraesCS2D02G490400",
    reduction = "umap", order = TRUE) +
  ggplot2::coord_fixed() +
  theme_dimred()
```

```{r}
Seurat::FeaturePlot(
    obj, features = "TraesCS2A02G490000",
    reduction = "umap", order = TRUE) +
  ggplot2::coord_fixed() +
  theme_dimred()
```

```{r}
Seurat::FeaturePlot(
    obj, features = "TraesCS2B02G518300",
    reduction = "umap", order = TRUE) +
  ggplot2::coord_fixed() +
  theme_dimred()
```

#### Suberin

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "Suberin")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "Suberin")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["Suberin"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["Suberin"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### CalloseSyn

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "CalloseSyn")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "CalloseSyn")
```

```{r fig.height=4, fig.width=14}
detected <- intersect(rownames(logfc_data$mtx), pathways[["CalloseSyn"]])
print(detected)
logfc_data$mtx[detected,][!is.na(logfc_data$mtx[detected,])]
```

#### Polyamine

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "Polyamine")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "Polyamine")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["Polyamine"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=6, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["Polyamine"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

### Immune related {.tabset}

#### CCNBLRR

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "CCNBLRR")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "CCNBLRR")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["CCNBLRR"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["CCNBLRR"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### NBLRR

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "NBLRR")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "NBLRR")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["NBLRR"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=10, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["NBLRR"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### TIR

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "TIR")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "TIR")
```

```{r fig.height=4, fig.width=14}
detected <- intersect(rownames(logfc_data$mtx), pathways[["TIR"]])
print(detected)
logfc_data$mtx[detected,][!is.na(logfc_data$mtx[detected,])]
```

#### PR

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "PR")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "PR")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["PR"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r}
TaPRs <- c(
  "PR1-like" = "TraesCS2A02G476800",
  "PR1-like" = "TraesCS2D02G317800",
  "PR1-like" = "TraesCS3A02G291000",
  "PR1-like" = "TraesCS3D02G290800",
  "PR1-like" = "TraesCS4A02G116400",
  "PR1-like" = "TraesCS4A02G251300",
  "PR1-like" = "TraesCS4B02G063600",
  "PR1-like" = "TraesCS4D02G062500",
  "PR1-like" = "TraesCS4D02G189200",
  #"PR1-like" = "TraesCS5D02G259800LC",
  "PR2-like" = "TraesCS6B02G188000",
  "PR2-like" = "TraesCS7B02G279900",
  "PR3-like" = "TraesCS2A02G033600",
  "PR3-like" = "TraesCS3D02G260300",
  "PR4-like" = "TraesCS2A02G551000",
  "PR4-like" = "TraesCS2D02G552100",
  "PR4-like" = "TraesCS3A02G517100",
  "PR4-like" = "TraesCS3D02G524700",
  "PR5-like" = "TraesCS2A02G290300",
  "PR5-like" = "TraesCS2B02G306800",
  "PR5-like" = "TraesCS2D02G288200",
  "PR5-like" = "TraesCS3A02G357800",
  "PR5-like" = "TraesCS3B02G390500",
  "PR5-like" = "TraesCS3D02G351700",
  "PR5-like" = "TraesCS4A02G177300",
  "PR5-like" = "TraesCS4B02G140500",
  "PR5-like" = "TraesCS4D02G135200",
  "PR5-like" = "TraesCS5A02G328500",
  "PR5-like" = "TraesCS5B02G358300",
  "PR5-like" = "TraesCS5D02G334400",
  "PR5-like" = "TraesCS6A02G129400",
  "PR5-like" = "TraesCS6B02G157700",
  "PR5-like" = "TraesCS6D02G118700",
  "PR5-like" = "TraesCS7A02G253800",
  "PR5-like" = "TraesCS7D02G252400",
  "PR14-like" = "TraesCS3B02G063400",
  "PR14-like" = "TraesCS3B02G063500",
  "PR14-like" = "TraesCS3B02G063600",
  "PR14-like" = "TraesCS3B02G063700"
)

TaPRs <- tibble::enframe(TaPRs, name = "Name", value = "ID")
```

```{r fig.height=12, fig.width=14}
plotExpressionHeatmap(
  cpm_se[rownames(cpm_se) %in% TaPRs$ID,],
  col_split_by = "cellType",
  col_anno_by = c("treatment", "time"),
  show_row_names = TRUE,
  row_labels = with(TaPRs, {
    Name[match(rownames(cpm_se)[rownames(cpm_se) %in% ID], ID)]
  })
)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, TaPRs$ID,
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types,
  row_labels = with(TaPRs,
    Name[match(intersect(rownames(logfc_data$mtx), ID), ID)]
  )
)
```

#### RLK

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "RLK")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "RLK")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["RLK"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=20, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["RLK"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### RBOH

Gene list got from:

> Reference: Hu, C.-H., Wei, X.-Y., Yuan, B., Yao, L.-B., Ma, T.-T., Zhang, P.-P., Wang, X., Wang, P.-Q., Liu, W.-T., Li, W.-Q., Meng, L.-S., & Chen, K.-M. (2018). Genome-Wide Identification and Functional Analysis of NADPH Oxidase Family Genes in Wheat During Development and Environmental Stress Responses. Frontiers in Plant Science, 9. https://doi.org/10.3389/fpls.2018.00906

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "RBOH")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "RBOH")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["RBOH"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r}
RBOH <- c(
  `TaNOX1-1AL` = "TraesCS1A02G290700",
  `TaNOX1-1BL` = "TraesCS1B02G300000",
  `TaNOX1-1DL` = "TraesCS1D02G289300",
  `TaNOX2-1AL` = "TraesCS1A02G347700",
  `TaNOX2-1BL` = "TraesCS1B02G362200",
  `TaNOX2-1DL` = "TraesCS1D02G350600",
  `TaNOX3-1BL` = "TraesCS1B02G295200",
  `TaNOX3-1DL` = "TraesCS1D02G284900",
  `TaNOX5-3AL` = "TraesCS3A02G354200",
  `TaNOX5-3BL` = "TraesCS3B02G386600",
  `TaNOX5-3DL` = "TraesCS3D02G347900",
  `TaNOX6-3AL` = "TraesCS3A02G280200",
  `TaNOX6-3BL` = "TraesCS3B02G314000",
  `TaNOX6-3DL` = "TraesCS3D02G279900",
  `TaNOX7-3AS` = "TraesCS3A02G182900",
  `TaNOX7-3AS` = "TraesCS3D02G187300",
  `TaNOX7-3AS` = "TraesCS3B02G212900",
  `TaNOX8-4BL` = "TraesCS4B02G358800",
  `TaNOX8-4DL` = "TraesCS4D02G352200",
  `TaNOX13-5AL` = "TraesCS5A02G527600",
  `TaNOX9-4BL` = "TraesCS4B02G327800",
  `TaNOX9-4DL` = "TraesCS4D02G324800",
  `TaNOX11-5AL` = "TraesCS5A02G499900",
  `TaNOX10-5AL` = "TraesCS5A02G301700",
  `TaNOX10-5BL` = "TraesCS5B02G299000",
  `TaNOX10-5BL` = "TraesCS5D02G306400",
  `TaNOX12-5AL` =	"TraesCS5A02G211800",
  `TaNOX12-5BL` =	"TraesCS5B02G212100",
  `TaNOX12-5DL` =	"TraesCS5D02G222100",
  `TaNOX14-5AS` =	"TraesCS5A02G093600",
  `TaNOX14-5BS` =	"TraesCS5B02G099700",
  `TaNOX14-5DS` =	"TraesCS5D02G105900",
  `TaNOX15-6AS` =	"TraesCS6A02G180600",
  `TaNOX15-6BS` =	"TraesCS6B02G205000",
  `TaNOX-like4-6DS` = "TraesCS6D02G164500",
  #`TaNOX-like1-1AL` = "TraesCS1B02G637700LC",
  #`TaNOX-like3-1BL` = "TraesCS1B02G637700LC",
  `TaNOX-like2-1BL` = "TraesCS1B02G295300",
  `TaNOX-like2-1BL` = "TraesCS1D02G285000",
  `TaFRO1-1AL` = "TraesCS1A02G284600",
  `TaFRO1-1BL` = "TraesCS1B02G293900",
  `TaFRO1-1DL` = "TraesCS1D02G283600",
  `TaFRO2-1BL` = "TraesCS1B02G295100",
  `TaNOX4-1DL` = "TraesCS1D02G284800",
  `TaNOX4-1DL` = "TraesCS1A02G285700",
  `TaFRO3-2AL` = "TraesCS2A02G312600",
  `TaFRO3-2BL` = "TraesCS2B02G329600",
  `TaFRO3-2DL` = "TraesCS2D02G311100",
  `TaFRO4-2AL` = "TraesCS2A02G396000",
  `TaFRO4-2BL` = "TraesCS2B02G414300",
  `TaFRO4-2DL` = "TraesCS2D02G393700"
)

RBOH <- tibble::enframe(RBOH, name = "Name", value = "ID")
```

```{r fig.height=8, fig.width=14}
plotExpressionHeatmap(
  cpm_se[rownames(cpm_se) %in% RBOH$ID,],
  col_split_by = "cellType",
  col_anno_by = c("treatment", "time"),
  show_row_names = TRUE,
  row_labels = with(RBOH, {
    Name[match(rownames(cpm_se)[rownames(cpm_se) %in% ID], ID)]
  })
)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, RBOH$ID,
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types,
  row_labels = with(RBOH,
    Name[match(intersect(rownames(logfc_data$mtx), ID), ID)]
  )
)
```

#### TaNPR

```{r}
TaNPRs <- c(
  "TaNPR1-A" = "TraesCS3A02G105400",
  "TaNPR1-B" = "TraesCS3B02G123800",
  "TaNPR1-D" = "TraesCS3D02G107500",
  "TaNPR2-A" = "TraesCS4A02G470500",
  "TaNPR2-D" = "TraesCS7D02G019000",
  "TaNPR3-A" = "TraesCS3A02G298800",
  "TaNPR3-B" = "TraesCS3B02G337700",
  "TaNPR3-D" = "TraesCS3D02G302900",
  "TaNPR4-A" = "TraesCS4A02G294400",
  "TaNPR4-B" = "TraesCS4B02G018900",
  "TaNPR4-D" = "TraesCS4D02G017500",
  "TaNPR5-A" = "TraesCS3A02G489000",
  "TaNPR5-B" = "TraesCS3B02G537400",
  "TaNPR5-D" = "TraesCS3D02G484100",
  "TaNPR6-A" = "TraesCS5A02G134700",
  "TaNPR6-B" = "TraesCS5B02G133700",
  "TaNPR6-D" = "TraesCS5D02G139600"
)

TaNPRs <- tibble::enframe(TaNPRs, name = "Name", value = "ID")
```

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "TaNPR")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "TaNPR")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["TaNPR"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotExpressionHeatmap(
  cpm_se[rownames(cpm_se) %in% TaNPRs$ID,],
  col_split_by = "cellType",
  col_anno_by = c("treatment", "time"),
  show_row_names = TRUE,
  row_labels = with(TaNPRs, {
    Name[match(rownames(cpm_se)[rownames(cpm_se) %in% ID], ID)]
  })
)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, TaNPRs$ID,
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types,
  row_labels = with(TaNPRs,
    Name[match(intersect(rownames(logfc_data$mtx), ID), ID)]
  )
)
```

#### RxNonly

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "RxNonly")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "RxNonly")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["RxNonly"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["RxNonly"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### GOSAR

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "GOSAR")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "GOSAR")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["GOSAR"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["GOSAR"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### NHP

```{r}
NHPs <- list(
  "FMO-like" = c(
    "TraesCS1D02G213900",
    "TraesCS2A02G071500",
    "TraesCS2A02G071600",
    "TraesCS2D02G070400",
    "TraesCS3A02G010900",
    "TraesCS3D02G000800",
    "TraesCS4A02G313200",
    "TraesCS4B02G366800",
    "TraesCS4B02G370200",
    "TraesCS4D02G269000",
    "TraesCS4D02G361000",
    "TraesCS5A02G348100",
    "TraesCS5A02G348400",
    "TraesCS5A02G349300",
    "TraesCS5B02G194300",
    "TraesCS5B02G349200",
    "TraesCS5D02G354600",
    "TraesCS6A02G150700",
    "TraesCS6B02G178800",
    "TraesCS6D02G139800",
    "TraesCS7A02G552000",
    "TraesCS7D02G060000"),
  "ALD1-like" = c(
    "TraesCS1B02G115200",
    "TraesCS3A02G310300",
    "TraesCS4A02G043800",
    "TraesCS4A02G116000",
    "TraesCS4B02G188200",
    "TraesCS4B02G264400",
    "TraesCS4B02G264500",
    "TraesCS4D02G189600",
    "TraesCS4D02G264500",
    "TraesCS5A02G236100",
    "TraesCS5B02G234400",
    "TraesCS5D02G243100"),
  "SARD4-like" = c(
    "TraesCS1A02G184900",
    "TraesCS1B02G192900",
    "TraesCS1D02G191900")
)

NHPs <- tibble::enframe(NHPs, name = "Name", value = "ID") |>
  tidyr::unnest(cols = ID)
```


```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "NHP")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "NHP")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["NHP"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=6, fig.width=14}
plotExpressionHeatmap(
  cpm_se[rownames(cpm_se) %in% NHPs$ID,],
  col_split_by = "cellType",
  col_anno_by = c("treatment", "time"),
  show_row_names = TRUE,
  row_labels = with(NHPs, {
    Name[match(rownames(cpm_se)[rownames(cpm_se) %in% ID], ID)]
  })
)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, NHPs$ID,
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types,
  row_labels = with(NHPs,
    Name[match(intersect(rownames(logfc_data$mtx), ID), ID)]
  )
)
```

### Metabolism related {.tabset}

#### gluconeogenesis

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "gluconeogenesis")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "gluconeogenesis")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["gluconeogenesis"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["gluconeogenesis"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### glycolysis

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "glycolysis")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "glycolysis")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["glycolysis"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["glycolysis"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### OxiPhos

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "OxiPhos")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "OxiPhos")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["OxiPhos"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=6, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["OxiPhos"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### Sweet

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "Sweet")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "Sweet")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["Sweet"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["Sweet"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### TaSUT

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "TaSUT")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "TaSUT")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["TaSUT"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["TaSUT"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### TaGST

Gene list got from:

> Wang, R., Ma, J., Zhang, Q., Wu, C., Zhao, H., Wu, Y., Yang, G., & He, G. (2019). Genome-wide identification and expression profiling of glutathione transferase gene family under multiple stresses and hormone treatments in wheat (Triticum aestivum L.). BMC Genomics, 20(1), 986. https://doi.org/10.1186/s12864-019-6374-x

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "TaGST")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "TaGST")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["TaGST"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r}
GSTs <- readxl::read_xlsx(
    path = "../data/12864_2019_6374_MOESM2_ESM.xlsx",
    skip = 3, col_names = c("Name", "ID"), range = "B4:C333") |>
  dplyr::mutate(ID = stringr::str_sub(ID, end = -3L))
```

```{r fig.height=20, fig.width=14}
plotExpressionHeatmap(
  cpm_se[rownames(cpm_se) %in% GSTs$ID,],
  col_split_by = "cellType",
  col_anno_by = c("treatment", "time"),
  show_row_names = FALSE,
  row_labels = with(GSTs, {
    Name[match(rownames(cpm_se)[rownames(cpm_se) %in% ID], ID)]
  })
)
```

```{r fig.height=20, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, GSTs$ID,
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types,
  row_labels = with(GSTs,
    Name[match(intersect(rownames(logfc_data$mtx), ID), ID)]
  ))
```

```{r fig.height=6, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, GSTs$ID,
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types,
  show_row_names = FALSE,
  asterisk = FALSE,
  row_labels = with(GSTs,
    Name[match(intersect(rownames(logfc_data$mtx), ID), ID)]
  ))
```

### Signaling {.tabset}

#### GOCaSignaling

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "GOCaSignaling")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "GOCaSignaling")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["GOCaSignaling"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["GOCaSignaling"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### MAPK

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "MAPK")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "MAPK")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["MAPK"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["MAPK"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

### Transcription Factors {.tabset}

#### Myb

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "Myb")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "Myb")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["Myb"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=20, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["Myb"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### NAC

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "NAC")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "NAC")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["NAC"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=14, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["NAC"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### WRKY

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "WRKY")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "WRKY")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["WRKY"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=14, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["WRKY"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### bZIP

> Liang, Y., Xia, J., Jiang, Y., Bao, Y., Chen, H., Wang, D., Zhang, D., Yu, J., & Cang, J. (2022). Genome-Wide Identification and Analysis of bZIP Gene Family and Resistance of TaABI5 (TabZIP96) under Freezing Stress in Wheat (Triticum aestivum). International Journal of Molecular Sciences, 23(4), 2351. https://doi.org/10.3390/ijms23042351

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "bZIP")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "bZIP")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["bZIP"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=7, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["bZIP"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### TaGRAS

> Liu, Y. & Wang, W. Characterization of the GRAS gene family reveals their contribution to the high adaptability of wheat. PeerJ 9, e10811 (2021).

```{r}
TaGRAS <- list(
  "DELLA" = c(
    "TraesCS1A02G413800",
    "TraesCS1B02G443800",
    "TraesCS1D02G421200",
    "TraesCS2A02G096800",
    "TraesCS2B02G112100",
    "TraesCS2D02G095300",
    "TraesCS3A02G233000",
    "TraesCS3B02G262400",
    "TraesCS3D02G220100",
    "TraesCS4A02G271000",
    "TraesCS4A02G466700",
    "TraesCS4B02G043100",
    "TraesCS4B02G342800",
    "TraesCS4D02G040400",
    "TraesCS4D02G338800",
    "TraesCS5A02G511900"
  ),
  "DLT" = c(
    "TraesCS4A02G430600",
    "TraesCS7A02G059000",
    "TraesCS7D02G053600"
  ),
  "HAM" = c(
    "TraesCS1A02G209400",
    "TraesCS1B02G223400",
    "TraesCS1D02G212800",
    "TraesCS4A02G088500",
    "TraesCS4A02G209800",
    "TraesCS4A02G482000",
    "TraesCS4B02G113300",
    "TraesCS4B02G215700",
    "TraesCS4B02G367300",
    "TraesCS4D02G110800",
    "TraesCS4D02G216200",
    "TraesCS4D02G360200",
    "TraesCS5A02G535000",
    "TraesCS6A02G247200",
    "TraesCS6B02G277400",
    "TraesCS6D02G229400",
    "TraesCS7A02G014200",
    "TraesCS7D02G011400"
  ),
  "LAS" = c(
    "TraesCS6A02G159500",
    "TraesCS6B02G191800",
    "TraesCS6D02G153000",
    "TraesCS7A02G382800",
    "TraesCS7B02G285500",
    "TraesCS7D02G379200"
  ),
  "LISCL" = c(
    "TraesCS1A02G030200",
    "TraesCS1A02G030300",
    "TraesCS1B02G037200",
    "TraesCS1B02G037300",
    "TraesCS1B02G037400",
    "TraesCS1B02G037500",
    "TraesCS1D02G030800",
    "TraesCS1D02G030900",
    "TraesCS1D02G031000",
    "TraesCS1D02G031200",
    "TraesCS3A02G067900",
    "TraesCS3A02G068000",
    "TraesCS3A02G360300",
    "TraesCS3B02G080900",
    "TraesCS3B02G081000",
    "TraesCS3B02G081100",
    "TraesCS3B02G392600",
    "TraesCS3D02G067700",
    "TraesCS3D02G067800",
    "TraesCS3D02G067900",
    "TraesCS3D02G354100",
    "TraesCS4A02G176200",
    "TraesCS4A02G176300",
    "TraesCS4A02G176400",
    "TraesCS4A02G176600",
    "TraesCS4A02G176500",
    "TraesCS4A02G176700",
    "TraesCS4A02G177100",
    "TraesCS4B02G140800",
    "TraesCS4B02G141100",
    "TraesCS4B02G141200",
    "TraesCS4B02G141300",
    "TraesCS4B02G141400",
    "TraesCS4B02G141500",
    "TraesCS4B02G336200",
    "TraesCS4D02G135500",
    "TraesCS4D02G135900",
    "TraesCS4D02G136000",
    "TraesCS4D02G136100",
    "TraesCS4D02G136200",
    "TraesCS4D02G136300",
    "TraesCS4D02G136400",
    "TraesCS4D02G136500",
    "TraesCS4D02G332000",
    "TraesCS5A02G080300",
    "TraesCS5A02G139000",
    "TraesCS5A02G481300",
    "TraesCS5A02G505900",
    "TraesCS5A02G506000",
    "TraesCS5A02G506100",
    "TraesCS5B02G138100",
    "TraesCS5B02G494500",
    "TraesCS5D02G092000",
    "TraesCS5D02G152700",
    "TraesCS5D02G494900",
    "TraesCS7A02G173000",
    "TraesCS7B02G077800",
    "TraesCS7B02G078300",
    "TraesCS7D02G174100",
    "TraesCS7D02G174300"
  ),
  "Os19" = c(
    "TraesCS2A02G304000",
    "TraesCS2B02G320700",
    "TraesCS2D02G302600"
  ),
  "Os4" = c(
    "TraesCS3A02G392400",
    "TraesCS3A02G392500",
    "TraesCS3A02G392600",
    "TraesCS3B02G424500",
    "TraesCS3B02G424600",
    "TraesCS3D02G385400",
    "TraesCS3D02G385500"
  ),
  "PAT1" = c(
    "TraesCS2A02G189600",
    "TraesCS2A02G214400",
    "TraesCS2A02G395400",
    "TraesCS2B02G217500",
    "TraesCS2B02G239400",
    "TraesCS2B02G413500",
    "TraesCS2D02G198200",
    "TraesCS2D02G220200",
    "TraesCS2D02G393200",
    "TraesCS3A02G408000",
    "TraesCS3B02G441600",
    "TraesCS3D02G403100",
    "TraesCS4A02G039000",
    "TraesCS4A02G465000",
    "TraesCS4B02G266700",
    "TraesCS4D02G266400",
    "TraesCS5A02G409800",
    "TraesCS5A02G414000",
    "TraesCS5D02G419400"
  ),
  "SCL3" = c(
    "TraesCS2A02G532900",
    "TraesCS2B02G562400",
    "TraesCS2D02G055700",
    "TraesCS2D02G534700",
    "TraesCS3A02G487500",
    "TraesCS3B02G535100",
    "TraesCS3D02G482600",
    "TraesCS4A02G185900",
    "TraesCS4D02G127800",
    "TraesCS5A02G135000",
    "TraesCS5B02G133800",
    "TraesCS5D02G139500",
    "TraesCS6D02G055100",
    "TraesCSU02G014000"
  ),
  "SCL4/7" = c(
    "TraesCS4A02G260600",
    "TraesCS4B02G054000",
    "TraesCS4D02G054000"
  ),
  "SCR" = c(
    "TraesCS1A02G304000",
    "TraesCS1B02G314500",
    "TraesCS1D02G303500",
    "TraesCS2A02G200700",
    "TraesCS2B02G228000",
    "TraesCS2D02G208500",
    "TraesCS4A02G191300",
    "TraesCS4B02G124000",
    "TraesCS4D02G122000",
    "TraesCS5A02G144000",
    "TraesCS5B02G143100",
    "TraesCS5D02G148000"
  ),
  "SHR" = c(
    "TraesCS1A02G319700",
    "TraesCS1B02G332000",
    "TraesCS1D02G319500",
    "TraesCS2A02G192600",
    "TraesCS2A02G194600",
    "TraesCS2A02G460200",
    "TraesCS2B02G212300",
    "TraesCS2B02G214600",
    "TraesCS2B02G481900",
    "TraesCS2D02G193300",
    "TraesCS2D02G194800",
    "TraesCS2D02G460500"
  )
)

TaGRAS <- tibble::enframe(TaGRAS, name = "Name", value = "ID") |>
  tidyr::unnest(cols = ID)
```

```{r fig.height=12, fig.width=14}
plotExpressionHeatmap(
  cpm_se[rownames(cpm_se) %in% TaGRAS$ID,],
  col_split_by = "cellType",
  col_anno_by = c("treatment", "time"),
  show_row_names = TRUE,
  row_labels = with(TaGRAS, {
    Name[match(rownames(cpm_se)[rownames(cpm_se) %in% ID], ID)]
  })
)
```

```{r fig.height=7, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, TaGRAS$ID,
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types,
  row_labels = with(TaGRAS,
    Name[match(intersect(rownames(logfc_data$mtx), ID), ID)]
  )
)
```

#### WOX

```{r}
WOX <- c(
  "TaWox1-A" = "TraesCS1A02G052000",
  "TaWox1-B" = "TraesCS1B02G069000",
  "TaWox1-D" = "TraesCS1D02G054000",
  "TaWox2-A" = "TraesCS1A02G399400",
  "TaWox2-B" = "TraesCS1B02G427400",
  "TaWox2-D" = "TraesCS1D02G406900",
  "TaWox3-A" = "TraesCS2A02G100700",
  "TaWox3-B" = "TraesCS2B02G117900",
  "TaWox3-D" = "TraesCS2D02G100200",
  "TaWox4-A" = "TraesCS2A02G491900",
  "TaWox5-A" = "TraesCS2A02G514000",
  "TaWox5-B" = "TraesCS2B02G542600",
  "TaWox5-D" = "TraesCS2D02G515600",
  "TaWox6-A" = "TraesCS3A02G073500",
  "TaWox6-B" = "TraesCS3B02G087800",
  "TaWox6-D" = "TraesCS3D02G073300",
  "TaWox7-A" = "TraesCS3A02G247200",
  "TaWox7-B" = "TraesCS3B02G272200",
  "TaWox7-D" = "TraesCS3D02G244300",
  "TaWox8-A" = "TraesCS3A02G341700",
  "TaWox8-B" = "TraesCS3B02G373800",
  "TaWox8-D" = "TraesCS3D02G335500",
  "TaWox8-Un" = "TraesCSU02G204800",
  "TaWox9-A" = "TraesCS3A02G358100",
  "TaWox9-B" = "TraesCS3B02G391100",
  "TaWox9-D" = "TraesCS3D02G352500",
  "TaWox10-A1" = "TraesCS3A02G358200",
  "TaWox10-A2" = "TraesCS3A02G358400",
  "TaWox10-B" = "TraesCS3B02G391200",
  "TaWox10-D1" = "TraesCS3D02G352600",
  "TaWox10-D2" = "TraesCS3D02G352700",
  "TaWox11-A" = "TraesCS3A02G368100",
  "TaWox11-B" = "TraesCS3B02G399800",
  "TaWox11-D" = "TraesCS3D02G361100",
  "TaWox12-A" = "TraesCS4A02G130200",
  "TaWox12-B" = "TraesCS4B02G174400",
  "TaWox12-D" = "TraesCS4D02G176400",
  "TaWox13-A" = "TraesCS5A02G085000",
  "TaWox13-B" = "TraesCS5B02G091000",
  "TaWox13-D" = "TraesCS5D02G097400",
  "TaWox14-A" = "TraesCS5A02G157300",
  "TaWox14-B" = "TraesCS5B02G156400",
  "TaWox14-D" = "TraesCS5D02G162600"
)

WOX <- tibble::enframe(WOX, name = "Name", value = "ID")
```

```{r fig.height=5, fig.width=14}
plotExpressionHeatmap(
  cpm_se[rownames(cpm_se) %in% WOX$ID,],
  col_split_by = "cellType",
  col_anno_by = c("treatment", "time"),
  show_row_names = TRUE,
  row_labels = with(WOX, {
    Name[match(rownames(cpm_se)[rownames(cpm_se) %in% ID], ID)]
  })
)
```

```{r}
intersect(logfc_data$mtx, WOX$ID)
```


### Response to Phytohormones {.tabset}

From Gene Ontology

#### JA

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "JA-resp")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "JA-resp")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["JA-resp"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=7, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["JA-resp"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### SA

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "SA-resp")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "SA-resp")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["SA-resp"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=7, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["SA-resp"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### Auxin

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "Auxin-resp")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "Auxin-resp")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["Auxin-resp"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=12, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["Auxin-resp"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### ABA

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "ABA-resp")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "ABA-resp")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["ABA-resp"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=12, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["ABA-resp"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### ET

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "ET-resp")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "ET-resp")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["ET-resp"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=12, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["ET-resp"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

### Phytohormones Biosynthesis {.tabset}

From Gene Ontology

#### JA

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "JA-syn")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "JA-syn")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["JA-syn"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=7, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["JA-syn"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### SA

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "SA-syn")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "SA-syn")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["SA-syn"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=7, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["SA-syn"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### Auxin

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "Auxin-syn")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "Auxin-syn")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["Auxin-syn"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=7, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["Auxin-syn"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### ABA

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "ABA-syn")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "ABA-syn")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["ABA-syn"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=7, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["ABA-syn"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### ET

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "ET-syn")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "ET-syn")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["ET-syn"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=7, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["ET-syn"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

## Analysis of Target Genes

### FHB related genes

```{r}
FHB <- list(
  MIN7 = c(
    "TraesCS2A02G202900",
    "TraesCS2B02G230000",
    "TraesCS2D02G212800"
  ),
  TaAOS = c(
    "TraesCS4A02G061900",
    "TraesCS4D02G238800",
    "TraesCS4B02G237600"
  ),
  TaJRL53 = "TraesCS4A02G430200",
  TaLAC4 = "TraesCS3B02G392700",
  `TaNACL-D1` = "TraesCS5D02G111300",
  TaRBL = c(
    "TraesCS3A02G078200",
    "TraesCS3B02G093100",
    "TraesCS3D02G078700"
  ),
  TaSHMT3 = "TraesCS3A02G385600",
  TaUGT5 = "TraesCS2B02G184000",
  TaUGT6 = "TraesCS5B02G436300",
  TaUGT12887 = "TraesCS5B02G148300",
  TaWRKY70 = "TraesCS2D02G489700",
  `TaWAK2A-800` = "TraesCS2A02G071800",
  `UDP-UGT` = c(
    "TraesCS2B02G068600",
    "TraesCS2B02G068700"
  ),
  TaTIR1 = "TraesCS1A02G091300",
  TaSnRK1a = "TraesCS1A02G350500",
  `TaLRRK-6D` = "TraesCS6D02G206100",
  TaMPT = "TraesCS5A02G236700",
  TaSAM = "TraesCS2A02G048600",
  `TaSHMT3A-1` = "TraesCS3A02G385600",
  TaCYP72A = "TraesCS3A01G532600"
)

FHB <- tibble::enframe(FHB, name = "Name", value = "ID") |>
  tidyr::unnest(cols = ID)
```

```{r fig.height=5, fig.width=14}
plotExpressionHeatmap(
  cpm_se[rownames(cpm_se) %in% FHB$ID,],
  col_split_by = "cellType",
  col_anno_by = c("treatment", "time"),
  show_row_names = TRUE,
  row_labels = with(FHB, {
    Name[match(rownames(cpm_se)[rownames(cpm_se) %in% ID], ID)]
  })
)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, FHB$ID,
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types,
  row_labels = with(FHB,
    Name[match(intersect(rownames(logfc_data$mtx), ID), ID)]
  )
)
```

### Mesophyll development

> Loudya, N., Mishra, P., Takahagi, K., Uehara-Yamaguchi, Y., Inoue, K., Bogre, L., Mochida, K., & López-Juez, E. (2021). Cellular and transcriptomic analyses reveal two-staged chloroplast biogenesis underpinning photosynthesis build-up in the wheat leaf. Genome Biology, 22(1), 151. https://doi.org/10.1186/s13059-021-02366-3

#### Chloroplast division

```{r}
chloroplast_division <- c(
  "TraesCS3D02G157200",
  "TraesCS2D02G257600",
  "TraesCS2A02G266200",
  "TraesCS2B02G275900",
  "TraesCS2A02G488400",
  "TraesCS2B02G516000",
  "TraesCS2D02G488700",
  "TraesCS1A02G276500",
  "TraesCS1B02G285700",
  "TraesCS1D02G276000",
  "TraesCS2A02G294500",
  "TraesCS2B02G310900",
  "TraesCS2D02G292200",
  "TraesCS4A02G315500",
  "TraesCS5B02G563900",
  "TraesCS5D02G550700",
  "TraesCS5B02G120600",
  "TraesCS5D02G124800",
  "TraesCS5A02G114600",
  "TraesCS6A02G066200",
  "TraesCS6B02G089500",
  "TraesCSU02G117700",
  "TraesCS1A02G243700",
  "TraesCS1B02G255100",
  "TraesCS1D02G243700"
)
```

```{r fig.height=7, fig.width=14}
plotExpressionHeatmap(
  cpm_se[rownames(cpm_se) %in% chloroplast_division,],
  col_split_by = "cellType",
  col_anno_by = c("treatment", "time"),
  show_row_names = TRUE
)
```

```{r fig.height=7, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, chloroplast_division,
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

```{r fig.height=10, fig.width=7}
Seurat::VlnPlot(obj,
  features = c(
    "TraesCS4A02G315500",
    "TraesCS5B02G563900",
    "TraesCS5D02G550700"),
  ncol = 1
)
```

```{r}
Seurat::VlnPlot(obj,
  features = c(
    "TraesCS3D02G157200"),
  ncol = 1
) + Seurat::NoLegend()
```

#### Protein import

```{r}
protein_import <- c(
  "TraesCS4A02G095300",
  "TraesCS4B02G209200",
  "TraesCS4D02G209900",
  "TraesCS4A02G072800",
  "TraesCS4B02G228900",
  "TraesCS4D02G230000",
  "TraesCS1A02G208000",
  "TraesCS1B02G222000",
  "TraesCS1D02G211300",
  "TraesCS1A02G093900",
  "TraesCS1B02G122000",
  "TraesCS1D02G102500",
  "TraesCS5B02G514000",
  "TraesCS5D02G514500",
  "TraesCS4A02G358400",
  "TraesCS2A02G138100",
  "TraesCS2B02G162000",
  "TraesCS2D02G141100",
  "TraesCS6A02G158800",
  "TraesCS6D02G153700",
  "TraesCS6B02G192500",
  "TraesCS2D02G306800",
  "TraesCS2A02G308600",
  "TraesCS2B02G325100",
  "TraesCS2A02G199800",
  "TraesCS2B02G227100",
  "TraesCS2D02G207600",
  "TraesCS6A02G229400",
  "TraesCS6B02G252100",
  "TraesCS6D02G205400",
  "TraesCS1A02G165800",
  "TraesCS1B02G182500",
  "TraesCS1D02G174900",
  "TraesCS1A02G165700",
  "TraesCS1B02G182400",
  "TraesCS1D02G175000",
  "TraesCS5A02G206800",
  "TraesCS5B02G205200",
  "TraesCS5D02G213000",
  "TraesCS5A02G206900",
  "TraesCS5A02G257700",
  "TraesCS5B02G205300",
  "TraesCS5B02G257000",
  "TraesCS5D02G213100",
  "TraesCS1A02G133100",
  "TraesCS1B02G151300",
  "TraesCS1D02G131800",
  "TraesCS5A02G106200",
  "TraesCS5B02G111200",
  "TraesCS5D02G117600"
)
```

```{r fig.height=7, fig.width=14}
plotExpressionHeatmap(
  cpm_se[rownames(cpm_se) %in% protein_import,],
  col_split_by = "cellType",
  col_anno_by = c("treatment", "time"),
  show_row_names = TRUE
)
```

```{r fig.height=7, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, protein_import,
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

### Benzoxazinoid

> Batyrshina, Z. S., Shavit, R., Yaakov, B., Bocobza, S., & Tzin, V. (2022). The transcription factor TaMYB31 regulates the benzoxazinoid biosynthetic pathway in wheat. Journal of Experimental Botany, 73(16), 5634–5649. https://doi.org/10.1093/jxb/erac204

```{r}
BXDs <- list(
  TaMYB31 = c(
    "TraesCS5A02G227400",
    "TraesCS5B02G226100",
    "TraesCS5D02G234800"
  ),
  IGPS1 = c(
    "TraesCS5A02G169500",
    "TraesCS5B02G166300",
    "TraesCS5D02G173900"
  ),
  IGPS2 = c(
    "TraesCS2A02G335200",
    "TraesCS2B02G348500",
    "TraesCS2D02G329500"
  ),
  IGPS3 = c(
    "TraesCS7A02G512300",
    "TraesCS7B02G423900",
    "TraesCS7D02G501700"
  ),
  Bx1 = c(
    "TraesCS1A02G428400",
    "TraesCS1B02G463100",
    "TraesCS1D02G437700",
    "TraesCS3A02G041800",
    "TraesCS3B02G041900",
    "TraesCS3D02G038500",
    "TraesCS4A02G097400",
    "TraesCS4B02G207100",
    "TraesCS4D02G207900",
    "TraesCS5A02G440700",
    "TraesCS5B02G444300",
    "TraesCS7A02G392800",
    "TraesCS7B02G294800",
    "TraesCS7D02G388400",
    "TraesCS7A02G308600",
    "TraesCS7B02G208700",
    "TraesCS7D02G305300"
  ),
  Bx2 = c(
    "TraesCS4A02G097500",
    "TraesCS4B02G207000",
    "TraesCS4D02G207800",
    "TraesCSU02G048600"
  ),
  Bx3 = c(
    "TraesCS2A02G045600",
    "TraesCS2B02G057800",
    "TraesCS5A02G008900",
    "TraesCS5B02G007200",
    "TraesCS5D02G014300"
  ),
  Bx4 = c(
    "TraesCS2B02G030500",
    "TraesCS2A02G045900",
    "TraesCS2B02G058000",
    "TraesCS2D02G044600",
    "TraesCS5A02G008800",
    "TraesCS5B02G007100",
    "TraesCS5D02G014200",
    "TraesCS5A02G318100",
    "TraesCS5B02G318600",
    "TraesCS5D02G014300",
    "TraesCS5A02G472300",
    "TraesCS5B02G485200",
    "TraesCS5D02G484900"
  ),
  Bx5 = c(
    "TraesCS5A02G008700",
    "TraesCS5B02G030000",
    "TraesCS5B02G029900",
    "TraesCS5B02G007000",
    "TraesCS5D02G014100"
  ),
  Bx6 = c(
    "TraesCS2A02G051700",
    "TraesCS2B02G066000",
    "TraesCSU02G024200",
    "TraesCSU02G156500"
  ),
  "Bx8/9" = c(
    "TraesCS6A02G059300",
    "TraesCS6B02G079600",
    "TraesCS6A02G059700",
    "TraesCS6B02G080100",
    "TraesCSU02G093300",
    "TraesCSU02G095600",
    "TraesCS7B02G016800",
    "TraesCS7D02G116700"
  ),
  Bx10 = c(
    "TraesCS2A02G472400",
    "TraesCS2B02G495100",
    "TraesCS1B02G481300",
    "TraesCS1B02G481400",
    "TraesCS1B02G481500",
    "TraesCS4A02G463400",
    "TraesCS2B02G110700",
    "TraesCS2D02G093900"
  )
)

BXDs <- tibble::enframe(BXDs, name = "Name", value = "ID") |>
  tidyr::unnest(cols = ID)
```

```{r fig.height=10, fig.width=14}
plotExpressionHeatmap(
  cpm_se[rownames(cpm_se) %in% BXDs$ID,],
  col_split_by = "cellType",
  col_anno_by = c("treatment", "time"),
  show_row_names = TRUE,
  row_labels = with(BXDs, {
    Name[match(rownames(cpm_se)[rownames(cpm_se) %in% ID], ID)]
  })
)
```

```{r fig.height=7, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, BXDs$ID,
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types,
  row_labels = with(BXDs,
    Name[match(intersect(rownames(logfc_data$mtx), ID), ID)]
  )
)
```

### Programmed Cell Death

```{r}
PCD <- list(
  TaATG1 = c(
    "TraesCS4A02G093800",
    "TraesCS4D02G211300"
  ),
  TaATG2 = c(
    "TraesCS6A02G169400",
    "TraesCS6B02G197400",
    "TraesCS7A02G208300",
    "TraesCS7D02G210700"
  ),
  TaATG3 = c(
    "TraesCS1A02G220200",
    "TraesCS1B02G233500",
    "TraesCS1D02G221700"
  ),
  TaATG4 = c(
    "TraesCSU02G021900",
    "TraesCS2D02G601400", 
    "TraesCS2A02G584900",
    "TraesCS6B02G190600"
  ),
  TaATG5 = c(
    "TraesCS6A02G050200",
    "TraesCS6A02G050300",
    "TraesCS6B02G066400",
    "TraesCS6B02G066500",
    "TraesCSU02G109200"
  ),
  TaATG6 = c(
    "TraesCS3A02G266600",
    "TraesCS3D02G266800",
    "TraesCS3B02G300100"
  ),
  TaATG7 = c(
    "TraesCS2D02G300300",
    "TraesCS3A02G220600",
    "TraesCS3B02G250900",
    "TraesCS3D02G231900"
  ),
  TaATG8a = c(
    "TraesCS2B02G491200"
  ),
  TaATG8c = c(
    "TraesCS2A02G468300"
  ),
  TaATG8g = c(
    "TraesCS2A02G224000",
    "TraesCS2B02G274200",
    "TraesCS2D02G229900"
  ),
  TaATG8h = c(
    "TraesCS5B02G155700",
    "TraesCS5D02G163400",
    "TraesCS5A02G158000"
  ),
  TaATG9 = c(
    "TraesCS4A02G079800",
    "TraesCS4B02G224200",
    "TraesCS4D02G224800"
  ),
  TaATG10 = c(
  "TraesCS2A02G352700",
  "TraesCS2B02G371600"
  ),
  TaATG11 = c(
    "TraesCS6A02G149000",
    "TraesCS6B02G177000",
    "TraesCS6D02G138400"
  ),
  TaATG12 = c(
    "TraesCS6A02G335800",
    "TraesCS6B02G366600",
    "TraesCS6D02G315600",
    "TraesCS7B02G070300"
  ),
  TaATG13 = c(
    "TraesCS2A02G387200",
    "TraesCS2B02G404900",
    "TraesCS2D02G384600",
    "TraesCS4A02G209500",
    "TraesCS4B02G112800",
    "TraesCS4D02G110500",
    "TraesCS6A02G238900",
    "TraesCS6B02G285600",
    "TraesCS6D02G221200"
  ),
  TaATG14 = c(
    "TraesCS1A02G308000",
    "TraesCS1B02G318700",
    "TraesCS1D02G307400",
    "TraesCS2A02G160200",
    "TraesCS2B02G186500",
    "TraesCS2D02G167600",
    "TraesCS5A02G141000",
    "TraesCS5B02G139300",
    "TraesCS5D02G151400"
  ),
  TaATG16 = "TraesCS5D02G394200",
  TaATG18 = c(
    "TraesCS1D02G253700",
    "TraesCS3B02G490900",
    "TraesCS3B02G110900",
    "TraesCS6B02G386200"
  ),
  TaATG22 = c(
    "TraesCS3A02G169200",
    "TraesCS3B02G197100",
    "TraesCS3B02G197300",
    "TraesCS3D02G172300",
    "TraesCS3D02G172600"
  ),
  TaATG27 = c(
    "TraesCS4A02G113700",
    "TraesCS4B02G190400",
    "TraesCS4D02G191800"
  ),
  TaATG101 = c(
    "TraesCS2A02G518700",
    "TraesCS2D02G520100"
  ),
  TaMCA1 = c(
    "TraesCS1D02G218100",
    "TraesCS1A02G215600",
    "TraesCS1B02G229100"
  ),
  TaMCA4 = c(
    "TraesCS3D02G322600",
    "TraesCS3A02G328800",
    "TraesCS3B02G359200"
  ),
  `TaMCA-Id` = "TraesCS4D02G358100",
  TaLls1 = c(
    "TraesCS4D02G309000",
    "TraesCS4B02G311100",
    "TraesCS4A02G411000"
  ),
  `HO-1` = "TraesCS7D02G371900",
  TaDAD2 = "TraesCS7D02G029800",
  TaLSD1 = c(
    "TraesCS1A02G279500",
    "TraesCS1B02G288600",
    "TraesCS1D02G278800"
  ),
  TaTrxh1 = "TraesCS2A02G243200",
  TOR = c(
    "TraesCS1D02G229300",
    "TraesCS1A02G227600",
    "TraesCS1B02G241900"
  ),
  EDS1 = c(
    "TraesCS5A02G202200",
    "TraesCS5B02G200800",
    "TraesCS5D02G208600"),
  EDS5 = c(
    "TraesCS6A02G066900",
    "TraesCS6B02G090100",
    "TraesCSU02G117300"),
  CAS = c(
    "TraesCS6A02G290600",
    "TraesCS6B02G320900",
    "TraesCS6D02G271900"),
  CHUP1 = c(
    "TraesCS3A02G235800",
    "TraesCS3B02G264100",
    "TraesCS3D02G235300"),
  TaMDAR6 = "TraesCS7A02G304000",
  `TaBI-1` = "TraesCS6B02G113900"
)

PCD <- tibble::enframe(PCD, name = "Name", value = "ID") |>
  tidyr::unnest(cols = ID)
```

```{r fig.height=12, fig.width=14}
plotExpressionHeatmap(
  cpm_se[rownames(cpm_se) %in% PCD$ID,],
  col_split_by = "cellType",
  col_anno_by = c("treatment", "time"),
  show_row_names = TRUE,
  row_labels = with(PCD, {
    Name[match(rownames(cpm_se)[rownames(cpm_se) %in% ID], ID)]
  })
)
```

```{r fig.height=4, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, PCD$ID,
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types,
  row_labels = with(PCD,
    Name[match(intersect(rownames(logfc_data$mtx), ID), ID)]
  )
)
```

TaLls1:

```{r fig.height=10, fig.width=7}
Seurat::VlnPlot(obj,
  features = c(
    "TraesCS4D02G309000",
    "TraesCS4B02G311100",
    "TraesCS4A02G411000"
  ),
  ncol = 1
) + Seurat::NoLegend()
```

TaMCA4:

```{r fig.height=10, fig.width=7}
Seurat::VlnPlot(obj,
  features = c(
    "TraesCS3D02G322600",
    "TraesCS3A02G328800",
    "TraesCS3B02G359200"
  ),
  ncol = 1
) + Seurat::NoLegend()
```

TaMCA1:

```{r fig.height=10, fig.width=7}
Seurat::VlnPlot(obj,
  features = c(
    "TraesCS1D02G218100",
    "TraesCS1A02G215600",
    "TraesCS1B02G229100"
  ),
  ncol = 1
) + Seurat::NoLegend()
```


## Analysis of Other RNA-seq results

### fusaoctaxin A {.tabset}

#### FA3FB3up

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "FA3FB3up")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "FA3FB3up")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["FA3FB3up"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=14, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["FA3FB3up"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### FA3hUp

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "FA3hUp")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "FA3hUp")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["FA3hUp"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=26, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["FA3hUp"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```

#### FAFB-3hUp

```{r fig.height=4, fig.width=14}
plotRadarGSVA(gsva_df, "FAFB-3hUp")
```

```{r fig.height=4, fig.width=14}
plotLineFacetGSVA(gsva_df, "FAFB-3hUp")
```

```{r}
printDEGNum(logfc_data$mtx, logfc_data$coldata, pathways[["FAFB-3hUp"]]) |>
  reactable::reactable(searchable = TRUE)
```

```{r fig.height=6, fig.width=14}
plotDEGsHeatmap(
  logfc_data$mtx, pathways[["FAFB-3hUp"]],
  logfc_data$heatanno, column_split = logfc_data$coldata$cell_types)
```
