---
title: "Differential Expression along Pseudotime"
author: "Altair Wei"
date: "2023-01-13"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source("../scripts/LoadUtils.R", chdir = TRUE)
```

```{r message=FALSE, warning=FALSE, results="hide"}
library(org.Taestivum.iwgsc.db)
```

```{r}
evaluateOptimalK <- function(data, nc = 15, seed = 1234) {
  wss <- vapply(
    X = seq_len(nc),
    FUN = function(i) {
      set.seed(seed)
      sum(kmeans(data, centers = i)$withinss)
    },
    FUN.VALUE = numeric(1))
  ggplot2::ggplot(data = data.frame(cluster = seq_len(nc), wss = wss),
                  mapping = ggplot2::aes(x = cluster, y = wss)) +
    ggplot2::geom_line(color = "grey") +
    ggplot2::geom_point(color = "steelblue", size = 2) +
    ggplot2::theme_bw() +
    ggplot2::labs(x = "Number of Clusters", y = "Within groups sum of squares")
}
```


## Parenchyma to Outer sheath

### Load Trajectory Data

```{r}
sce_os_all <- readRDS(Sys.glob("../results/ObjectCache/TrajectoryInference/sce_all_slingshot_os_all_*.rds"))
```

### Trajectories

```{r}
plotSlingshotCurveOnReduc(sce_os_all, dimred = "PHATE",
                    linewidth = 1, point_size = 0.5,
                    lineage_rename = c("Lineage1" = "Va_2",
                                       "Lineage2" = "Va_4",
                                       "Lineage3" = "Va_3")) +
  ggplot2::coord_fixed() +
  theme_dimred()
```

### Highly Variable Genes

```{r eval=FALSE}
# Fit GAM for all genes will take too much memory. So
# we select features based on two different methods. See
# https://github.com/statOmics/tradeSeq/issues/150
sce_os_all <- scry::devianceFeatureSelection(sce_os_all, assay = "logcounts", sorted = FALSE)
rowData(sce_os_all)$geneVar <- scran::modelGeneVar(sce_os_all, density.weights = FALSE)
hvg.vars <- scran::getTopHVGs(rowData(sce_os_all)$geneVar, n = 5000)
sorted <- sort(rowData(sce_os_all)$binomial_deviance, decreasing = TRUE)
hvg.dev <- names(sorted)[1:5000]
hvgs <- union(hvg.dev, hvg.vars)
```

> At any given abundance, we assume that the variation in expression for most genes is driven by uninteresting processes like sampling noise.
> Under this assumption, the fitted value of the trend at any given gene's abundance represents an estimate of its uninteresting variation, which we call the technical component.
> We then define the biological component for each gene as the difference between its total variance and the technical component.
> This biological component represents the "interesting" variation for each gene and can be used as the metric for HVG selection.
>
> Strictly speaking, the interpretation of the fitted trend as the technical component assumes that the expression profiles of most genes are dominated by random technical noise.
> In practice, all expressed genes will exhibit some non-zero level of biological variability due to events like transcriptional bursting.
> Thus, it would be more appropriate to consider these estimates as technical noise plus "uninteresting" biological variation, under the assumption that most genes do not participate in the processes driving interesting heterogeneity across the population.

```{r fig.height=4, fig.width=10, eval=FALSE}
layout(matrix(c(1, 2), nrow = 1))

fit.var <- metadata(rowData(sce_os_all)$geneVar)
plot(fit.var$mean, fit.var$var, xlab = "Mean of log-expression",
    ylab = "Variance of log-expression",
    main="Feature Selection with Variance")
points(fit.var$mean[hvg.dev], fit.var$var[hvg.dev], col = "red", pch = 16, cex = 0.5)
curve(fit.var$trend(x), col = "dodgerblue", add = TRUE, lwd = 2)

selected <- sorted * NA
selected[hvg.vars] <- sorted[hvg.vars]
plot(sorted, type="l", xlab="ranked genes",
     ylab="binomial deviance", main="Feature Selection with Deviance")
segments(
  x0 = seq_along(selected), y0 = 0,
  x1 = seq_along(selected), y1 = selected,
  lty = 1, lwd = 0.03)
abline(v = 5000, lty = 2, col = "red")
```

### Fit Negative Binomial Model

```{r}
icMat <- xfun::cache_rds(
  file = "icMat_os_all.rds",
  dir = "../results/ObjectCache/TrajectoryInference/",
  expr = tradeSeq::evaluateK(sce_os_all, k = 3:10)
)
```

```{r}
invisible(tradeSeq::plot_evalutateK_results(icMat, k = 3:10, aicDiff = 2))
```

The middle panels show that the drop in AIC levels off if the number of knots is increased beyond 8, and we will choose that number of knots to fit the tradeSeq models.

```{r}
tradeSce_os_all <- xfun::cache_rds(
  file = "tradeseq_gam_os_all.rds",
  dir = "../results/ObjectCache/TrajectoryInference/",
  expr = tradeSeq::fitGAM(
    # fitGAM needs counts rather than logcounts, see
    # the source code of SCE-version fitGAM.
    counts = sce_os_all,
    nknots = 8,
    sce = TRUE,
    verbose = TRUE,
    parallel = TRUE,
    BPPARAM = BiocParallel::MulticoreParam(workers = 1)
  )
)
```


### Differential Expression

#### Run Test

> In large datasets with thousands of single cells, small fold changes can become statistically significant due to the very high sample size. However, these differences may not always be biologically meaningful. To tackle this problem, all DE tests in tradeSeq have an `l2fc` argument which specifies the absolute value of the log2 fold-change cut-off to test against. For example, setting `l2fc=log2(2)` will test which genes have a fold change that is significantly higher than 2 or significantly lower than 1/2 within or between the lineages of interest.

似乎 `tradeSeq::patternTest` 中的 `l2fc` 只影响 global 的结果。我看了下代码，pairwise=TRUE 且 sce=TRUE 的条件下，调用了 `getEigenStatGAM` 而非 `getEigenStatGAMFC`，因此没有用上 l2fc 参数。

```{r}
patternRes_os_all <- xfun::cache_rds(
  file = "tradeseq_patternRes_os_all.rds",
  dir = "../results/ObjectCache/TrajectoryInference/",
  expr = tradeSeq::patternTest(
    tradeSce_os_all, global = TRUE, pairwise = FALSE, l2fc = log2(2))
)
```

```{r}
patternRes_os_all |> str()
```


```{r}
patternRes_df <- patternRes_os_all |>
  tibble::rownames_to_column("gene") |>
  dplyr::mutate(fdr = p.adjust(pvalue, "fdr"))

str(patternRes_df)
```

#### Check Results

```{r}
patternRes_df |>
  dplyr::arrange(fdr) |>
  ggplot2::ggplot(ggplot2::aes(
    x = fcMedian, y = -log(pvalue), color = fdr < 0.01)) +
  ggplot2::geom_point() +
  ggplot2::scale_x_log10() +
  NULL
```

```{r}
patternRes_df |>
  dplyr::arrange(fdr) |>
  ggplot2::ggplot(ggplot2::aes(
    x = fcMedian, fill = fdr < 0.01)) +
  ggplot2::geom_density(alpha = 0.5) +
  ggplot2::scale_x_log10() +
  NULL
```

#### Gene Clustering

```{r}
patternGenes <- patternRes_df |>
  dplyr::filter(fdr <= 0.01) |>
  dplyr::pull(gene) |>
  unique()

length(patternGenes)
```

```{r}
yhatSmooth <- tradeSeq::predictSmooth(tradeSce_os_all, gene = patternGenes, nPoints = 100, tidy = FALSE)
yhatSmoothScaled <- t(scale(t(yhatSmooth)))
```

Determine optimal number of clusters:

```{r}
evaluateOptimalK(yhatSmoothScaled)
```

```{r eval=FALSE}
nbct <- lapply(
  X = c("kl", "ch", "hartigan", "ccc", "scott", "marriot", "trcovw",
        "tracew", "friedman", "rubin", "cindex", "db", "silhouette",
        "duda", "pseudot2", "beale", "ratkowsky", "ball", "ptbiserial",
        "frey", "mcclain", "dunn", "hubert", "sdindex", "dindex", "sdbw"),
  FUN = function(index) {
    cat("Start calculate:", index, "\n")
    tryCatch(
      NbClust::NbClust(
        yhatSmoothScaled, min.nc = 2, max.nc = 15,
        method = "kmeans", index = index)$Best.nc,
      error = function(e) {
        cat("Failed to calculate", index, "\n")
        NULL
      }
    )
  })

nbct
```

```{r}
set.seed(1124)
fit_km <- kmeans(yhatSmoothScaled, 6, nstart = 25)
```

```{r}
avgtime <- split(
    x = as.data.frame(t(fit_km$centers)),
    f = stringr::str_match(
      colnames(fit_km$centers), "(lineage\\d)_")[, 2]) |>
  vapply(
    FUN.VALUE = numeric(nrow(fit_km$centers)),
    FUN = function(df) apply(df, 2, which.max)) |>
  base::rowMeans()

avgtime
```

```{r}
row_split <- factor(
  x = paste0("G", fit_km$cluster),
  levels = paste0("G", names(sort(avgtime))))
```

#### Heatmap

```{r fig.height=7, fig.width=14}
ph_res <- plotPseudotimeHeatmap(
  yhatSmoothScaled,
  palette = colorRamps::matlab.like,
  row_split = row_split,
  row_title_rot = 0,
  show_row_dend = FALSE,
  column_split = stringr::str_match(
    colnames(yhatSmoothScaled), "(lineage\\d)_")[, 2],
  cluster_row_slices = FALSE
) |> ComplexHeatmap::draw()
```

#### Enrichment Analysis

```{r}
gene_list <- split(names(fit_km$cluster), paste0("G", fit_km$cluster))
gene_list <- gene_list[levels(row_split)]
```

```{r}
comp_enr <- clusterProfiler::compareCluster(
  gene_list,
  fun = "enrichGO",
  OrgDb = "org.Taestivum.iwgsc.db",
  keyType = "GID",
  ont = "BP",
  universe = rownames(tradeSce_os_all)
)
```

```{r fig.height=9, fig.width=8}
comp_enr |>
  enrichRepresent() |>
  enrichplot::dotplot(
    showCategory = NULL,
    by = "count", label_format = 50)
```

## Chlorenchyma

### Load Trajectory Data

```{r}
sce_chl_all <- readRDS(Sys.glob("../results/ObjectCache/TrajectoryInference/sce_all_slingshot_chl_all_*.rds"))
```

### Trajectories

```{r}
plotSlingshotCurveOnReduc(sce_chl_all, dimred = "PHATE",
                    linewidth = 1, point_size = 0.5) +
  ggplot2::coord_fixed() +
  theme_dimred()
```

### Fit Negative Binomial Model

```{r}
icMat <- xfun::cache_rds(
  file = "icMat_chl_all.rds",
  dir = "../results/ObjectCache/TrajectoryInference/",
  expr = tradeSeq::evaluateK(sce_chl_all, k = 3:10)
)
```

```{r}
invisible(tradeSeq::plot_evalutateK_results(icMat, k = 3:10, aicDiff = 2))
```

Optimal knot = 8

```{r}
tradeSce_chl_all <- xfun::cache_rds(
  file = "tradeseq_gam_chl_all.rds",
  dir = "../results/ObjectCache/TrajectoryInference/",
  clean = FALSE,
  expr = tradeSeq::fitGAM(
    # fitGAM needs counts rather than logcounts, see
    # the source code of SCE-version fitGAM.
    counts = sce_chl_all,
    nknots = 8,
    sce = TRUE,
    verbose = TRUE,
    parallel = TRUE,
    BPPARAM = BiocParallel::MulticoreParam(workers = 1)
  )
)
```

### Differential Expression

#### Run Test

```{r}
patternRes_chl_all <- xfun::cache_rds(
  file = "tradeseq_patternRes_chl_all.rds",
  dir = "../results/ObjectCache/TrajectoryInference/",
  expr = tradeSeq::patternTest(
    tradeSce_chl_all, global = TRUE, pairwise = FALSE, l2fc = log2(3))
)
```


```{r}
patternRes_df <- patternRes_chl_all |>
  tibble::rownames_to_column("gene") |>
  dplyr::mutate(fdr = p.adjust(pvalue, "fdr"))

str(patternRes_df)
```

```{r}
patternRes_df |>
  dplyr::filter(gene == "TraesCS5B02G039000")
```


#### Check Results

> Rather than attaching strong probabilistic interpretations to the p-values (which, as in most RNA-seq applications, would involve a variety of hard-to-verify assumptions and would not necessarily add much value to the analysis), we view the p-values simply as useful numerical summaries for ranking the genes for further inspection.

```{r}
patternRes_df |>
  dplyr::arrange(fdr) |>
  ggplot2::ggplot(ggplot2::aes(
    x = fcMedian, y = -log(pvalue), color = fdr < 0.01)) +
  ggplot2::geom_point() +
  ggplot2::scale_x_log10() +
  NULL
```

```{r}
patternRes_df |>
  dplyr::arrange(fdr) |>
  ggplot2::ggplot(ggplot2::aes(
    x = fcMedian, fill = fdr < 0.01)) +
  ggplot2::geom_density(alpha = 0.5) +
  ggplot2::scale_x_log10() +
  NULL
```

#### Gene Clustering

```{r}
patternGenes <- patternRes_df |>
  dplyr::filter(fdr <= 0.01) |>
  dplyr::pull(gene) |>
  unique()

length(patternGenes)
```

```{r}
yhatSmooth <- tradeSeq::predictSmooth(tradeSce_chl_all, gene = patternGenes, nPoints = 100, tidy = FALSE)
yhatSmoothScaled <- t(scale(t(yhatSmooth)))
```

Determine optimal number of clusters:

```{r}
evaluateOptimalK(yhatSmoothScaled)
```

Optimal k=4

```{r}
set.seed(1116)
fit_km <- kmeans(yhatSmoothScaled, 4, nstart = 25)
```

```{r}
avgtime <- split(
    x = as.data.frame(t(fit_km$centers)),
    f = stringr::str_match(
      colnames(fit_km$centers), "(lineage\\d)_")[, 2]) |>
  vapply(
    FUN.VALUE = numeric(nrow(fit_km$centers)),
    FUN = function(df) apply(df, 2, which.max)) |>
  base::rowMeans()

avgtime
```

```{r}
row_split <- factor(
  x = paste0("G", fit_km$cluster),
  levels = paste0("G", names(sort(avgtime))))
```

#### Heatmap

```{r fig.height=7, fig.width=14}
ph_res <- plotPseudotimeHeatmap(
  yhatSmoothScaled,
  palette = colorRamps::matlab.like,
  row_split = row_split,
  row_title_rot = 0,
  show_row_dend = FALSE,
  column_split = stringr::str_match(
    colnames(yhatSmoothScaled), "(lineage\\d)_")[, 2],
  cluster_row_slices = FALSE
) |> ComplexHeatmap::draw()
```

#### Enrichment Analysis

```{r}
gene_list <- split(names(fit_km$cluster), paste0("G", fit_km$cluster))
gene_list <- gene_list[levels(row_split)]
```

```{r}
comp_enr <- clusterProfiler::compareCluster(
  gene_list,
  fun = "enrichGO",
  OrgDb = "org.Taestivum.iwgsc.db",
  keyType = "GID",
  ont = "BP",
  universe = rownames(tradeSce_chl_all)
)
```

```{r fig.height=9, fig.width=8}
comp_enr |>
  clusterProfiler::simplify(cutoff = 0.8) |>
  enrichRepresent() |>
  enrichplot::dotplot(
    showCategory = NULL,
    by = "count", label_format = 50)
```
