---
title: "Differential State Analysis"
output: html_document
---

```{r setup, include=FALSE}
library(magrittr)
library(patchwork)
library(SingleCellExperiment)
source("../scripts/LoadUtils.R")
BPPARAM <- BiocParallel::MulticoreParam(workers = 6)
```

## Prepare Data

### Cell Annotation

```{r}
all_celltypes <- c(
  "Gu", "Ep_1", "Ep_2",
  "Me_1", "Me_2", "Me_3", "Me_4", "Me_5", "Me_6",
  "Va_1", "Va_2", "Va_3", "Va_4",
  "BS", "CC", "MPV_1", "MPV_2")

all_celltype_colors <- ggthemes::tableau_color_pal("Tableau 20")(length(all_celltypes))
names(all_celltype_colors) <- all_celltypes
```

```{r}
all_body_layers <- c(
  "Gu"    = "L1",
  "Ep_1"  = "L1",
  "Ep_2"  = "L1",
  "Me_1"  = "L2",
  "Me_2"  = "L2",
  "Me_3"  = "L2",
  "Me_4"  = "L2",
  "Me_5"  = "L2",
  "Me_6"  = "L2",
  "Va_1"  = "L2",
  "Va_2"  = "L2",
  "Va_3"  = "L2",
  "Va_4"  = "L2",
  "BS"    = "L3",
  "CC"    = "L3",
  "MPV_1" = "L3",
  "MPV_2" = "L3"
)

all_tissues <- c(
  Gu   = "Stomata",
  Ep_1 = "Epidermis",
  Ep_2 = "Epidermis",
  Me_1 = "Chlorenchyma",
  Me_2 = "Chlorenchyma",
  Me_3 = "Chlorenchyma",
  Me_4 = "Parenchyma",
  Me_5 = "Parenchyma",
  Me_6 = "Parenchyma",
  Va_1 = "Chlorenchyma",
  Va_2 = "Outer sheath",
  Va_3 = "Outer sheath",
  Va_4 = "Outer sheath",
  BS = "Inner sheath",
  CC = "Phloem",
  MPV_1 = "Procambium",
  MPV_2 = "Procambium"
)
```

### Convert Data Format

`muscat` 包要求输入数据需要预先经过处理，比如整合、聚类和细胞注释等等，因此这里我们直接加载已经处理过的 `Seurat` 对象。

```{r}
obj <- readRDS(Sys.glob("../results/ObjectCache/IntegrationAndAnnotation/obj_annotated_*.rds"))
```

```{r read-muscat-sce}
sce <- Seurat::as.SingleCellExperiment(obj, assay = "RNA")
sce$sample <- make.names(sce$sample)
sce$group <- make.names(sce$group)
```

```{r}
sce$ident <- forcats::fct_relevel(sce$ident, all_celltypes)
sce$body_layer <- do.call(
  dplyr::recode, c(list(.x = sce$ident), as.list(all_body_layers)))
sce$tissue <- do.call(
  dplyr::recode, c(list(.x = sce$ident), as.list(all_tissues)))
```

```{r}
sce <- muscat::prepSCE(sce,
  kid = "tissue", # cell type
  gid = "group", # group IDs (ctrl/stim)
  sid = "sample", # sample IDs (ctrl/stim.1234)
  drop = FALSE
)
```

```{r}
rm(obj)
invisible(gc())
```

### Cell Abundance

差异状态分析依赖于细胞亚群的大小，因此群体规模太小的亚群应该事先排除掉。

```{r cluster-prop-table}
cluster_stats <- table(sce$cluster_id, sce$sample_id)
cluster_stats %>% as.data.frame.matrix()
```

看看各个类群在各自样本中所占的比例：

```{r cluster-abundance-barplot, fig.height=7, fig.width=14, warning=FALSE}
p1 <- scater::plotReducedDim(
    sce,
    dimred = "UMAP",
    point_size = 1,
    colour_by = "cluster_id",
    text_by = "cluster_id",
    text_size = 4,
    theme_size = 14) +
  ggthemes::scale_color_tableau() +
  ggplot2::coord_fixed() +
  remove_legend() +
  theme_dimred()

p2 <- rhapsodykit::barplot_cluster_abundance(sce) +
  ggthemes::scale_fill_tableau() +
  NULL

p2$data$sample_id <- stringr::str_replace(
  p2$data$sample_id, "X(\\dDPI)\\.(.*)\\.(\\d)", "\\1-\\2-\\3")

(p <- p1 + p2 + patchwork::plot_layout(widths = c(2, 2)))
```

### Functions

```{r}
upset_contrasts_list <- function(input, ...) {
  sets_list <- input$table %>%
    purrr::map(function(x) {
      purrr::map(x, "gene") %>%
        purrr::compact()
    })

  sets_list <- Filter(\(x) length(x) > 0, sets_list)

  sets_metadata <- sets_list %>%
    purrr::imap(function(contrast, key) {
      cbind(
        sets = sprintf(
          "%s [%s]",
          stringr::str_replace(key,
            "X(\\dDPI)\\.(.*)-X.*", "\\1-\\2"),
          names(contrast)),
        contrast = key,
        cluster = names(contrast)
      )
    }) %>%
    do.call(rbind, .) %>%
    as.data.frame()

  sets_to_plot <- sets_list %>%
    unlist(recursive = FALSE)
  
  names(sets_to_plot) <- sets_metadata$sets

  sets_to_plot
}
```

## Perform Differential State

### Pseudobulk Aggregation

#### For biologist

Pseudobulk CPM mean expression matrix:

```{r}
pseudobulk_cpm_mean <- xfun::cache_rds(
  file = "pseudobulk_cpm_mean.rds",
  dir = "../results/ObjectCache/Pseudobulk/",
  expr = local({
    splits <- unique(sce$treatment)
    names(splits) <- splits
    pb_list <- splits |>
      lapply(function(tr) sce[, sce$treatment == tr]) |>
      lapply(function(sub) {
        pb <- rhapsodykit::calculate_pseudo_bulk(
          sub, type = "cpm", fun = "mean", by = c("time", "cluster_id"))
        as.list(pb@assays@data)
      })
  
    pb_list <- unlist(pb_list, recursive = FALSE, use.names = TRUE) |>
      purrr::imap(function(mtx, name) {
        colnames(mtx) <- paste(name, colnames(mtx), sep = ".")
        mtx
      })
  
    pb_mtx <- do.call(cbind, pb_list)

    writexl::write_xlsx(
      x = as.data.frame(pb_mtx, row.names = rownames(pb_mtx)) |>
        tibble::rownames_to_column("Gene"),
      path = "../results/DataPlots/pseudobulk_cpm_mean.xlsx")
    
    pb_mtx
  })
)
```

```{r}
pseudobulk_cpm_mean_df <- xfun::cache_rds(
  file = "pseudobulk_df_cpm_mean.rds",
  dir = "../results/ObjectCache/Pseudobulk/",
  expr = as.data.frame(pseudobulk_cpm_mean) |>
    tibble::rownames_to_column("Gene") |>
    tidyr::pivot_longer(-Gene, names_to = "bulk", values_to = "CPM") |>
    tidyr::separate(bulk, sep = "\\.", into = c("treatment", "time", "cellType"))
)
```

#### For differential state

差异状态分析有两类方法，其中之一是基于聚合数据的差异基因表达方法，这种方法的稳定性和表现都很不错。

所以我们首先将单细胞数据聚合成 pseudo-bulk 数据：

```{r calc-pseudo-bulk, warning=FALSE}
pb_list <- xfun::cache_rds(
  file = "pb_list.rds",
  rerun = FALSE,
  dir = "../results/ObjectCache/DifferentialState/",
  expr = {
    list(
      "counts" = rhapsodykit::calculate_pseudo_bulk(sce, "counts", BPPARAM = BPPARAM),
      "logcounts" = rhapsodykit::calculate_pseudo_bulk(sce, "logcounts", BPPARAM = BPPARAM)
      #"cpm" = rhapsodykit::calculate_pseudo_bulk(sce, "cpm")
      #"vstresiduals" = rhapsodykit::calculate_pseudo_bulk(sce, "vstresiduals")
    )
  }
)
```

变量 `pb` 实际上也是一个 `SingleCellExperiment` 对象，每一个 assay 都是一个 cluster 的 pseudo-bulk 数据。

```{r}
pb_list$counts |> assays()
```

```{r}
pb_list$counts |> assay("Stomata") |> str()
```

### Variance of Samples

我们需要通过 MDS plot 查看一下 pseudo-bulk 是否能够反应样本间的相似性。注意，在这里我们使用基于 logcounts 的数据而不是 counts 数据。

```{r pb-mds-plot}
pb_mds_list <- xfun::cache_rds(
  file = "pb_mds_list.rds",
  dir = "../results/ObjectCache/DifferentialState/",
  rerun = FALSE,
  expr = local({
    data1 <- muscat::pbMDS(pb_list$counts)$data
    data1$time <- stringr::str_extract(data1$group_id, "\\dDPI")
    data1$treatment <- stringr::str_extract(data1$group_id, "MOCK|PNR2|TR4")

    data2 <- muscat::pbMDS(pb_list$logcounts)$data
    data2$time <- stringr::str_extract(data2$group_id, "\\dDPI")
    data2$treatment <- stringr::str_extract(data2$group_id, "MOCK|PNR2|TR4")

    list(counts = data1, logcounts = data2)
  })
)
```

然后我们展示该分布：

```{r mds-plot-by-time, fig.height=7, fig.width=14}
plist <- lapply(pb_mds_list, function(pb_mds) {
  pb_mds %>%
    ggplot2::ggplot(ggplot2::aes_string(
      x = "MDS1", y = "MDS2",
      color = "cluster_id", shape = "treatment")
    ) +
    ggplot2::geom_point(size = 2, alpha = 1) +
    ggplot2::facet_wrap(~time) +
    ggthemes::scale_color_tableau() +
    ggplot2::scale_shape_manual(values = c("MOCK" = 15, "PNR2" = 8, "TR4" = 10)) +
    ggplot2::guides(
      color = ggplot2::guide_legend(override.aes = list(alpha = 1))) +
    ggplot2::theme_bw() +
    ggplot2::theme(
      aspect.ratio = 1,
      axis.text = ggplot2::element_text(color = "black"),
      panel.grid.minor = ggplot2::element_blank(),
      panel.grid.major = ggplot2::element_line(linewidth = 0.2, color = "lightgrey")
    )
})

plist$counts <- plist$counts + ggplot2::ggtitle("Pseudobulk Counts")
plist$logcounts <- plist$logcounts + ggplot2::ggtitle("Pseudobulk Logcounts")

(p <- patchwork::wrap_plots(plist, guides = "collect") &
    center_plot_title())
```

我们以细胞群集为分面来看看：

```{r mds-plot-by-cluster, fig.height=7, fig.width=14}
plot_mds_by_cluster <- function(df, clusters = NULL) {
  if (!is.null(clusters))
    df <- dplyr::filter(df, cluster_id %in% clusters)

  df %>%
    ggplot2::ggplot(ggplot2::aes_string(
      x = "MDS1", y = "MDS2",
      color = "treatment", size = "time", alpha = "time")) +
    ggplot2::geom_point() +
    ggplot2::facet_wrap(~cluster_id, nrow = 2) +
    ggplot2::scale_color_brewer(type = "qual", palette = "Dark2") +
    ggplot2::scale_alpha_manual(values = c(
      "0DPI" = 1, "1DPI" = .8, "2DPI" = .6, "3DPI" = .4)) +
    ggplot2::scale_size_manual(values = c(
      "0DPI" = 1, "1DPI" = 2, "2DPI" = 3, "3DPI" = 4)) +
    ggplot2::guides(
      color = ggplot2::guide_legend(override.aes = list(alpha = 1))) +
    ggplot2::theme_bw() +
    ggplot2::theme(
      axis.text = ggplot2::element_text(color = "black"),
      panel.grid.minor = ggplot2::element_blank(),
      panel.grid.major = ggplot2::element_line(
        linewidth = 0.2, color = "lightgrey")
    )
}

(p <- plot_mds_by_cluster(pb_mds_list$logcounts))
```

### Pseudobulk differential expression

可以使用 `muscat::pbDS` 函数完成任务，不过当要比较的 group 超过两个时，我们需要首先构建 design 矩阵以及 contrast 矩阵：

我们看看实验相关的信息，主要是 sample 属于哪个 group ：

```{r}
metadata(sce)$experiment_info
```

在需要提供的数据中，design 矩阵起着指示器的作用，它用于表明样本属于哪个组。而 `model.matrix` 的统计学含义以及用法可以参考 [Expressing design formula in R](https://genomicsclass.github.io/book/pages/expressing_design_formula.html) 。contrast 矩阵用于表明我们想要比较哪两个组，以 `stim-ctrl` 字符串的形式指定，实验组在 `-` 前，对照组在后。我在 `rhapsodykit` 中提供了一个方便的帮助函数，自动构建 design 和 contrast 矩阵。

```{r run-ds-analysis}
# run DS analysis
contrast_names <- c(
  "X1DPI.PNR2-X1DPI.MOCK",
  "X1DPI.TR4-X1DPI.MOCK",
  "X1DPI.PNR2-X1DPI.TR4",

  "X2DPI.PNR2-X2DPI.MOCK",
  "X2DPI.TR4-X2DPI.MOCK",
  "X2DPI.PNR2-X2DPI.TR4",

  "X3DPI.PNR2-X3DPI.MOCK",
  "X3DPI.TR4-X3DPI.MOCK",
  "X3DPI.PNR2-X3DPI.TR4"
)

ds_res <- xfun::cache_rds(
  file = "ds_res.rds",
  dir = "../results/ObjectCache/DifferentialState/",
  rerun = FALSE,
  expr = {
    list(
      edgeR = rhapsodykit::pseudobulk_diff_state(
        sce, pb_list$counts,
        method = "edgeR",
        contrasts = contrast_names,
        filter = "genes",
        BPPARAM = BPPARAM
      ),
      DESeq2 = rhapsodykit::pseudobulk_diff_state(
        sce, pb_list$counts,
        method = "DESeq2",
        contrasts = contrast_names,
        filter = "genes",
        BPPARAM = BPPARAM
      )
    )
  }
)
```

```{r}
str(ds_res$DESeq2$table$`X1DPI.PNR2-X1DPI.MOCK`, max.level = 1)
```

差异基因表达分析的结果储存在 `res_list` 列表中，而其成员则是 `contrast` 参数所指定的比较组合。每个比较组合又是一个列表，并且包含了针对每个 cluster 的差异表达分析结果。

```{r}
# access results table for 1st comparison
cluster1 <- ds_res$DESeq2$table[["X1DPI.PNR2-X1DPI.MOCK"]][[1]]
head(cluster1)
```

1DPI-PNR2 vs. 1DPI-MOCK 比较组合中 cluster 1 的差异基因表达结果如上所示。

## Handle DS Results

### Filtering and summary

Here we only filter results with P-adjusted value smaller than 0.05:

```{r}
ds_sig <- ds_res$DESeq2 %>%
  rhapsodykit::diff_state_significant(fdr_limit = 0.05, logfc_limit = 0)
```

```{r show-diff-results}
rhapsodykit::diff_state_summary(ds_sig) |>
  dplyr::rename(
    `Cell Type` = cluster_id,
    `No. DS` = num_of_DS,
    `Pct. DS` = percent_of_DS,
    Contrast = contrast)
```

### Filter by epxression frequency

除了查看基因表达的变化幅度和显著性外，我们还可以看看基因表达的频率，即一个样本或组别中表达给定基因的细胞所占的比例。

```{r gene-expr-freq, eval=FALSE}
ds_res$DESeq2 %>%
  rhapsodykit::diff_state_significant(fdr_limit = 0.05, logfc_limit = 0) %>%
  rhapsodykit::expr_freq_filter(sce) %>%
  rhapsodykit::diff_state_summary()
```

We don't filter DEGs by expression frequency.

## Visualization of DS Results

### Visualize No. of DEGs

```{r}
genecount_df <- rhapsodykit::diff_state_format(ds_sig)
```

```{r eval=FALSE}
writexl::write_xlsx(genecount_df, "../results/DataPlots/DS_DESeq2_Significant_DEGs.xlsx")
```

```{r diff-gene-count-plot, fig.height=7, fig.width=7}
genecount_df$cluster_id <- factor(genecount_df$cluster_id, unique(all_tissues))
genecount_df$time <- stringr::str_extract(genecount_df$contrast, "\\dDPI")
genecount_df$contrast <- factor(genecount_df$contrast, levels = contrast_names)
genecount_df$regu_type <- ifelse(genecount_df$logFC > 0, "Up-regulated", "Down-regulated")

df_to_plot <- genecount_df |>
  dplyr::group_by(time, contrast, cluster_id, regu_type) |>
  dplyr::count() |>
  dplyr::mutate(n = ifelse(regu_type == "Up-regulated", n, -1 * n)) |>
  tidyr::pivot_wider(names_from = "cluster_id", values_from = "n", values_fill = NA) |>
  tidyr::pivot_longer(cols = -c(time, contrast, regu_type), names_to = "cluster_id", values_to = "n") |>
  dplyr::mutate(contrast = stringr::str_replace(contrast,
      "X(\\dDPI)\\.(.*)-X(\\dDPI)\\.(.*)", "\\2 vs. \\4"))

df_to_plot$cluster_id <- factor(df_to_plot$cluster_id, unique(all_tissues))
df_to_plot$contrast <- factor(df_to_plot$contrast, c("PNR2 vs. MOCK", "TR4 vs. MOCK", "PNR2 vs. TR4"))

p <- ggplot2::ggplot(
    data = df_to_plot,
    mapping = ggplot2::aes(x = cluster_id, y = n, fill = contrast)) +
  ggplot2::geom_bar(
    data = subset(df_to_plot, regu_type == "Up-regulated"),
    stat = "identity",
    position = ggplot2::position_dodge2(preserve = "single", padding = 0)) +
  ggplot2::geom_bar(
    data = subset(df_to_plot, regu_type == "Down-regulated"),
    stat = "identity", alpha = 0.5,
    position = ggplot2::position_dodge2(preserve = "single", padding = 0)) +
  ggplot2::scale_fill_brewer(palette = "Set1") +
  ggplot2::ylab("Number of DS genes") +
  ggplot2::xlab("Cell Types") +
  ggplot2::theme_bw() +
  ggplot2::theme(
    axis.text.x = ggplot2::element_text(angle = 45, hjust = 1),
    legend.position = "right",
    strip.background.x = ggplot2::element_blank(),
    strip.text.x = ggplot2::element_blank()) +
  ggplot2::facet_grid(time ~ cluster_id, scales = "free_x") +
  NULL

p
```

edgeR 与 DESeq2 发现的基因数量大体趋势是一致的，只不过两种方法计算出的 p_adj.loc 可能有所差别。

### Volcano plot to show DEGs {.tabset}

```{r results='asis'}
df_to_plot <- ds_res$DESeq2 %>% rhapsodykit::diff_state_format()

for (dp in c("1DPI", "2DPI", "3DPI")) {
  cat("####", dp, "{.tabset}\n\n")
  for (tr in list(c("PNR2", "MOCK"),
                  c("TR4", "MOCK"),
                  c("PNR2", "TR4"))) {

    cont = sprintf("X%s.%s-X%s.%s", dp, tr[[1]], dp, tr[[2]])
    title = sprintf("%s-%s vs. %s-%s", dp, tr[[1]], dp, tr[[2]])
    cat("#####", title, "\n\n")

    dplyr::filter(df_to_plot, contrast == cont) |>
      rhapsodykit::volcano_diff_state(title = title) |>
      print()

    cat("\n\n")
  }
  cat("\n\n")
}
```

### Intersections between cell types {.tabset}

```{r}
plot_updown <- function(ds, contrast) {
  p1 <- ds %>%
    rhapsodykit::diff_state_pull("X1DPI.PNR2-X1DPI.MOCK", 1:8, "gene") %>%
    rhapsodykit::from_list_to_upset() %>%
    rhapsodykit::upset_dataframe() %>%
    rhapsodykit::upset_plot(n_intersections = 30)

  up_list <- ds %>%
    rhapsodykit::diff_state_filter(logFC > 0) %>%
    rhapsodykit::diff_state_pull(contrast, seq_along(ds$data), "gene")

  down_list <- ds %>%
    rhapsodykit::diff_state_filter(logFC < 0) %>%
    rhapsodykit::diff_state_pull(contrast, seq_along(ds$data), "gene")

  p2 <- rhapsodykit::upset_updown_regulated(
    up_list, down_list, n_intersections = 30)

  p1 / p2
}
```

```{r results='asis', fig.height=10, fig.width=10}
for (dp in c("1DPI", "2DPI", "3DPI")) {
  cat("####", dp, "{.tabset}\n\n")
  for (tr in list(c("PNR2", "MOCK"),
                  c("TR4", "MOCK"),
                  c("PNR2", "TR4"))) {

    cont = sprintf("X%s.%s-X%s.%s", dp, tr[[1]], dp, tr[[2]])
    title = sprintf("%s-%s vs. %s-%s", dp, tr[[1]], dp, tr[[2]])
    cat("#####", title, "\n\n")

    p <- plot_updown(ds_sig, cont) &
      ggplot2::ggtitle(title)
    print(p)

    cat("\n\n")
  }
  cat("\n\n")
}
```

### The concordance between treatments

```{r}
plot_updown_contrast <- function(ds, contrast, clusters = NULL) {
  p1 <- ds %>%
    rhapsodykit::diff_state_subset(
      contrast = contrast,
      clusters = clusters) %>%
    upset_contrasts_list() %>%
    rhapsodykit::from_list_to_upset() %>%
    rhapsodykit::upset_dataframe() %>%
    rhapsodykit::upset_plot(
      order_by = "degree",
      n_intersections = 100)

  up_list <- ds %>%
    rhapsodykit::diff_state_subset(
      contrast = contrast,
      clusters = clusters) %>%
    rhapsodykit::diff_state_filter(logFC > 0) %>%
    upset_contrasts_list()

  down_list <- ds %>%
    rhapsodykit::diff_state_subset(
      contrast = contrast,
      clusters = clusters) %>%
    rhapsodykit::diff_state_filter(logFC < 0) %>%
    upset_contrasts_list()

  p2 <- rhapsodykit::upset_updown_regulated(
    up_list, down_list,
    order_by = "degree", n_intersections = 100)

  p1 / p2
}
```

因为有不同的 contrast ，所以无法显示上调与下调的基因？也对，某个基因可能在一个集合（cellType或者contrast）里面上调，但在另一个集合里面下调，这种基因会被漏掉。这个困局对任何在 UpsetPlot 上显示上下调基因的绘图来说都是存在的。

#### Compare all DEG list

```{r fig.height=20, fig.width=20}
plot_updown_contrast(
  ds_sig,
  contrast = c(
    "X1DPI.PNR2-X1DPI.MOCK",
    "X1DPI.TR4-X1DPI.MOCK",
    "X2DPI.PNR2-X2DPI.MOCK",
    "X2DPI.TR4-X2DPI.MOCK",
    "X3DPI.PNR2-X3DPI.MOCK",
    "X3DPI.TR4-X3DPI.MOCK")
)
```

#### Compare by time {.tabset}

```{r fig.height=14, fig.width=18, results='asis'}
for (dp in c("1DPI", "2DPI", "3DPI")) {
  cat("#####", dp, "\n\n")
  contrast <- sprintf(c(
    "X%s.PNR2-X%s.MOCK",
    "X%s.TR4-X%s.MOCK"), dp, dp)
  p <- plot_updown_contrast(
    ds_sig, contrast = contrast)
  print(p)
  cat("\n\n")
}
```

#### Compare by cell type {.tabset}

```{r fig.height=14, fig.width=14, results='asis'}
for (cls in unique(all_tissues)) {
  cat("#####", cls, "\n\n")
  p <- plot_updown_contrast(
    ds_sig,
    contrast = c(
      "X1DPI.PNR2-X1DPI.MOCK",
      "X1DPI.TR4-X1DPI.MOCK",
      "X2DPI.PNR2-X2DPI.MOCK",
      "X2DPI.TR4-X2DPI.MOCK",
      "X3DPI.PNR2-X3DPI.MOCK",
      "X3DPI.TR4-X3DPI.MOCK"),
    clusters = cls)
  print(p)
  cat("\n\n")
}
```

### Overview of DEGs

#### LogFC Scatters

```{r fig.height=7, fig.width=14}
df_to_plot <- ds_res$DESeq2 |>
  rhapsodykit::diff_state_significant(fdr_limit = 0.05, logfc_limit = 0) |>
  rhapsodykit::diff_state_format() |>
  dplyr::filter(!stringr::str_detect(contrast, "X\\dDPI\\.PNR2-X\\dDPI\\.TR4")) |>
  dplyr::select(gene, cluster_id, contrast, logFC) |>
  dplyr::mutate(
    time = stringr::str_extract(contrast, "\\dDPI"),
    contrast = stringr::str_replace(contrast,
      "X(\\dDPI)\\.(.*)-X(\\dDPI)\\.(.*)", "\\2-\\4")) |>
  tidyr::pivot_wider(names_from = "time", values_from = "logFC", values_fill = 0) |>
  tidyr::pivot_longer(cols = dplyr::ends_with("DPI"), names_to = "time", values_to = "logFC") |>
  tidyr::pivot_wider(names_from = contrast, values_from = logFC, values_fill = 0)

p <- df_to_plot |>
  ggplot2::ggplot(ggplot2::aes(
    x = `PNR2-MOCK`, y = `TR4-MOCK`, color = cluster_id)) +
  ggplot2::geom_point(shape = 1, size = 1) +
  ggthemes::scale_color_tableau() +
  ggplot2::facet_grid(time ~ cluster_id) +
  ggplot2::theme_bw() +
  remove_legend()

p
```

#### LogFC Heatmap

```{r}
logfc_df <- ds_res$DESeq2 |>
  rhapsodykit::diff_state_significant(fdr_limit = 0.05, logfc_limit = 0) |>
  rhapsodykit::diff_state_format() |>
  dplyr::filter(!stringr::str_detect(contrast, "X\\dDPI\\.PNR2-X\\dDPI\\.TR4")) |>
  dplyr::select(gene, cluster_id, contrast, logFC) |>
  dplyr::mutate(
    time = stringr::str_extract(contrast, "\\dDPI"),
    contrast = stringr::str_replace(contrast,
      "X(\\dDPI)\\.(.*)-X(\\dDPI)\\.(.*)", "\\2-\\4")) |>
  tidyr::pivot_wider(names_from = "time", values_from = "logFC", values_fill = NA) |>
  tidyr::pivot_longer(cols = dplyr::ends_with("DPI"), names_to = "time", values_to = "logFC")

logfc_mtx <- logfc_df |>
  tidyr::pivot_wider(names_from = c(cluster_id, contrast, time),
                     names_sep = "#",
                     values_from = logFC) |>
  tibble::column_to_rownames("gene") |>
  as.matrix()

str_coldata <- strsplit(colnames(logfc_mtx), "#")

mtx_coldata <- data.frame(
  colidx = seq_along(colnames(logfc_mtx)),
  cell_types = factor(sapply(str_coldata, function(x) x[[1]]),
                      levels = unique(all_tissues)),
  contrast = factor(sapply(str_coldata, function(x) x[[2]]),
                    #levels = c("PNR2-MOCK", "TR4-MOCK", "PNR2-TR4")),
                    levels = c("PNR2-MOCK", "TR4-MOCK")),
  times = factor(sapply(str_coldata, function(x) x[[3]]),
                 levels = c("1DPI", "2DPI", "3DPI"))
)

mtx_coldata <- mtx_coldata[with(mtx_coldata, order(cell_types, contrast, times)),]

logfc_mtx <- logfc_mtx[, mtx_coldata$colidx]
```

在热图中用灰色表示不显著的基因，即在差异基因列表中缺失的那些。有些基因明明不显著，但 logFC 可能非常大，会干扰判断。

学习 dendsort: modular leaf ordering methods for dendrogram representations in R

学习 ComplexHeatmap 中 outliers 的处理方式，避免其影响聚类效果。

```{r fig.height=7, fig.width=10}
dend_mtx <- logfc_mtx
dend_mtx[is.na(dend_mtx)] <- 0
dend_row_hc <- dendsort::dendsort(fastcluster::hclust(dist(dend_mtx)))
```

```{r fig.height=7, fig.width=14}
ha = ComplexHeatmap::HeatmapAnnotation(
      which = "column",
      df = mtx_coldata[-1],
      col = list(
        cell_types = structure(
          ggthemes::tableau_color_pal()(
            length(levels(mtx_coldata$cell_types))),
          names = levels(mtx_coldata$cell_types)
        ),
        contrast = structure(
          c("#E41A1C", "#377EB8"),
          names = levels(mtx_coldata$contrast)
        ),
        times = structure(
          RColorBrewer::brewer.pal(3, "YlGn"),
          names = levels(mtx_coldata$times)
        )
      ))

p <- ComplexHeatmap::Heatmap(
    matrix = logfc_mtx,
    name = "Log2 Fold Change",
    column_split = mtx_coldata$cell_types,
    use_raster = TRUE,
    na_col = "grey",
    cluster_rows = dend_row_hc,
    cluster_columns = FALSE,
    show_column_names = FALSE,
    show_row_names = FALSE,
    top_annotation = ha,
    heatmap_legend_param = list(
      title_position = "leftcenter-rot",
      legend_height  = grid::unit(4, "cm")
    ))

p <- ComplexHeatmap::draw(p, merge_legend = TRUE)
```

