# 2DPI 样品的差异状态分析 (Differential State analysis)

## 数据准备

```{r message=FALSE, warning=FALSE}
library(magrittr)
library(patchwork)

obj <- readRDS("../results/IntegratedAnalysis/2DPI-All/Seurat_Object_Combined.rds")
```

## 数据概览

### 样本细胞降维度图

```{r warning=FALSE}
Seurat::DimPlot(obj, reduction = "tsne", group.by = "group")
Seurat::DimPlot(obj, reduction = "tsne", label = TRUE, repel = TRUE)
```

### 细胞身份注释

```{r fig.width=15, fig.height=8}
Seurat::DefaultAssay(obj) <- "RNA"

Seurat::DotPlot(
  obj,
  features = list(
    "FDH" = c(
      "TraesCS4B02G297500",
      "TraesCS4D02G296400",
      "TraesCS4A02G007400"
    ),
    "ATML1" = c(
      "TraesCS2A02G474000",
      "TraesCS2D02G473700"
    ),
    "DCR" = c(
      "TraesCS1A02G341300",
      "TraesCS1D02G343400"
    ),
    # EP3 是排水孔相关基因
    "EP3" = c(
      "TraesCS2A02G350700",
      "TraesCS2D02G348800",
      "TraesCS6D02G199500",
      "TraesCS6A02G216100"
    ),
    "ALMT12" = c(
      "TraesCS1D02G194000",
      "TraesCS1A02G189900",
      "TraesCS1B02G192000"
    ),
    "MYB60" = c(
      "TraesCS4A02G322200",
      "TraesCS5D02G552200"
    ),
    "HIC" = c(
      "TraesCS4D02G226100"
    ),
    "RBCS" = c(
      "TraesCS2A02G066800",
      #"TraesCS2B02G079200",
      #"TraesCS2D02G065200",
      #"TraesCS2D02G065300",
      "TraesCS5A02G165400",
      #"TraesCS5A02G165700",
      #"TraesCS5B02G162600",
      #"TraesCS5B02G162800",
      #"TraesCS5D02G169600",
      "TraesCS5D02G169900"
    ),
    "CAB3" = c(
      "TraesCS7A02G276400",
      #"TraesCS1D02G411300",
      "TraesCS1B02G317500",
      #"TraesCS7D02G276300",
      #"TraesCS5B02G353200",
      #"TraesCS5A02G350600",
      "TraesCS1A02G403300"
    ),
    "LHCB2.1" = c(
      "TraesCS5D02G329200",
      "TraesCS5B02G322900",
      "TraesCS5A02G322500"
    ),
    "CA1" = c(
      #"TraesCS7D02G443400",
      "TraesCS7B02G354800",
      "TraesCS3A02G230000",
      #"TraesCS3D02G223300",
      "TraesCS3B02G259300"
    ),
    "AOC2" = c(
      "TraesCS6D02G314300",
      "TraesCS6A02G334800",
      "TraesCS6B02G365200"
    ),
    "SULTR3;4" = c(
      "TraesCS7A02G088700",
      "TraesCS4A02G388000",
      "TraesCS7D02G084100"
    ),
    "TaGSr" = c(
      "TraesCS4B02G240900",
      "TraesCS4D02G240700",
      "TraesCS4A02G063800"
    ),
    "gl-OXO" = c(
      "TraesCS4D02G032000",
      "TraesCS4B02G033300",
      "TraesCS4A02G279200",
      "TraesCS4D02G031800"
    ),
    "TaSUT1" = c(
      "TraesCS4A02G016400",
      "TraesCS4B02G287800",
      "TraesCS4D02G286500"
    ),
    "CPIII" = c(
      "TraesCS6B02G050700",
      "TraesCS6D02G041700",
      "TraesCS6A02G036100"
    ),
    "AT1G62510" = c(
      "TraesCS2A02G424800",
      "TraesCS2B02G444500",
      #"TraesCS2D02G422700",
      #"TraesCS2D02G422800",
      "TraesCS2A02G424861"
    ),
    # Bulliform
    "PFA-DSP2" = c(
      "TraesCS5B02G163200",
      "TraesCS5D02G170400"
    )
  )
) +
  ggplot2::theme(
    axis.text.x = ggplot2::element_text(angle = 90, vjust = 0.5, hjust = 1),
    strip.text.x = ggplot2::element_text(angle = 90)
  )
```

依据 marker 基因对类群重命名：

```{r}
Seurat::DefaultAssay(obj) <- "integrated"

obj <- Seurat::RenameIdents(
  obj,
  `0` = "Mesophyll Cells 4",
  `1` = "Mesophyll Cells 1",
  `2` = "Vascular Cells 4",
  `3` = "Cortex Cells 2",
  `4` = "Mesophyll Cells 5",
  `5` = "Vascular Cells 5",
  `6` = "Mesophyll Cells 2",
  `7` = "Mesophyll Cells 3",
  `8` = "Cortex Cells 3",
  `9` = "Vascular Cells 2",
  `10` = "Epidermal Cells",
  `11` = "Vascular Cells 7",
  `12` = "Unknown",
  `13` = "Vascular Cells 6",
  `14` = "Vascular Cells 1",
  `15` = "Guardian Cells",
  `16` = "Cortex Cells 1"
)

sce <- rhapsodykit::prepare_muscat_sce(obj, group_make_names = TRUE)
```

### 相对细胞群集丰度

差异状态分析依赖于细胞亚群的大小，因此群体规模太小的亚群应该事先排除掉。

```{r}
cluster_stats <- table(sce$cluster_id, sce$sample_id)
cluster_stats
```

看看各个类群在各自样本中所占的比例：

```{r warning=FALSE}
cluster_prop <- prop.table(cluster_stats, 2)
cluster_prop_df <- reshape2::melt(
  cluster_prop,
  varnames = c("cluster_id", "sample_id"),
  value.name = "frequency",
  as.is = TRUE
)

cluster_prop_df$cluster_id <- factor(type.convert(cluster_prop_df$cluster_id))
cluster_prop_df$group_id <- sce$group_id[match(cluster_prop_df$sample_id, sce$sample_id)]

p_cluster_percent <- cluster_prop_df %>%
  ggplot2::ggplot(ggplot2::aes(
    x = sample_id, y = frequency, fill = cluster_id)) +
  ggplot2::scale_y_continuous(breaks = seq(0, 1, 0.2), expand = c(0, 0)) +
  ggplot2::facet_wrap("group_id", ncol = 1, scales = "free_y") +
  ggplot2::geom_bar(stat = "identity", col = "white",  width = 1, size = 0.2) + 
  ggplot2::scale_x_discrete(expand = c(0, 0)) +
  ggplot2::coord_flip() +
  ggplot2::theme(
    aspect.ratio = NULL, 
    panel.grid = ggplot2::element_blank(), panel.spacing = grid::unit(1, "mm"), 
    strip.text = ggplot2::element_blank(),  strip.background = ggplot2::element_blank())
```

TODO: 计算丰度变化的显著性。

> Statistical significance of differences in the proportion of clusters between separate arrays and sequencing runs was performed using a Chi-square test. (Macosko, Evan Z., Anindita Basu, Rahul Satija, James Nemesh, Karthik Shekhar, Melissa Goldman, Itay Tirosh, et al. 2015. “Highly Parallel Genome-Wide Expression Profiling of Individual Cells Using Nanoliter Droplets.” Cell 161 (5): 1202–14. https://doi.org/10.1016/j.cell.2015.05.002.)

### 样本细胞数量

```{r}
# cell counts by sample
sample_cells_df <- reshape2::melt(table(sce$sample_id), varnames = "sample_id", value.name = "n_cells")
sample_cells_df$group_id <- sce$group_id[match(sample_cells_df$sample_id, sce$sample_id)]

p_sample_cells <- sample_cells_df %>%
  ggplot2::ggplot(ggplot2::aes(
    x = sample_id, y = n_cells, fill = group_id)) +
  ggplot2::scale_y_continuous("nb. of cells", limits = c(0, 7500), 
        breaks = seq(0, 7500, 2500), expand = c(0, 0)) +
  ggplot2::facet_wrap("group_id", ncol = 1, scales = "free_y") +
  ggplot2::geom_bar(stat = "identity", col = "white",  width = 1, size = 0.2) + 
  ggplot2::scale_x_discrete(expand = c(0, 0)) +
  ggplot2::coord_flip() +
  ggplot2::theme(aspect.ratio = NULL, 
    panel.grid = ggplot2::element_blank(), panel.spacing = grid::unit(1, "mm"), 
    strip.text = ggplot2::element_blank(),  strip.background = ggplot2::element_blank())
```

我们将两幅图合并起来：

```{r fig.height=5, fig.width=14}
p_cluster_percent + p_sample_cells
```


## 差异状态分析

### 实施 DS 分析

#### 将单细胞聚合成 pseudo-bulk 数据


```{r warning=FALSE}
pb_list <- list(
  "counts" = rhapsodykit::calculate_pseudo_bulk(sce, "counts"),
  "logcounts" = rhapsodykit::calculate_pseudo_bulk(sce, "logcounts"),
  "cpm" = rhapsodykit::calculate_pseudo_bulk(sce, "cpm")
  #"vstresiduals" = rhapsodykit::calculate_pseudo_bulk(sce, "vstresiduals")
)
```

我们需要通过 MDS plot 查看一下 pseudo-bulk 是否能够反应样本间的相似性：

```{r}
mds_plot_list <- lapply(pb_list, function(pb) {
  pb_mds <- muscat::pbMDS(pb)
  pb_mds +
    ggplot2::scale_shape_manual(values = c(15, 8, 10)) +
    ggplot2::theme(legend.position = "right", legend.box = "horizontal")
})
```

首先我们看看 `logcounts` 类型 pseudo-bulk 数据：

```{r}
mds_plot_list$logcounts
```

#### 样品水平的分析: 基于聚合的方法

我们看看实验相关的信息，主要是 sample 属于哪个 group ：

```{r message=FALSE}
S4Vectors::metadata(sce)$experiment_info
```

```{r}
# run DS analysis
ds_res <- list(
  edgeR = rhapsodykit::pseudobulk_diff_state(
    sce, pb_list$counts,
    method = "edgeR",
    contrasts = c(
      "X2DPI.PNR2-X2DPI.MOCK",
      "X2DPI.TR4-X2DPI.MOCK",
      "X2DPI.PNR2-X2DPI.TR4"
    ),
    verbose = FALSE
  ),
  DESeq2 = rhapsodykit::pseudobulk_diff_state(
    sce, pb_list$counts,
    method = "DESeq2",
    contrasts = c(
      "X2DPI.PNR2-X2DPI.MOCK",
      "X2DPI.TR4-X2DPI.MOCK",
      "X2DPI.PNR2-X2DPI.TR4"
    ),
    verbose = FALSE
  )
)

str(ds_res$DESeq2$table, max.level = 1)
```

TODO: 意外的是 `muscat::pbDS` 有过滤机制，不符合要求的 subpopulation 就被移除了。暂时先不去研究过滤的条件。

```{r}
setdiff(levels(sce$cluster_id), names(ds_res$DESeq2$table[[1]]))
```


### 对结果可视化

#### 可视化 DS 基因的数量

我们可以查看一下 edgeR 和 DESeq2 两种方法发现的差异基因数量：

```{r fig.height=7, fig.width=7}
patchwork::wrap_plots(
  ds_res$edgeR %>%
    rhapsodykit::diff_state_significant(0.01, 2) %>%
    rhapsodykit::genecount_diff_state() +
    ggplot2::ggtitle("edgeR"),
  ds_res$DESeq2 %>%
    rhapsodykit::diff_state_significant(0.01, 2) %>%
    rhapsodykit::genecount_diff_state() +
    ggplot2::ggtitle("DESeq2"),
  ncol = 1,
  guides = "collect"
)
```

我们首先画一下 `X2DPI.PNR2-X2DPI.MOCK` 火山图：

```{r}
df_to_plot <- ds_res$DESeq2 %>% rhapsodykit::diff_state_format()
dplyr::filter(df_to_plot, contrast == "X2DPI.PNR2-X2DPI.MOCK") %>%
  rhapsodykit::volcano_diff_state(title = "2DPI-PNR2 vs. 2DPI-MOCK")
```

然后看一下 `X2DPI.TR4-X2DPI.MOCK` 的火山图：

```{r}
dplyr::filter(df_to_plot, contrast == "X2DPI.TR4-X2DPI.MOCK") %>%
  rhapsodykit::volcano_diff_state(title = "2DPI-TR4 vs. 2DPI-MOCK")
```

最后比较一下 `X2DPI.PNR2-X2DPI.TR4`:

```{r}
dplyr::filter(df_to_plot, contrast == "X2DPI.PNR2-X2DPI.TR4") %>%
  rhapsodykit::volcano_diff_state(title = "2DPI-PNR2 vs. 2DPI-TR4")
```


#### 细胞亚群间差异状态的一致性

我们可以通过集合的可视化来看看细胞亚群之间的 DS 基因是否有交集：

```{r}
tbl_filtered <- ds_res$DESeq2 %>%
  rhapsodykit::diff_state_significant(0.01, 2) %>%
  .$table %>% .[["X2DPI.PNR2-X2DPI.MOCK"]]

de_gs_by_k <- purrr::map(tbl_filtered, "gene")

ds_res$DESeq2 %>%
  rhapsodykit::diff_state_significant(0.01, 2) %>%
  rhapsodykit::diff_state_pull("X2DPI.PNR2-X2DPI.MOCK", 1:12, "gene") %>%
  rhapsodykit::from_list_to_upset() %>%
  rhapsodykit::upset_dataframe() %>%
  rhapsodykit::upset_plot()

ds_res$DESeq2 %>%
  rhapsodykit::diff_state_significant(0.01, 2) %>%
  rhapsodykit::diff_state_pull("X2DPI.TR4-X2DPI.MOCK", 1:12, "gene") %>%
  rhapsodykit::from_list_to_upset() %>%
  rhapsodykit::upset_dataframe() %>%
  rhapsodykit::upset_plot()

ds_res$DESeq2 %>%
  rhapsodykit::diff_state_significant(0.01, 2) %>%
  rhapsodykit::diff_state_pull("X2DPI.PNR2-X2DPI.TR4", 1:12, "gene") %>%
  rhapsodykit::from_list_to_upset() %>%
  rhapsodykit::upset_dataframe() %>%
  rhapsodykit::upset_plot()
```

```{r}
plot_updown <- function(contrast) {
  up_list <- ds_res$DESeq2 %>%
    rhapsodykit::diff_state_filter(logFC > 0) %>%
    rhapsodykit::diff_state_significant(0.01, 2) %>%
    rhapsodykit::diff_state_pull(contrast, 1:12, "gene")
  
  down_list <- ds_res$DESeq2 %>%
    rhapsodykit::diff_state_filter(logFC < 0) %>%
    rhapsodykit::diff_state_significant(0.01, 2) %>%
    rhapsodykit::diff_state_pull(contrast, 1:12, "gene")
  
  rhapsodykit::upset_updown_regulated(up_list, down_list)
}

plot_updown("X2DPI.PNR2-X2DPI.MOCK") + ggplot2::ggtitle("2DPI PNR2 vs. MOCK")
plot_updown("X2DPI.TR4-X2DPI.MOCK") + ggplot2::ggtitle("2DPI TR4 vs. MOCK")
plot_updown("X2DPI.PNR2-X2DPI.TR4") + ggplot2::ggtitle("2DPI PNR2 vs. TR4")
```

#### 样本间差异状态的一致性

```{r}
upset_contrasts <- function(input) {
  sets_list <- input$table %>%
    purrr::map(function(x) {
      purrr::map(x, "gene") %>%
        purrr::compact()
    })
  
  sets_metadata <- sets_list %>%
    purrr::imap(function(contrast, key) {
      cbind(
        sets = sprintf("%s [%s]", key, names(contrast)),
        contrast = key,
        cluster = names(contrast)
      )
    }) %>%
    do.call(rbind, .) %>%
    as.data.frame()

  sets_to_plot <- sets_list %>%
    unlist(recursive = FALSE)
  
  names(sets_to_plot) <- sets_metadata$sets

  sets_to_plot %>%
    rhapsodykit::from_list_to_upset() %>%
    UpSetR::upset(
      nsets = 6,
      cutoff = 5,
      text.scale = c(1, 1, 1, 1, 1, 0.75),
      set.metadata = list(
        data = sets_metadata,
        plots = list(
          list(
            type = "matrix_rows",
            column = "contrast",
            alpha = 0.4,
            colors = c(
              "X2DPI.PNR2-X2DPI.MOCK" = "green",
              "X2DPI.TR4-X2DPI.MOCK" = "navy",
              "X2DPI.PNR2-X2DPI.TR4" = "yellow"
            )
          )
        )
      )
    )
}

ds_res$DESeq2 %>%
  rhapsodykit::diff_state_significant(0.01, 2) %>%
  upset_contrasts()
```

## 基因功能分析

### 基于 GO 注释的功能分析

首先我们要拿到小麦 GO 注释数据：

```{r}
# Get GO Data
# quote="" argument is neccessary for reading the complete table,
# because it disables quoting.
go_data <- read.table(
  "../../data/GOData.tsv", header = TRUE, sep = "\t", quote = "") %>%
  tibble::as_tibble()

go_data <- go_data %>%
  dplyr::filter(GO_ID != "")

go2gene <- go_data %>%
  dplyr::select(GO_ID, Gene_ID)
go2name <- go_data %>%
  dplyr::select(GO_ID, GO_Name)
```

#### 差异状态基因的生物主题比较

我们可以试试 clusterProfiler 生物主题比较功能：

```{r fig.height=7, fig.width=7}
compare_list <- list(
  "VCs_6.2DPI.TR4.MOCK" = ds_res$DESeq2 %>%
    rhapsodykit::diff_state_significant(0.01, 2) %>%
    rhapsodykit::diff_state_pull("X2DPI.TR4-X2DPI.MOCK", "Vascular Cells 6", "gene"),
  "VCs_6.2DPI.PNR2.MOCK" = ds_res$DESeq2 %>%
    rhapsodykit::diff_state_significant(0.01, 2) %>%
    rhapsodykit::diff_state_pull("X2DPI.PNR2-X2DPI.MOCK", "Vascular Cells 6", "gene"),
  "MCs_2.2DPI.TR4.MOCK" = ds_res$DESeq2 %>%
    rhapsodykit::diff_state_significant(0.01, 2) %>%
    rhapsodykit::diff_state_pull("X2DPI.TR4-X2DPI.MOCK", "Mesophyll Cells 2", "gene"),
  "MCs_2.2DPI.PNR2.MOCK" = ds_res$DESeq2 %>%
    rhapsodykit::diff_state_significant(0.01, 2) %>%
    rhapsodykit::diff_state_pull("X2DPI.PNR2-X2DPI.MOCK", "Mesophyll Cells 2", "gene")
)

comp_enr <- clusterProfiler::compareCluster(compare_list, fun = "enricher", TERM2GENE = go2gene, TERM2NAME = go2name)

enrichplot::dotplot(comp_enr, showCategory = 10, by = "count") +
  #ggplot2::scale_y_discrete(expand = ggplot2::expansion(c(0, 0), c(1, 1))) +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1))
```

然后看看 GO 词条之间的关联：

```{r fig.height=7, fig.width=10}
comp_enr %>%
  enrichplot::pairwise_termsim() %>%
  enrichplot::emapplot(legend_n = 2)
```

#### 对差异状态基因做富集分析

要做富集分析我们首先要拿到一个基因列表。

```{r fig.height=7, fig.width=14}
p1 <- ds_res$DESeq2 %>%
  rhapsodykit::diff_state_significant(0.01, 2) %>%
  rhapsodykit::diff_state_pull("X2DPI.TR4-X2DPI.MOCK", "Vascular Cells 6", "gene") %>%
  rhapsodykit::enrichment_analysis(go_data) %>%
  rhapsodykit::enrich_barplot(x = "Count", show_category = 20, title = "Vascular Cells 6 of 2DPI-TR4 vs. 2DPI-MOCK")

p2 <- ds_res$DESeq2 %>%
  rhapsodykit::diff_state_significant(0.01, 2) %>%
  rhapsodykit::diff_state_pull("X2DPI.PNR2-X2DPI.MOCK", "Vascular Cells 6", "gene") %>%
  rhapsodykit::enrichment_analysis(go_data) %>%
  rhapsodykit::enrich_barplot(x = "Count", show_category = 20, title = "Vascular Cells 6 of 2DPI-PNR2 vs. 2DPI-MOCK")

p3 <- ds_res$DESeq2 %>%
  rhapsodykit::diff_state_significant(0.01, 2) %>%
  rhapsodykit::diff_state_pull("X2DPI.PNR2-X2DPI.TR4", "Vascular Cells 6", "gene") %>%
  rhapsodykit::enrichment_analysis(go_data) %>%
  rhapsodykit::enrich_barplot(x = "Count", show_category = 20, title = "Vascular Cells 6 of 2DPI-PNR2 vs. 2DPI-TR4")

p1 + p2 + p3
```

现在我们可以从基因与 GO 的关系绘图：

```{r fig.height=8, fig.width=12}
sig_genes <- ds_res$DESeq2 %>%
  rhapsodykit::diff_state_significant(0.01, 2)
genes <- sig_genes %>%
  rhapsodykit::diff_state_pull("X2DPI.TR4-X2DPI.MOCK", "Vascular Cells 6", "gene")
fc <- sig_genes %>%
  rhapsodykit::diff_state_pull("X2DPI.TR4-X2DPI.MOCK", "Vascular Cells 6", "logFC")
names(fc) <- genes

enr <- genes %>%
  clusterProfiler::enricher(TERM2GENE = go2gene, TERM2NAME = go2name)

rhapsodykit::enrich_cnetplot(enr, fold_change = fc, show_category = 20)
```

如果我们不将基因显示为 node ，而是将 GO 词条之间的链接定义为重叠基因的数量，也就是说 GO 词条之间重叠基因越多他们在图上的距离也就越近。这个功能叫做 Enrichment Map 。

```{r fig.height=7, fig.width=7}
enr %>%
  enrichplot::pairwise_termsim() %>%
  enrichplot::emapplot(cex_label_category = 0.6)
```


当要展示很多 GO 词条时，网络图就会变得非常复杂，我们可以试试 clusterProfiler 的 Heatmap-like functional classification 功能。不过这个热图似乎没有聚类，也就不是很方便人类阅读和理解了。TODO: 

```{r fig.height=7, fig.width=10}
enr %>%
  clusterProfiler::heatplot(foldChange = fc) +
  ggplot2::theme(axis.text.x = ggplot2::element_blank())
```

使用 UpSet 图似乎看起来更有意义。

```{r}
enr %>% enrichplot::upsetplot()
```

#### 对差异状态基因做 GSEA 分析

基于差异基因表达的富集分析可以找到差异较大的基因，但它发现不了那些虽然差异较小的，但在一组相关基因中以协调的方式得到证明的基因。因此 Gene Set Enrichment Analysis (GSEA) 直接关注这个问题。

GSEA 将一个基因集内各基因的统计数据聚合起来，因此有可能检测到一个预定义的基因集中所有基因以小而协调的方式变化的情况。因为许多相关的表型差异很可能是由一组基因中微小但一致的变化表现出来的。

基因是根据其表型进行排名的。给定一个预先的基因集 S (例如，共享相同 DO 类别的基因) ，GSEA 的目标是确定 S 的成员是随机分布在整个排序好的基因列表 (L)，还是主要存在于顶部或底部。.

For gene set enrichment analysis, we need a ranked list of genes. Suppose you are importing your own data from a csv file and the file contains two columns, one for gene ID (no duplicated allowed) and another one for fold change, you can prepare your own geneList via the following command:

```{r}
## 我们不需要依据 FDR 来筛选基因，只需要 logFC 来排序
gsea_df <- ds_res$DESeq2 %>%
  rhapsodykit::diff_state_pull("X2DPI.TR4-X2DPI.MOCK", "Vascular Cells 6", c("gene", "logFC"))

## feature 1: numeric vector
geneList <- gsea_df[,2]

## feature 2: named vector
names(geneList) <- as.character(gsea_df[,1])

## feature 3: decreasing order
geneList <- sort(geneList, decreasing = TRUE)

```

问题是从哪里去找通路数据库呢？似乎 [Plant Reactome](https://plantreactome.gramene.org/) 很不错的样子。

似乎可以用 `clusterProfiler::GSEA()` 函数：

```{r}
gsea <- clusterProfiler::GSEA(geneList, TERM2GENE = go2gene, TERM2NAME = go2name)

rhapsodykit::enrich_cnetplot(gsea, fold_change = geneList, show_category = 20)
```

```{r fig.height=7, fig.width=7}
gsea %>%
  enrichplot::pairwise_termsim() %>%
  enrichplot::emapplot(cex_label_category = 0.6)
```

```{r fig.height=7, fig.width=10}
gsea %>%
  enrichplot::upsetplot()
```


```{r fig.height=7, fig.width=14}
gsea %>%
  enrichplot::ridgeplot()
```

```{r}
gsea %>%
  enrichplot::gseaplot(geneSetID = 3, title = gsea$Description[3])
```

```{r fig.height=7, fig.width=14}
gsea %>%
  enrichplot::gseaplot2(geneSetID = 1:5)
```

```{r}
gsea %>%
  enrichplot::gsearank(geneSetID = 3, title = gsea$Description[3])
```
