---
title: "Pseudobulk Gene Expression Analysis"
author: "Altair Wei"
date: '2022-08-03'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
source("../scripts/LoadUtils.R", chdir = TRUE)
```

```{r message=FALSE, warning=FALSE, results="hide"}
library(org.Taestivum.iwgsc.db)
orgdb <- org.Taestivum.iwgsc.db
```

## 生成 Pseudobulk 数据

```{r}
obj <- readRDS(Sys.glob(
  "../results/ObjectCache/IntegrationAndAnnotation/obj_annotated_*.rds"))
```

CPM 其实与 LogNormalize 非常接近，只是不进行 log 转化。

```{r}
expr <- xfun::cache_rds(
  file = "pseudobulk_cpm_mean.rds",
  dir = "../results/ObjectCache/Pseudobulk/",
  expr = pseudobulk(obj, type = "cpm", fun = "mean")
)
```

```{r}
df <- xfun::cache_rds(
  file = "pseudobulk_df_cpm_mean.rds",
  dir = "../results/ObjectCache/Pseudobulk/",
  rerun = TRUE,
  expr = as.data.frame(expr) |>
    tibble::rownames_to_column("Gene") |>
    tidyr::pivot_longer(-Gene, names_to = "bulk", values_to = "CPM") |>
    tidyr::separate(bulk, sep = "\\.", into = c("treatment", "time", "cellType"))
)
```


## 读取通路数据

```{r}
filelist <- Sys.glob("../data/pathways/*.txt")
names(filelist) <- basename(filelist) |> stringr::str_remove(stringr::fixed(".txt"))

pathways <- lapply(filelist, function(filename) {
  genes <- stringr::str_trim(readLines(filename))
  genes[genes != ""]
})
```

## 通路活性计算

```{r}
expr_gsva <- GSVA::gsva(
  expr, pathways, method = "gsva", kcdf = "Gaussian")

gsva_df <- as.data.frame(expr_gsva) |>
    tibble::rownames_to_column("Pathway") |>
    tidyr::pivot_longer(-Pathway, names_to = "bulk", values_to = "ES") |>
    tidyr::separate(bulk, sep = "\\.", into = c("treatment", "time", "cellType"))
```

```{r}
gsva_df$cellType <- factor(gsva_df$cellType, levels = c(
  "CC", "BS", "MPV_1", "MPV_2",
  "Va_1", "Va_2", "Va_3", "Va_4",
  "Me_1", "Me_2", "Me_3", "Me_4", "Me_5", "Me_6",
  "Ep_1", "Ep_2", "Gu"))
```

### 光合与氧化磷酸化

```{r fig.height=16, fig.width=16}
dplyr::filter(gsva_df,
    Pathway %in% c("OxiPhos", "PhotosynAnt", "PhotosynCfix", "PhotosynPSEA")) |>
  ggplot2::ggplot(
    ggplot2::aes(
        x = cellType,
        y = ES,
        group = treatment,
        color = treatment,
        fill = treatment
      )) +
  ggplot2::geom_polygon(size = 1, alpha = .1) +
  ggplot2::geom_point() +
  see::coord_radar() +
  ggplot2::facet_grid(vars(Pathway), vars(time)) +
  see::theme_radar() +
  NULL
```

### 苯丙烷代谢

```{r}
PAL <- c("Aromatic amino acid", "General phenylpropanoid", "Lignin biosynthesis", "Flavanones", "Polyamine", "HCAAs")
pal_df <- dplyr::filter(gsva_df, Pathway %in% PAL)
pal_df$Pathway <- factor(pal_df$Pathway, levels = PAL)
```

```{r fig.height=24, fig.width=16}
pal_df |>
  ggplot2::ggplot(
    mapping = ggplot2::aes(
        x = cellType,
        y = ES,
        group = treatment,
        color = treatment,
        fill = treatment
      )) +
  ggplot2::geom_polygon(size = 1, alpha = .1) +
  ggplot2::geom_point() +
  see::coord_radar() +
  ggplot2::facet_grid(vars(Pathway), vars(time)) +
  see::theme_radar() +
  NULL
```

### 免疫通路

```{r fig.height=16, fig.width=16}
dplyr::filter(gsva_df,
    Pathway %in% c("CalloseSyn", "TaNPR", "MAPK", "PR")) |>
  ggplot2::ggplot(
    ggplot2::aes(
        x = cellType,
        y = ES,
        group = treatment,
        color = treatment,
        fill = treatment
      )) +
  ggplot2::geom_polygon(size = 1, alpha = .1) +
  ggplot2::geom_point() +
  see::coord_radar() +
  ggplot2::facet_grid(vars(Pathway), vars(time)) +
  see::theme_radar() +
  NULL
```

### NLRs

```{r fig.height=16, fig.width=16}
dplyr::filter(gsva_df,
    Pathway %in% c("CCNBLRR", "NBLRR", "RxNonly", "RLK")) |>
  ggplot2::ggplot(
    ggplot2::aes(
        x = cellType,
        y = ES,
        group = treatment,
        color = treatment,
        fill = treatment
      )) +
  ggplot2::geom_polygon(size = 1, alpha = .1) +
  ggplot2::geom_point() +
  see::coord_radar() +
  ggplot2::facet_grid(vars(Pathway), vars(time)) +
  see::theme_radar() +
  NULL
```

### 信号

```{r fig.height=16, fig.width=16}
dplyr::filter(gsva_df,
    Pathway %in% c("GOCaSignaling", "GOSAR", "Myb", "WRKY")) |>
  ggplot2::ggplot(
    ggplot2::aes(
        x = cellType,
        y = ES,
        group = treatment,
        color = treatment,
        fill = treatment
      )) +
  ggplot2::geom_polygon(size = 1, alpha = .1) +
  ggplot2::geom_point() +
  see::coord_radar() +
  ggplot2::facet_grid(vars(Pathway), vars(time)) +
  see::theme_radar() +
  NULL
```

### 糖酵解

```{r fig.height=8, fig.width=16}
dplyr::filter(gsva_df,
    Pathway %in% c("glycolysis", "gluconeogenesis")) |>
  ggplot2::ggplot(
    ggplot2::aes(
        x = cellType,
        y = ES,
        group = treatment,
        color = treatment,
        fill = treatment
      )) +
  ggplot2::geom_polygon(size = 1, alpha = .1) +
  ggplot2::geom_point() +
  see::coord_radar() +
  ggplot2::facet_grid(vars(Pathway), vars(time)) +
  see::theme_radar() +
  NULL
```

### 其他

```{r fig.height=16, fig.width=16}
dplyr::filter(gsva_df,
    Pathway %in% c("TaSUT", "Sweet", "Suberin", "NAC")) |>
  ggplot2::ggplot(
    ggplot2::aes(
        x = cellType,
        y = ES,
        group = treatment,
        color = treatment,
        fill = treatment
      )) +
  ggplot2::geom_polygon(size = 1, alpha = .1) +
  ggplot2::geom_point() +
  see::coord_radar() +
  ggplot2::facet_grid(vars(Pathway), vars(time)) +
  see::theme_radar() +
  NULL
```
