---
title: "Slides for PhD Defense"
author: "Altair Wei"
date: "2023-01-31"
output: html_document
---

```{r setup, include=FALSE}
library(magrittr)
source("../scripts/LoadUtils.R", chdir = TRUE)
knitr::opts_chunk$set(dev = "png")
scv <- reticulate::import("scvelo")
plt <- reticulate::import("matplotlib.pyplot")
```

```{python}
import scvelo as scv
import matplotlib.pyplot as plt
```

## Load Data

### Annotations

```{r message=FALSE, warning=FALSE, results="hide"}
library(org.Taestivum.iwgsc.db)
orgdb <- org.Taestivum.iwgsc.db
```

```{r}
blue_to_orange <- rev(ggthemes::ggthemes_data[[
  c("tableau", "color-palettes", "ordered-diverging", "Orange-Blue Diverging")]][["value"]])
```

### MOCK samples

```{r}
obj_mock <- readRDS(Sys.glob("../results/ObjectCache/IntegrationAndAnnotation/obj_mock_annotated_*.rds"))
sce_mock <- Seurat::as.SingleCellExperiment(obj_mock, assay = "RNA")
sce_mock <- muscat::prepSCE(sce_mock,
  kid = "ident", # subpopulation assignments
  gid = "group", # group IDs (ctrl/stim)
  sid = "sample", # sample IDs (ctrl/stim.1234)
  drop = FALSE
)
```

```{r}
mock_celltypes <- c(
  "Gu", "Ep_α", "Ep_β",
  "Me_α", "Me_β", "Me_γ", "Me_δ", "Me_ε",
  "BS", "CC", "Va_α", "Va_β", "Va_γ", "Va_δ", "MPV")

mock_celltype_names <- c(
  "Guard cell", "Epidermal α", "Epidermal β",
  
  "Mesophyll α", "Mesophyll β", "Mesophyll γ",
  "Mesophyll δ", "Mesophyll ε",
  
  "Vascular bundle sheath", "Phloem companion cell",
  "Vascular α", "Vascular β", "Vascular γ",
  "Vascular δ", "Midvein provascular cell"
)

mock_celltype_colors <- ggthemes::tableau_color_pal("Tableau 20")(length(mock_celltypes))
#mock_celltype_colors <- scater:::.get_palette("tableau20")[seq_along(mock_celltypes)]
names(mock_celltype_colors) <- mock_celltypes
scales::show_col(mock_celltype_colors)
```

```{r}
mock_body_layers <- c(
  "Gu"   = "L1",
  "Ep_α" = "L1",
  "Ep_β" = "L1",
  "Me_α" = "L2",
  "Me_β" = "L2",
  "Me_γ" = "L2",
  "Me_δ" = "L2",
  "Me_ε" = "L2",
  "BS"   = "L3",
  "CC"   = "L3",
  "Va_α" = "L3",
  "Va_β" = "L3",
  "Va_γ" = "L3",
  "Va_δ" = "L3",
  "MPV"  = "L3"
)
```

```{r}
sce_mock$cluster_id <- forcats::fct_relevel(sce_mock$cluster_id, mock_celltypes)
sce_mock$body_layer <- do.call(
  dplyr::recode, c(list(.x = sce_mock$cluster_id), as.list(mock_body_layers)))
Seurat::Idents(obj_mock) <- forcats::fct_relevel(Seurat::Idents(obj_mock), mock_celltypes)
```

### All samples

```{r}
obj <- readRDS(Sys.glob("../results/ObjectCache/IntegrationAndAnnotation/obj_annotated_*.rds"))
sce <- Seurat::as.SingleCellExperiment(obj, assay = "RNA")
sce <- muscat::prepSCE(sce,
  kid = "ident", # subpopulation assignments
  gid = "group", # group IDs (ctrl/stim)
  sid = "sample", # sample IDs (ctrl/stim.1234)
  drop = FALSE
)
```

```{r}
all_celltypes <- c(
  "Gu", "Ep_1", "Ep_2",
  "Me_1", "Me_2", "Me_3", "Me_4", "Me_5", "Me_6",
  "Va_1", "Va_2", "Va_3", "Va_4",
  "BS", "CC", "MPV_1", "MPV_2")

all_celltype_colors <- ggthemes::tableau_color_pal("Tableau 20")(length(all_celltypes))
#mock_celltype_colors <- scater:::.get_palette("tableau20")[seq_along(mock_celltypes)]
names(all_celltype_colors) <- all_celltypes
scales::show_col(all_celltype_colors)
```

```{r}
all_body_layers <- c(
  "Gu"    = "L1",
  "Ep_1"  = "L1",
  "Ep_2"  = "L1",
  "Me_1"  = "L2",
  "Me_2"  = "L2",
  "Me_3"  = "L2",
  "Me_4"  = "L2",
  "Me_5"  = "L2",
  "Me_6"  = "L2",
  "Va_1"  = "L2",
  "Va_2"  = "L2",
  "Va_3"  = "L2",
  "Va_4"  = "L2",
  "BS"    = "L3",
  "CC"    = "L3",
  "MPV_1" = "L3",
  "MPV_2" = "L3"
)

all_tissues <- c(
  Gu   = "Stomata",
  Ep_1 = "Epidermis",
  Ep_2 = "Epidermis",
  Me_1 = "Chlorenchyma",
  Me_2 = "Chlorenchyma",
  Me_3 = "Chlorenchyma",
  Me_4 = "Parenchyma",
  Me_5 = "Parenchyma",
  Me_6 = "Parenchyma",
  Va_1 = "Chlorenchyma",
  Va_2 = "Outer sheath",
  Va_3 = "Outer sheath",
  Va_4 = "Outer sheath",
  BS = "Inner sheath",
  CC = "Phloem",
  MPV_1 = "Procambium",
  MPV_2 = "Procambium"
)
```

```{r}
sce$cluster_id <- forcats::fct_relevel(sce$cluster_id, all_celltypes)
Seurat::Idents(obj) <- forcats::fct_relevel(Seurat::Idents(obj), all_celltypes)
sce$body_layer <- do.call(
  dplyr::recode, c(list(.x = sce$cluster_id), as.list(all_body_layers)))
sce$tissue <- do.call(
  dplyr::recode, c(list(.x = sce$cluster_id), as.list(all_tissues)))
```

### DS Results

```{r}
logfc_data <- readRDS(Sys.glob(
  "../results/ObjectCache/DifferentialState/ds_deg_logfc_data_*.rds"))
```

### Gene List

```{r}
filelist <- Sys.glob("../data/pathways/*.txt")
names(filelist) <- basename(filelist) |> stringr::str_remove(stringr::fixed(".txt"))

pathways <- lapply(filelist, function(filename) {
  genes <- stringr::str_trim(readLines(filename))
  genes[genes != ""]
})
```

### GSVA data

```{r}
gsva_df <- readRDS(Sys.glob(
  "../results/ObjectCache/Pseudobulk/pseudobulk_8celltypes_gsva_df_*.rds"))
```

## Quality Control and Integration

```{r fig.height=13, fig.width=14}
p <- scater::plotReducedDim(
    sce_mock[, sample(seq_len(ncol(sce_mock)))],
    dimred = "UMAP",
    colour_by = "sample_id",
    other_fields = "time",
    point_size = 1,
    point_alpha = 1,
    theme_size = 20) +
  legend_override(
    key = "color", values = list(size = 4),
    title = "sample") +
  ggplot2::coord_fixed() +
  ggplot2::facet_wrap(~ time) +
  ggthemes::scale_color_tableau("Tableau 20") +
  empty_strip() +
  ggplot2::theme(
    legend.position = "bottom",
    legend.justification = "center")

p
```

## Cell annotation

#### Cluster and resolution tree

```{r}
library(clustree)
obj_mock$RNA_snn_res.1.2 <- NULL
obj_mock$RNA_snn_res.1.4 <- NULL
obj_mock$RNA_snn_res.1.6 <- NULL
```

```{r fig.height=8, fig.width=14}
p1 <- scater::plotReducedDim(sce_mock,
    dimred = "UMAP",
    point_size = 1,
    point_alpha = 1,
    colour_by = "RNA_snn_res.0.6",
    text_size = 6,
    theme_size = 18) +
  ggplot2::guides(
    color = ggplot2::guide_legend(
      override.aes = list(size = 4),
      title = "cluster")) +
  ggplot2::coord_fixed() +
  theme_dimred()

p2 <- clustree::clustree(
    obj_mock, prefix = "RNA_snn_res.",
    edge_width = 1) +
  legend_override("color", list(size = 4), title = "resolution")

(p <- p1 + p2 +
    patchwork::plot_layout(widths = c(3, 2)))
```

#### Marker and LCM-seq

```{r}
marker_cosg <- COSG::cosg(
 obj_mock,
 groups = 'all',
 assay = 'RNA',
 slot = 'data',
 mu = 1,
 n_genes_user = 300
)
```

```{r}
LCM_design <- c(
  "VC_1" = "vb.",
  "VC_2" = "vb.",
  "VC_3" = "vb.",
  "CC_1" = "bp.",
  "CC_2" = "bp.",
  "CC_3" = "bp.",
  "MC3_1" = "dp.",
  "MC3_2" = "dp.",
  "MC3_3" = "dp.",
  "MC_1" = "me.",
  "MC_2" = "me.",
  "MC_3" = "me.",
  "EC_1" = "ep.",
  "EC_2" = "ep.",
  "EC_3" = "ep."
)

CIBER <- deconvLCM(
  seurat = obj_mock,
  lcm_file = "../results/LCMSeq/gene_quanti/counts/all.featureCounts",
  markers = unique(unlist(marker_cosg$names)),
  design = LCM_design,
  mc.cores = 4
)
```

```{r fig.height=12, fig.width=14}
p1 <- Seurat::DotPlot(obj_mock, features = KNOWN_MARKERS) +
  ggplot2::theme(
    axis.text.x = ggplot2::element_blank(),
    strip.text.x = ggplot2::element_text(size = 11),
    legend.position = "bottom",
    legend.title = ggplot2::element_text(vjust = 1)
  )

p2 <- deconvScatter(CIBER, LCM_design, ncol = 5, theme_size = 14) +
  ggplot2::theme(
    axis.text.x = ggplot2::element_text(angle = 0, hjust = 0.5, vjust = 0.5))

(p <- p1 / p2 +
    patchwork::plot_layout(heights = c(1, 1)))
```

```{r}
mtx_tpm <- xfun::cache_rds(
  file = "lcmseq_tpm_mtx.rds",
  dir = "../results/ObjectCache/IntegrationAndAnnotation/",
  expr = local({
    LCM_samples <- c(
      "VC_1" = "vb.1",
      "VC_2" = "vb.2",
      "VC_3" = "vb.3",
      "CC_1" = "bp.1",
      "CC_2" = "bp.2",
      "CC_3" = "bp.3",
      "MC3_1" = "dp.1",
      "MC3_2" = "dp.2",
      "MC3_3" = "dp.3",
      "MC_1" = "me.1",
      "MC_2" = "me.2",
      "MC_3" = "me.3",
      "EC_1" = "ep.1",
      "EC_2" = "ep.2",
      "EC_3" = "ep.3"
    )

    tpm_files <- Sys.glob("../results/LCMSeq/gene_quanti/tpm/*.out")
    names(tpm_files) <- stringr::str_remove(
      basename(Sys.glob("../results/LCMSeq/gene_quanti/tpm/*.out")),
      stringr::fixed(".out"))

    df <- mapply(
      SIMPLIFY = FALSE,
      USE.NAMES = FALSE,
      tpm_files,
      names(tpm_files),
      FUN = function(fi, name) {
        readr::read_tsv(fi, show_col_types = FALSE) |>
          dplyr::select(Gene_Id, TPM) |>
          dplyr::mutate(Sample = name)
      }) |> dplyr::bind_rows()

    df_tpm <- df |>
      dplyr::mutate(Sample = dplyr::recode(Sample, !!!LCM_samples)) |>
      dplyr::group_by(Gene_Id, Sample) |>
      dplyr::summarise(n = length(TPM), TPM = list(TPM)) |>
      # Filter out same chloroplast and mitochondria genes
      dplyr::filter(n == 1) |>
      tidyr::unnest(TPM) |>
      dplyr::select(Gene_Id, Sample, TPM) |>
      tidyr::pivot_wider(names_from = Sample, values_from = TPM, values_fill = 0)

    df_tpm <- df_tpm[, c("Gene_Id", LCM_samples)]

    writexl::write_xlsx(df_tpm, path = "../results/DataPlots/lcmseq_mock_tpm.xlsx")

    mtx_tpm <- df_tpm |>
      tibble::column_to_rownames("Gene_Id") |>
      as.matrix()

    mtx_tpm
  })
)
```


### Known marker genes

Suc2 (AT1G22710) for Companion cell: no homologs in wheat, rice and maize

SEOR1 (At3g01680), SEOR2 (At3g01670), RTM1 (At1g05760) and RTM2 (At5g04890) for Sieve element

#### Phloem parenchyma

SWEET11 (AT3G48740) and SWEET12 (AT5G23660) for phloem parenchyma

```{r}
phloem_parenchyma <- list(
  "SWEET11" = c(
    "TraesCS6B02G015200",
    "TraesCSU02G206000",
    "TraesCS6D02G009500",
    "TraesCS6D02G009600",
    "TraesCS6D02G367400",
    "TraesCS6D02G009700",
    "TraesCS6B02G421700",
    "TraesCS6B02G015300",
    "TraesCS6A02G009100",
    "TraesCS6D02G012100",
    "TraesCS6A02G009000",
    "TraesCS6B02G015100",
    "TraesCS1A02G050400",
    "TraesCS7D02G263100",
    "TraesCS4B02G182000",
    "TraesCS5A02G289200",
    "TraesCS4A02G122100",
    "TraesCS6D02G367200",
    "TraesCS7A02G147300",
    "TraesCS6B02G421500",
    "TraesCS6B02G421600",
    "TraesCS4D02G183800",
    "TraesCS6A02G382500",
    "TraesCS7B02G050500",
    "TraesCS6A02G382400",
    "TraesCS6B02G421800",
    "TraesCS7D02G149000",
    "TraesCS1B02G066900",
    "TraesCS7B02G160000",
    "TraesCS5B02G288600",
    "TraesCS1D02G052700",
    "TraesCS7A02G261100",
    "TraesCS5D02G297100",
    "TraesCS6A02G382600",
    "TraesCS6D02G367300"
  ),
  "OsNPF2.2" = c(
    "TraesCS2D02G044200",
    "TraesCS5B02G039100",
    "TraesCS5A02G004400",
    "TraesCS7D02G076900",
    "TraesCS5A02G037900",
    "TraesCS5D02G045300",
    "TraesCS2A02G045500",
    "TraesCS5B02G001100",
    "TraesCS2D02G044000",
    "TraesCS2B02G057600",
    "TraesCS2B02G057700",
    "TraesCS5D02G012500"
  )
)

Seurat::DotPlot(obj_mock, features = phloem_parenchyma) +
  ggplot2::coord_flip() +
  ggplot2::facet_grid(
    facets = feature.groups ~ .,
    scales = "free_y",
    space = "free_y") +
  ggplot2::theme(
    axis.text.x = ggplot2::element_text(angle = 45, hjust = 1),
    strip.text.y = ggplot2::element_text(angle = 0)
  )
```

```{r}
Seurat::FeaturePlot(
    obj_mock, features = "TraesCS6D02G367400",
    reduction = "umap", order = TRUE) +
  ggplot2::coord_fixed() +
  theme_dimred()
```

```{r}
Seurat::FeaturePlot(
    obj_mock, features = "TraesCS5B02G001100",
    reduction = "umap", order = TRUE) +
  ggplot2::coord_fixed() +
  theme_dimred()
```


#### Companion cell

```{r}
companion_cell <- list(
  # Same as TaSUT1
  "OsSUT1" = c(
    "TraesCS4A02G016400",
    "TraesCS4D02G286500",
    "TraesCS4B02G287800"
  ),
  "OsDOF11" = c(
    "TraesCS6A02G274000",
    "TraesCS6B02G301500",
    "TraesCS6D02G254200"
  )
)

Seurat::DotPlot(obj_mock, features = companion_cell) +
  ggplot2::coord_flip() +
  ggplot2::facet_grid(
    facets = feature.groups ~ .,
    scales = "free_y",
    space = "free_y") +
  ggplot2::theme(
    axis.text.x = ggplot2::element_text(angle = 45, hjust = 1),
    strip.text.y = ggplot2::element_text(angle = 0)
  )
```


#### Bundle sheath

SCL23 (AT5G41920) and SCR (AT3G54220) for bundle sheath

```{r}
bundle_sheath <- list(
  "SCL23" = c(
    "TraesCS2A02G200700",
    "TraesCS2D02G208500",
    "TraesCS2B02G228000"
  ),
  "SWEET13" = c(
    "TraesCS6A02G382400",
    "TraesCS6B02G421700",
    "TraesCS6B02G421800",
    "TraesCS6D02G367300",
    "TraesCS6D02G367400",
    "TraesCS6A02G382500",
    "TraesCS6A02G382600",
    "TraesCS6B02G421500",
    "TraesCS6B02G421600",
    "TraesCS6D02G367200"
  )
)

Seurat::DotPlot(obj_mock, features = bundle_sheath) +
  ggplot2::coord_flip() +
  ggplot2::facet_grid(
    facets = feature.groups ~ .,
    scales = "free_y",
    space = "free_y") +
  ggplot2::theme(
    axis.text.x = ggplot2::element_text(angle = 45, hjust = 1),
    strip.text.y = ggplot2::element_text(angle = 0)
  )
```

表达量太低。

#### Xylem parenchyma

AAP6 (AT5G49630)，AtHKT1 (AT4G10310) and GLR3.6 (AT3G51480) for xylem parenchyma

```{r}
xylem_parenchyma <- list(
  "AAP6" = c(
    "TraesCS2A02G268200",
    "TraesCS2B02G266600",
    "TraesCS2D02G256000"
  ),
  "AtHKT1" = c(
    "TraesCS6D02G144500",
    "TraesCS6B02G182600",
    "TraesCS2B02G451700",
    "TraesCS2D02G428500"
  ),
  "GLR3.6" = c(
    "TraesCSU02G107900",
    "TraesCS6A02G049900",
    "TraesCS6A02G049000",
    "TraesCS2A02G455200",
    "TraesCS6B02G065400",
    "TraesCS6B02G065700",
    "TraesCS2B02G477300",
    "TraesCS2D02G455600",
    "TraesCSU02G108800",
    "TraesCS6B02G066200",
    "TraesCSU02G107990",
    "TraesCS6A02G048700",
    "TraesCS6A02G048800",
    "TraesCSU02G107700",
    "TraesCS6B02G065500"
  )
)
```

```{r fig.height=7, fig.width=14}
p1 <- Seurat::DotPlot(obj_mock, features = xylem_parenchyma) +
  ggplot2::coord_flip() +
  ggplot2::facet_grid(
    facets = feature.groups ~ .,
    scales = "free_y",
    space = "free_y") +
  ggplot2::theme(
    axis.text.x = ggplot2::element_text(angle = 45, hjust = 1),
    strip.text.y = ggplot2::element_text(angle = 0)
  )

p2 <- Seurat::DotPlot(obj, features = xylem_parenchyma) +
  ggplot2::coord_flip() +
  ggplot2::facet_grid(
    facets = feature.groups ~ .,
    scales = "free_y",
    space = "free_y") +
  ggplot2::theme(
    axis.text.x = ggplot2::element_text(angle = 45, hjust = 1),
    strip.text.y = ggplot2::element_text(angle = 0)
  )

(p <- p1 + p2)
```

The "Percent Expressed" is too low.

```{r}
Seurat::FeaturePlot(obj_mock, features = "TraesCS2B02G451700", reduction = "umap", order = TRUE) +
  ggplot2::coord_fixed() +
  theme_dimred()
```

### Cell population relationship

#### RNA velocity of MOCK cells

```{r}
adata_mock <- readPickle(Sys.glob(
  "../results/ObjectCache/TrajectoryInference/unitvelo_mock_config_1_*.pickle"))
```

```{r}
sce_mock_slim <- sce_mock[, adata_mock$obs_names$to_list()]
```

```{r}
adata_mock$obsm$update(list(
  X_umap = reducedDim(sce_mock_slim, "UMAP")
))

adata_mock$obs$cellType <- droplevels(sce_mock_slim$cluster_id)
```

```{r}
scv$tl$latent_time(adata_mock, min_likelihood=NULL)
sce_mock_slim$latent_time <- adata_mock$obs$latent_time
```

```{r}
scv$tl$velocity_embedding(adata_mock, basis = "umap")
mock_vector_df <- velociraptor::gridVectors(
  adata_mock$obsm["X_umap"],
  adata_mock$obsm["velocity_umap"],
  resolution = 30)
```

```{r fig.height=6, fig.width=14}
p1 <- scater::plotReducedDim(
    sce_mock_slim,
    colour_by="cluster_id",
    dimred = "UMAP",
    point_size = 0.8,
    point_alpha = 1,
    theme_size = 14) +
  ggplot2::scale_color_manual(values = mock_celltype_colors) +
  ggplot2::geom_segment(
    data = mock_vector_df,
    mapping = ggplot2::aes_string(
      x = "start.1",
      y = "start.2", 
      xend = "end.1",
      yend = "end.2"),
    arrow = grid::arrow(
      length = grid::unit(0.06, "inches"),
      angle = 25,
      type = "closed"),
    size = 0.5) +
  ggplot2::coord_fixed() +
  legend_override("color", list(size = 4, alpha = 1), title = "cluster") +
  theme_dimred()

p2 <- scater::plotReducedDim(
    sce_mock_slim,
    dimred = "UMAP",
    point_size = 0.8,
    colour_by = "latent_time",
    theme_size = 14
  ) +
  ggplot2::scale_color_gradientn(colours = blue_to_orange) +
  ggplot2::coord_fixed() +
  remove_legend_title() +
  center_plot_title() +
  theme_dimred()

(p <- p1 + p2)
```

```{r include=FALSE, results='hide'}
rm(adata_mock, sce_mock_slim)
gc()
```

### Reconstruction of cell identity

#### Me_ε marker

```{r fig.height=6, fig.width=14}
p1 <- Seurat::FeaturePlot(obj_mock, features = "TraesCS1B02G406200",
                    reduction = "umap", order = TRUE) +
  ggplot2::coord_fixed() +
  theme_dimred()

p2 <- Seurat::VlnPlot(obj_mock, features = "TraesCS1B02G406200") +
  ggplot2::xlab("Cell Populations") +
  remove_legend()

(p <- patchwork::wrap_plots(p1, p2, nrow = 1))
```

#### Va_β marker

```{r fig.height=6, fig.width=14}
p1 <- Seurat::FeaturePlot(obj_mock, features = "TraesCS6A02G041700",
                    reduction = "umap", order = TRUE) +
  ggplot2::coord_fixed() +
  theme_dimred()

p2 <- Seurat::VlnPlot(obj_mock, features = "TraesCS6A02G041700") +
  ggplot2::xlab("Cell Populations") +
  remove_legend()

(p <- patchwork::wrap_plots(p1, p2, nrow = 1))
```

### Temporal transcriptional programs

```{r}
filelist <- paste0("../data/pathways/Photosyn", c("Ant", "Cfix", "PSEA"), ".txt")
names(filelist) <- basename(filelist) |> stringr::str_remove(stringr::fixed(".txt"))
photosyn <- lapply(filelist, function(filename) {
  genes <- stringr::str_trim(readLines(filename))
  genes[genes != ""]
})
```

```{r}
obj_mock@meta.data[, names(photosyn)] <- as.data.frame(t(calculateModuleScore(logcounts(sce_mock), features = photosyn)))
```

```{r fig.height=7, fig.width=14}
p1 <- Seurat::VlnPlot(
  obj_mock,
  features = c("nCount_RNA"),
  pt.size = 0, log = TRUE)

p234 <- Seurat::VlnPlot(
  obj_mock,
  features = c("PhotosynAnt",
               "PhotosynCfix", "PhotosynPSEA"),
  ncol = 2, pt.size = 0)

p234 <- p234 &
  ggplot2::ylab("Module Score")

plist <- patchwork::wrap_plots(p1, p234[[1]], p234[[2]], p234[[3]])

p <- plist &
  ggplot2::scale_fill_manual(values = mock_celltype_colors) &
  ggplot2::xlab("Cell Population") &
  remove_legend()

p
```


## Wheat Responses to Fungal Invasion

### Cell label transfer

```{r}
sce$predicted.id <- factor(sce$predicted.id, mock_celltypes)
df <- colData(sce)[, c(
    "cluster_id", "predicted.id", "tissue",
    "body_layer", "prediction.score.max")] |>
  as.data.frame()
```

```{r}
p1 <- scater::plotReducedDim(sce, dimred = "UMAP",
                             colour_by = "predicted.id",
                             text_by = "predicted.id",
                             point_size = 0.5,
                             theme_size = 14) +
  ggplot2::scale_color_manual(values = mock_celltype_colors) +
  legend_override("color", list(size = 4, alpha = 1), title = "cluster") +
  ggplot2::coord_fixed() +
  theme_dimred()

p3 <- scater::plotReducedDim(sce, dimred = "UMAP",
                             colour_by = "cluster_id",
                             text_by = "cluster_id",
                             point_size = 0.5,
                             theme_size = 14) +
  ggplot2::scale_color_manual(values = all_celltype_colors) +
  legend_override("color", list(size = 4, alpha = 1), title = "cluster") +
  ggplot2::coord_fixed() +
  theme_dimred()
```

```{r}
df_sankey <- df |>
  dplyr::group_by(predicted.id, cluster_id, tissue, body_layer) |>
  dplyr::summarise(n = dplyr::n(), .groups = "drop") |>
  dplyr::mutate(freq = n / sum(n))

# When the strata are defined by a character or factor variable,
# they default to the order of the variable.
# Duplicated axis values will broke the order of stratum
df_sankey$predicted.id <- forcats::fct_recode(
  df_sankey$predicted.id, Gu_ = "Gu", BS_ = "BS", CC_ = "CC")
```

```{r fig.height=7, fig.width=14}
p4 <- ggplot2::ggplot(
  data = df_sankey,
  mapping = ggplot2::aes(
    axis1 = predicted.id,
    axis2 = cluster_id,
    axis3 = tissue,
    axis4 = body_layer,
    y = freq)) +
  ggalluvial::geom_alluvium(
    mapping = ggplot2::aes(fill = cluster_id)) +
  ggplot2::scale_fill_manual(values = all_celltype_colors) +
  ggalluvial::geom_stratum(alpha = 0) +
  ggplot2::geom_text(
    stat = ggalluvial::StatStratum,
    # Change axis back to original ones
    mapping = ggplot2::aes(label = forcats::fct_recode(
      ggplot2::after_stat(stratum), Gu = "Gu_", BS = "BS_", CC = "CC_"))) +
  ggplot2::scale_y_continuous(expand = c(0, 0.01)) +
  ggplot2::scale_x_continuous(expand = c(0, 0.001)) +
  ggplot2::theme_void() +
  remove_legend()

p4
```

```{r}
p7 <- ggplot2::ggplot(df, ggplot2::aes(
    x = cluster_id, y = prediction.score.max, fill = cluster_id)) +
  ggplot2::geom_boxplot(outlier.size = .5) +
  ggplot2::scale_fill_manual(values = all_celltype_colors) +
  ggplot2::xlab("cluster") +
  ggplot2::theme_bw(14) +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1)) +
  remove_legend()
```

```{r fig.height=16, fig.width=14}
p <- (p1 + p3) / p4 / p7 +
  patchwork::plot_layout(heights = c(1, 1, 0.5)) +
  patchwork::plot_annotation(tag_levels = "A") &
  font_plot_tag(16)

p
```

### Known Pathways {.tabset}

#### Photosynthesis

```{r fig.height=7, fig.width=14}
photos <- pathways[c("PhotosynAnt", "PhotosynCfix", "PhotosynPSEA")] |>
  tibble::enframe(name = "Pathway", "Gene") |>
  tidyr::unnest(cols = Gene)

p <- plotDEGsHeatmap(
  logfc_data$mtx,
  photos$Gene,
  logfc_data$heatanno,
  column_split = logfc_data$coldata$cell_types,
  show_row_names = FALSE,
  asterisk = FALSE,
  row_split = with(photos,
    Pathway[match(intersect(rownames(logfc_data$mtx), Gene), Gene)]
  ))

#plotpreview(p, width = 14, height = 7)
```

#### Lignin biogenesis

TODO: 这里应该保留最细粒度的 GSVA。

```{r fig.height=7, fig.width=14}
lignin <- pathways[c(
    "Aromatic amino acid",
    "General phenylpropanoid",
    "Lignin biosynthesis")] |>
  tibble::enframe(name = "Pathway", "Gene") |>
  tidyr::unnest(cols = Gene)

p1 <- plotLineFacetGSVA(gsva_df, unique(lignin$Pathway)) +
  ggplot2::facet_grid(Pathway ~ cellType)

p1
```

```{r fig.height=10, fig.width=14}
p2 <- plotDEGsHeatmap(
  logfc_data$mtx,
  lignin$Gene,
  logfc_data$heatanno,
  column_split = logfc_data$coldata$cell_types,
  show_row_names = FALSE,
  asterisk = FALSE,
  row_split = with(lignin,
    Pathway[match(intersect(rownames(logfc_data$mtx), Gene), Gene)]
  ),
  row_title_gp = grid::gpar(fontsize = 8))
```

```{r include=FALSE, eval=FALSE}
ggpreview(p1, height = 7, width = 14)
plotpreview(p2, width = 14, height = 10)
```

#### PTI related

```{r fig.height=5, fig.width=14}
p1 <- plotLineFacetGSVA(gsva_df, "RLK")

p1
```

```{r fig.height=5, fig.width=14}
p2 <- plotDEGsHeatmap(
  logfc_data$mtx,
  pathways[["RLK"]],
  logfc_data$heatanno,
  column_split = logfc_data$coldata$cell_types,
  show_row_names = FALSE,
  asterisk = FALSE)
```

```{r include=FALSE, eval=FALSE}
ggpreview(p1, height = 3, width = 14)
plotpreview(p2, height = 6, width = 14)
```

#### ETI related

```{r fig.height=5, fig.width=14}
p1 <- plotLineFacetGSVA(gsva_df, "NBLRR")

p1
```

```{r fig.height=5, fig.width=14}
p2 <- plotDEGsHeatmap(
  logfc_data$mtx,
  pathways[["NBLRR"]],
  logfc_data$heatanno,
  column_split = logfc_data$coldata$cell_types,
  show_row_names = FALSE,
  asterisk = FALSE)
```

```{r include=FALSE, eval=FALSE}
ggpreview(p1, height = 3, width = 14)
plotpreview(p2, height = 6, width = 14)
```

#### ROS produce and reduce

```{r fig.height=5, fig.width=14}
ROS <- pathways[c("RBOH", "TaGST")] |>
  tibble::enframe(name = "Pathway", "Gene") |>
  tidyr::unnest(cols = Gene)

p1 <- plotLineFacetGSVA(gsva_df, unique(ROS$Pathway)) +
  ggplot2::facet_grid(Pathway ~ cellType)

p1
```

```{r fig.height=10, fig.width=14}
p2 <- plotDEGsHeatmap(
  logfc_data$mtx,
  ROS$Gene,
  logfc_data$heatanno,
  column_split = logfc_data$coldata$cell_types,
  show_row_names = FALSE,
  asterisk = FALSE,
  row_split = with(ROS,
    Pathway[match(intersect(rownames(logfc_data$mtx), Gene), Gene)]
  ))
```

```{r include=FALSE, eval=FALSE}
ggpreview(p1, height = 5, width = 14)
plotpreview(p2, height = 6, width = 14)
```

#### Outer sheath immunity responses

```{r fig.height=12, fig.width=14}
Os_immune <- pathways[
    c("RLK", "NBLRR", "RBOH", "Aromatic amino acid",
      "General phenylpropanoid", "TaGST", "Polyamine")] |>
  tibble::enframe(name = "Pathway", "Gene") |>
  tidyr::unnest(cols = Gene)

p1 <- plotLineFacetGSVA(gsva_df, unique(Os_immune$Pathway)) +
  ggplot2::facet_grid(Pathway ~ cellType)

p1
```

```{r fig.height=10, fig.width=14}
p2 <- plotDEGsHeatmap(
  logfc_data$mtx,
  Os_immune$Gene,
  logfc_data$heatanno,
  column_split = logfc_data$coldata$cell_types,
  show_row_names = FALSE,
  asterisk = FALSE,
  row_title_gp = grid::gpar(fontsize = 6),
  row_split = with(Os_immune,
    Pathway[match(intersect(rownames(logfc_data$mtx), Gene), Gene)]
  ))
```

```{r include=FALSE, eval=FALSE}
plotpreview(p2, height = 8, width = 14)
```

## Cell Dynamics of L2 Body Layer

### Cell Type Proportion

```{r}
ca_res_all <- readRDS(Sys.glob("../results/ObjectCache/DifferentialAbundance/ca_res_nonclust_all_*.rds"))
ca_res_all$results$cellTypes <- forcats::fct_relevel(ca_res_all$results$cellTypes, all_celltypes)
```

```{r fig.height=8, fig.width=12}
df <- rhapsodykit::calculateClusterProportion(
  sce$cluster_id, sce$sample_id, sce$group_id)

p1 <- df %>%
  ggplot2::ggplot(ggplot2::aes(
    x = sample_id, y = frequency, fill = cluster_id)) +
  ggplot2::facet_wrap(~group_id, nrow = 1, scales = "free_x") +
  ggplot2::geom_bar(
    stat = "identity", col = "white",
    width = 1, size = 0.2, position = "stack") +
  ggplot2::scale_y_continuous(breaks = seq(0, 1, 0.2), expand = c(0, 0)) +
  ggplot2::scale_x_discrete(expand = c(0, 0)) +
  ggplot2::scale_fill_manual(values = all_celltype_colors) +
  ggplot2::theme(
    aspect.ratio = NULL,
    panel.grid = ggplot2::element_blank(),
    panel.spacing = grid::unit(1, "mm"),
    strip.text = ggplot2::element_blank(),
    strip.background = ggplot2::element_blank()) +
  rotate_x_labels()

p2 <- df %>%
  ggplot2::ggplot(ggplot2::aes(
    x = sample_id, y = frequency, fill = group_id)) +
  ggplot2::facet_wrap(~cluster_id, ncol = 3) +
  ggplot2::geom_bar(
    stat = "identity", col = "white",
    width = 1, size = 0.2,
    position = ggplot2::position_dodge()) +
  ggthemes::scale_fill_tableau("Classic Cyclic") +
  ggplot2::scale_y_continuous(expand = c(0, 0)) +
  ggplot2::theme_bw() +
  ggplot2::theme(
    strip.text.x = ggplot2::element_text(size = 10)) +
  verticalize_x_labels()

p3 <- plot_bootstrap_distribution(ca_res_all, ncol = 3) +
  ggplot2::theme(
    strip.text.x = ggplot2::element_text(size = 10)) +
  ggplot2::scale_color_brewer(type = "qual", palette = "Dark2") +
  rotate_x_labels() +
  empty_strip()

(p <- p1+ p3 +
  patchwork::plot_layout(widths  = c(1, 1), guides = "collect"))
```

```{r}
p <- rhapsodykit::calculateClusterProportion(
  sce$body_layer, sce$sample_id, sce$group_id) |>
  dplyr::mutate(percent = scales::label_percent(accuracy = 0.1)(frequency)) |>
  ggplot2::ggplot(ggplot2::aes(
    x = sample_id, y = frequency, fill = cluster_id)) +
  ggplot2::facet_wrap(~group_id, nrow = 1, scales = "free_x") +
  ggplot2::geom_bar(
    stat = "identity", col = "white",
    width = 1, size = 0.2, position = "stack") +
  ggplot2::geom_text(
    mapping = ggplot2::aes(label = percent),
    size = 2, stat = "identity",
    position = ggplot2::position_fill(vjust = 0.5)) +
  ggplot2::scale_y_continuous(breaks = seq(0, 1, 0.2), expand = c(0, 0)) +
  ggplot2::scale_x_discrete(expand = c(0, 0)) +
  ggplot2::guides(fill = ggplot2::guide_legend(title = "body_layer")) +
  ggplot2::theme(
    aspect.ratio = NULL,
    panel.grid = ggplot2::element_blank(),
    panel.spacing = grid::unit(1, "mm"),
    strip.text = ggplot2::element_blank(),
    strip.background = ggplot2::element_blank()) +
  rotate_x_labels()

p
```

```{r include=FALSE, results='hide'}
rm(ca_res_all)
gc()
```

### Co-varing neighborhood analysis

```{r}
sce_all_mock <- sce[, sce$treatment == "MOCK"]

sce_all_mock$time_val <- as.numeric(factor(
  sce_all_mock$time, levels = c("0DPI", "1DPI", "2DPI", "3DPI")))
sce_all_mock$sample <- as.character(sce_all_mock$sample_id)

sce_all_mock <- association.SingleCellExperiment(
  sce_all_mock, test_var = "time_val", samplem_key = "sample_id", dimred = "HARMONY")
```

```{r}
df_ncorrs_mock <- data.frame(
  time_ncorrs = sce_all_mock$cna_ncorrs,
  cellType = sce_all_mock$tissue,
  treatment = "MOCK",
  fdr_5p_t = metadata(sce_all_mock)$cnaRes$fdr_5p_t)

df_ncorrs_mock$cellType <- forcats::fct_reorder(
  df_ncorrs_mock$cellType, df_ncorrs_mock$time_ncorrs)

p1 <- plotNcorrRidges(sce_all_mock, group_by = "tissue") +
  ggthemes::scale_fill_tableau("Tableau 10") +
  ggplot2::scale_y_discrete(limits=rev, expand = ggplot2::expansion(add = c(0, 4))) +
  NULL

p1
```

```{r}
sce_all_pnr2 <- sce[, sce$treatment == "PNR2" | sce$time == "0DPI"]

sce_all_pnr2$time_val <- as.numeric(factor(
  sce_all_pnr2$time, levels = c("0DPI", "1DPI", "2DPI", "3DPI")))
sce_all_pnr2$sample <- as.character(sce_all_pnr2$sample_id)

sce_all_pnr2 <- association.SingleCellExperiment(
  sce_all_pnr2, test_var = "time_val", samplem_key = "sample_id", dimred = "HARMONY")
```

```{r}
df_ncorrs_pnr2 <- data.frame(
  time_ncorrs = sce_all_pnr2$cna_ncorrs,
  cellType = sce_all_pnr2$tissue,
  treatment = "PNR2",
  fdr_5p_t = metadata(sce_all_pnr2)$cnaRes$fdr_5p_t)

df_ncorrs_pnr2$cellType <- forcats::fct_relevel(
  df_ncorrs_pnr2$cellType, levels(df_ncorrs_mock$cellType))

p2 <- plotNcorrRidges(sce_all_pnr2, group_by = "tissue") +
  ggthemes::scale_fill_tableau("Tableau 10") +
  ggplot2::scale_y_discrete(limits=rev, expand = ggplot2::expansion(add = c(0, 4))) +
  NULL

p2
```

```{r}
sce_all_tr4 <- sce[, sce$treatment == "TR4" | sce$time == "0DPI"]

sce_all_tr4$time_val <- as.numeric(factor(
  sce_all_tr4$time, levels = c("0DPI", "1DPI", "2DPI", "3DPI")))
sce_all_tr4$sample <- as.character(sce_all_tr4$sample_id)

sce_all_tr4 <- association.SingleCellExperiment(
  sce_all_tr4, test_var = "time_val", samplem_key = "sample_id", dimred = "HARMONY")
```

```{r}
df_ncorrs_tr4 <- data.frame(
  time_ncorrs = sce_all_tr4$cna_ncorrs,
  cellType = sce_all_tr4$tissue,
  treatment = "TR4",
  fdr_5p_t = metadata(sce_all_tr4)$cnaRes$fdr_5p_t)

df_ncorrs_tr4$cellType <- forcats::fct_relevel(
  df_ncorrs_tr4$cellType, levels(df_ncorrs_mock$cellType))

p3 <- plotNcorrRidges(sce_all_tr4, group_by = "tissue") +
  ggthemes::scale_fill_tableau("Tableau 10") +
  ggplot2::scale_y_discrete(limits=rev, expand = ggplot2::expansion(add = c(0, 4))) +
  NULL

p3
```

```{r fig.height=5, fig.width=14}
df_ncorrs <- dplyr::bind_rows(df_ncorrs_mock, df_ncorrs_pnr2, df_ncorrs_tr4)
df_ncorrs$cellType <- forcats::fct_reorder(
  df_ncorrs$cellType, df_ncorrs$time_ncorrs)

fdr_df <- purrr::imap(
  .x = list(
    MOCK = sce_all_mock,
    PNR2 = sce_all_pnr2,
    TR4  = sce_all_tr4),
  .f = function(x, treat) {
    fdr <- metadata(x)$cnaRes$fdr_5p_t
    data.frame(
      treatment = treat,
      posfdr = fdr,
      negfdr = -fdr)
  }) |> dplyr::bind_rows()

p2 <- ggplot2::ggplot(df_ncorrs, ggplot2::aes(
    x = time_ncorrs, y = cellType, fill = cellType)) +
  ggplot2::ylab("Cell Population") +
  ggplot2::xlab("Neighborhood coefficients of time") +
  ggridges::geom_density_ridges(scale = 4) +
  ggplot2::geom_vline(
    data = fdr_df, mapping = ggplot2::aes(xintercept = posfdr),
    linetype = "dashed") +
  ggplot2::geom_vline(
    data = fdr_df, mapping = ggplot2::aes(xintercept = negfdr),
    linetype = "dashed") +
  ggplot2::facet_wrap(~ treatment) +
  ggplot2::scale_fill_manual(values = structure(
    ggthemes::tableau_color_pal("Tableau 10")(length(TISSUE_TYPES)),
    names = TISSUE_TYPES)) +
  ggplot2::scale_y_discrete(limits=rev, expand = ggplot2::expansion(add = c(0, 4.5))) +
  ggridges::theme_ridges(14) +
  empty_strip(
    strip.text = ggplot2::element_text(size = 14),
    panel.spacing = grid::unit(2, "lines")) +
  remove_legend()

p2
```

```{r include=FALSE, results='hide'}
rm(df_ncorrs_mock, sce_all_pnr2, df_ncorrs_pnr2, sce_all_tr4, df_ncorrs_tr4, df_ncorrs)
gc()
```

### RNA Velocity and cluster connectivity

```{r}
adata_all_mock <- readPickle(Sys.glob(
  "../results/ObjectCache/TrajectoryInference/unitvelo_adata_mock_*.pickle"))
adata_all_pnr2 <- readPickle(Sys.glob(
  "../results/ObjectCache/TrajectoryInference/unitvelo_adata_pnr2_*.pickle"))
adata_all_tr4 <- readPickle(Sys.glob(
  "../results/ObjectCache/TrajectoryInference/unitvelo_adata_tr4_*.pickle"))
```

```{r}
sce_sub_list <- lapply(
  X = list(
    MOCK = adata_all_mock,
    PNR2 = adata_all_pnr2,
    TR4  = adata_all_tr4),
  FUN = function(adata) {
    sce_sub <- sce[, adata$obs_names$to_list()]
    adata$obsm$update(list(X_umap = reducedDim(sce_sub, "UMAP")))
    scv$tl$velocity_embedding(adata, basis = "umap")
    reducedDim(sce_sub, "velocity_umap") <- adata$obsm["velocity_umap"]
    sce_sub
  })
```

```{r fig.height=8, fig.width=14}
plist <- purrr::imap(sce_sub_list, function(sce_sub, name) {
  colnames(reducedDim(sce_sub, "UMAP")) <- NULL
  vector_df <- velociraptor::gridVectors(
      reducedDim(sce_sub, "UMAP"),
      reducedDim(sce_sub, "velocity_umap"),
      resolution = 30)
  scater::plotReducedDim(
      sce_sub,
      colour_by="cluster_id",
      dimred = "UMAP",
      point_size = 0.8,
      point_alpha = 1,
      theme_size = 14) +
    ggplot2::scale_color_manual(values = all_celltype_colors) +
    ggplot2::geom_segment(
      data = vector_df,
      mapping = ggplot2::aes_string(
        x = "start.1",
        y = "start.2", 
        xend = "end.1",
        yend = "end.2"),
      arrow = grid::arrow(
        length = grid::unit(0.06, "inches"),
        angle = 25,
        type = "closed"),
      size = 0.5) +
    ggplot2::coord_fixed() +
    ggplot2::ggtitle(name) +
    center_plot_title(face = "plain") +
    theme_dimred()
})

(p <- patchwork::wrap_plots(plist) +
    patchwork::plot_layout(guides = "collect") &
    ggplot2::theme(legend.position = "bottom") &
    ggplot2::guides(color = ggplot2::guide_legend(
      override.aes = list(size = 4, alpha = 1),
      title = "cluster", nrow = 1, byrow = TRUE)))
```

```{r}
adata <- runPAGA(
  sce, group_key = "cluster_id",
  assay.X = "logcounts", use.dimred = "HARMONY"
)
```

```{r}
adata$obsm$update(list(
  X_umap = reducedDim(sce, "UMAP")
))
```

```{r}
p <- plotPagaGraph(adata, basis = "umap", edge_cex = 2, node_cex = 2) +
  ggplot2::coord_fixed() +
  ggplot2::xlab("UMAP 1") +
  ggplot2::ylab("UMAP 2") +
  theme_dimred()

p
```

```{r include=FALSE, results='hide'}
rm(adata_all_mock, adata_all_pnr2, adata_all_tr4)
gc()
```

### Differential abundance analysis

```{r}
da_ctr_1dpi <- readRDS(Sys.glob(
  "../results/ObjectCache/DifferentialAbundance/da_ctr_1dpi_*.rds"))
da_ctr_2dpi <- readRDS(Sys.glob(
  "../results/ObjectCache/DifferentialAbundance/da_ctr_2dpi_*.rds"))
da_ctr_3dpi <- readRDS(Sys.glob(
  "../results/ObjectCache/DifferentialAbundance/da_ctr_3dpi_*.rds"))
```

```{r}
p_da_1dpi <- patchwork::wrap_plots(
    rhapsodykit::plotDACellScore(
      da_ctr_1dpi[[1]], "umap", label = TRUE) +
      ggplot2::ggtitle(names(da_ctr_1dpi)[[1]]),
    rhapsodykit::plotDACellScore(
      da_ctr_1dpi[[2]], "umap", label = TRUE) +
      ggplot2::ggtitle(names(da_ctr_1dpi)[[2]])) +
  patchwork::plot_annotation(tag_levels = "A") &
  ggplot2::scale_color_gradientn(colours = blue_to_orange) &
  ggplot2::xlab("UMAP 1") &
  ggplot2::ylab("UMAP 2") &
  ggplot2::coord_fixed() &
  center_plot_title() &
  font_plot_tag(16) &
  theme_dimred()

p_da_2dpi <- patchwork::wrap_plots(
    rhapsodykit::plotDACellScore(
      da_ctr_2dpi[[1]], "umap", label = TRUE) +
      ggplot2::ggtitle(names(da_ctr_2dpi)[[1]]),
    rhapsodykit::plotDACellScore(
      da_ctr_2dpi[[2]], "umap", label = TRUE) +
      ggplot2::ggtitle(names(da_ctr_2dpi)[[2]])) +
  patchwork::plot_annotation(tag_levels = "A") &
  ggplot2::scale_color_gradientn(colours = blue_to_orange) &
  ggplot2::xlab("UMAP 1") &
  ggplot2::ylab("UMAP 2") &
  ggplot2::coord_fixed() &
  center_plot_title() &
  font_plot_tag(16) &
  theme_dimred()

p_da_3dpi <- patchwork::wrap_plots(
    rhapsodykit::plotDACellScore(
      da_ctr_3dpi[[1]], "umap", label = TRUE) +
      ggplot2::ggtitle(names(da_ctr_3dpi)[[1]]),
    rhapsodykit::plotDACellScore(
      da_ctr_3dpi[[2]], "umap", label = TRUE) +
      ggplot2::ggtitle(names(da_ctr_3dpi)[[2]])) +
  patchwork::plot_annotation(tag_levels = "A") &
  ggplot2::scale_color_gradientn(colours = blue_to_orange) &
  ggplot2::xlab("UMAP 1") &
  ggplot2::ylab("UMAP 2") &
  ggplot2::coord_fixed() &
  center_plot_title() &
  font_plot_tag(16) &
  theme_dimred()
```

```{r fig.height=14, fig.width=14}
( p <- p_da_1dpi / p_da_2dpi / p_da_3dpi)
```

```{r include=FALSE, results='hide'}
rm(da_ctr_1dpi, da_ctr_2dpi, da_ctr_3dpi)
gc()
```

### Slingshot trajectory inference

#### Mesophyll state transitions

```{r}
sce_me_all <- readRDS(Sys.glob("../results/ObjectCache/TrajectoryInference/sce_all_slingshot_me_all_*.rds"))
sce_me_1dpi <- readRDS(Sys.glob("../results/ObjectCache/TrajectoryInference/sce_all_slingshot_me_1dpi_*.rds"))
sce_me_2dpi <- readRDS(Sys.glob("../results/ObjectCache/TrajectoryInference/sce_all_slingshot_me_2dpi_*.rds"))
sce_me_3dpi <- readRDS(Sys.glob("../results/ObjectCache/TrajectoryInference/sce_all_slingshot_me_3dpi_*.rds"))
```

```{r fig.height=8, fig.width=14}
p1 <- scater::plotReducedDim(
    sce_me_all, dimred = "PHATE", theme_size = 14,
    colour_by = "cellType", text_by = "cellType",
    point_size = 1) +
  legend_override("color", list(size = 3, alpha = 1),
                  title = "cluster")

p2 <- plotSlingshotCurveOnReduc(
  sce_me_all, dimred = "PHATE",
  linewidth = 1.2, theme_size = 14, point_size = 1,
  lineage_rename = c(
    "Lineage1" = "Chlorenchymal silencing",
    "Lineage2" = "Chlorenchymal activation"),
  lineage_label_size = 4)

p3 <- plotLineageCurveOnReduc(
    sce_me_all, 1, dimred = "PHATE",
    linewidth = 1.2, theme_size = 14,
    point_size = 1) +
  ggplot2::ggtitle("Chlorenchymal silencing") +
  center_plot_title(face = "plain")

p4 <- plotLineageCurveOnReduc(
  sce_me_all, 2, dimred = "PHATE",
  linewidth = 1.2, theme_size = 14,
  point_size = 1) +
  ggplot2::ggtitle("Chlorenchymal activation") +
  center_plot_title(face = "plain")

p <- p1 + p2 + p3 + p4 +
  patchwork::plot_layout(ncol = 2) &
  ggplot2::coord_fixed() &
  theme_dimred()

p
```

```{r fig.height=12, fig.width=14}
plist <- purrr::imap(
  .x = list(
    "1DPI" = sce_me_1dpi,
    "2DPI" = sce_me_2dpi,
    "3DPI" = sce_me_3dpi),
  .f = function(x, name) {
    plotPseudotimeDensity(x, c(
      "Lineage1" = "Chlorenchymal silencing",
      "Lineage2" = "Chlorenchymal activation"))
  }
)

(p <- patchwork::wrap_plots(plist, ncol = 1, guides = "collect") &
    ggplot2::scale_x_continuous(expand = ggplot2::expansion(mult = 0.01)))
```

```{r include=FALSE, results='hide'}
rm(sce_me_all, sce_me_1dpi, sce_me_2dpi, sce_me_3dpi)
gc()
```


## Proof-of-concept studies

### ISH for Subepidermalization

```{r}
TMV_N_like <- c(
  "TraesCS1B02G406200",
  "TraesCS1D02G387400",
  "TraesCS1A02G343500",
  "TraesCS1B02G413100",
  "TraesCS1D02G178700",
  "TraesCS1D02G183600",
  "TraesCS1D02G212600",
  "TraesCS2A02G240400",
  "TraesCS2A02G242400",
  "TraesCS2A02G240700",
  "TraesCS2B02G002300",
  "TraesCS2B02G252400",
  "TraesCS2B02G331800",
  "TraesCS2B02G521400",
  "TraesCS2B02G601100",
  "TraesCS3A02G405700",
  "TraesCS3A02G430500",
  "TraesCS3A02G485300",
  "TraesCS3B02G066800",
  "TraesCS3B02G189500",
  "TraesCS4A02G155900",
  "TraesCS4A02G155200",
  "TraesCS4A02G272300",
  "TraesCS4D02G150400",
  "TraesCS5B02G039000",
  "TraesCS5B02G040100",
  "TraesCS5B02G096700",
  "TraesCS5B02G423500",
  "TraesCS5D02G221000",
  "TraesCS5D02G345100",
  "TraesCS5D02G345900",
  "TraesCS5D02G488100",
  "TraesCS6B02G402800",
  "TraesCS7A02G289000",
  "TraesCS7B02G396300",
  "TraesCS7D02G124000",
  "TraesCS7D02G255100",
  "TraesCS7D02G256100",
  "TraesCSU02G243900")

nlike_counts <- Matrix::rowSums(SeuratObject::GetAssayData(obj, slot = "counts")[
    TMV_N_like, SeuratObject::WhichCells(obj, idents = "Me_2"), drop = TRUE])

TMV_N_like <- names(sort(nlike_counts))
```

```{r fig.height=7, fig.width=14}
p1 <- Seurat::DotPlot(obj, features = TMV_N_like) +
  ggplot2::coord_flip() +
  ggplot2::xlab("Cell Population") +
  rotate_x_labels()

p2 <- Seurat::DotPlot(obj,
    features = TMV_N_like,
    group.by = "group",
    idents = "Me_2") +
  ggplot2::coord_flip() +
  rotate_x_labels()

(p <- p1 + p2)
```

```{r}
TMV_N_like_tree <- c(
  "TraesCS7D02G124000",
  "TraesCS7D02G256100",
  "TraesCS1D02G178700",
  "TraesCS2B02G002300",
  "TraesCS3B02G189500",
  "TraesCS4D02G150400",
  "TraesCS3A02G405700",
  "TraesCS4A02G155200",
  "TraesCS5B02G423500",
  "TraesCS1D02G212600",
  "TraesCS5D02G345900",
  "TraesCS7D02G255100",
  "TraesCS1A02G343500",
  "TraesCS5D02G345100",
  "TraesCS5B02G096700",
  "TraesCS3A02G485300",
  "TraesCS1B02G413100",
  "TraesCS2B02G521400",
  "TraesCS1D02G183600",
  "TraesCSU02G243900",
  "TraesCS5B02G040100",
  "TraesCS2B02G331800",
  "TraesCS2A02G240700",
  "TraesCS2A02G240400",
  "TraesCS2B02G601100",
  "TraesCS5D02G221000",
  "TraesCS1B02G406200",
  "TraesCS1D02G387400",
  "TraesCS3A02G430500",
  "TraesCS2B02G252400",
  "TraesCS7B02G396300",
  "TraesCS7A02G289000",
  "TraesCS4A02G155900",
  "TraesCS5B02G039000",
  "TraesCS5D02G488100",
  "TraesCS2A02G242400",
  "TraesCS6B02G402800",
  "TraesCS4A02G272300",
  "TraesCS3B02G066800"
)

p <- Seurat::DotPlot(obj, features = rev(TMV_N_like_tree)) +
  ggplot2::coord_flip() +
  ggplot2::xlab("Cell Population") +
  rotate_x_labels()

p
```


```{r}
me_2_marker <- c(
  "SERN13"      = "TraesCS7D02G256100",
  "SERN12-2B"   = "TraesCS2B02G521400",
  "SERN11"      = "TraesCSU02G243900",
  "SERN10-3A"   = "TraesCS3A02G405700",
  "SERN10-5B"   = "TraesCS5B02G423500",
  "SERN9-3A"    = "TraesCS3A02G430500",
  "SERN8"       = "TraesCS2A02G242400",
  "SERN7"       = "TraesCS4A02G272300",
  "SERN6"       = "TraesCS1D02G183600",
  "SERN5"       = "TraesCS6B02G402800",
  "SERN4"       = "TraesCS1B02G406200",
  "SERN3"       = "TraesCS7D02G255100",
  "SERN2"       = "TraesCS1B02G413100",
  "SERN1"       = "TraesCS5B02G039000"
)

Seurat::DotPlot(obj, features = setNames(me_2_marker, NULL)) +
  ggplot2::coord_flip() +
  ggplot2::scale_x_discrete(labels = names(me_2_marker)) +
  ggplot2::ylab("Cell Population") +
  ggplot2::theme_bw() +
  ggplot2::theme(
    axis.text.x = ggplot2::element_text(angle = 45, hjust = 1),
    strip.text.y = ggplot2::element_text(angle = 0),
    panel.grid = ggplot2::element_blank()
  )
```

```{r fig.height=7, fig.width=14}
p1 <- Seurat::VlnPlot(
    obj, features = rev(me_2_marker),
    stack = TRUE, flip = TRUE, fill.by = "ident") +
  ggplot2::xlab("Cell Population") +
  remove_legend()

p2 <- Seurat::DotPlot(obj, features = setNames(me_2_marker, NULL)) +
  ggplot2::coord_flip() +
  ggplot2::scale_x_discrete(labels = names(me_2_marker)) +
  ggplot2::ylab("Cell Population") +
  ggplot2::theme_bw(14) +
  ggplot2::guides(
    size = ggplot2::guide_legend(title = "Percent\nExpressed"),
    colour = ggplot2::guide_colorbar(title = "Average\nExpression")) +
  ggplot2::theme(
    axis.text.x = ggplot2::element_text(angle = 45, hjust = 1, colour = "black"),
    axis.text.y = ggplot2::element_text(colour = "black"),
    panel.grid = ggplot2::element_blank())

p3 <- Seurat::FeaturePlot(
    obj, features = "TraesCS5B02G039000",
    reduction = "umap", order = TRUE, label = TRUE) +
  ggplot2::coord_fixed() +
  theme_dimred() +
  ggplot2::theme(
    legend.position = c(0.8, 0.1),
    plot.title = ggplot2::element_blank()
  )

(p <- p2 + p3 +
    patchwork::plot_layout(widths = c(2, 3)) +
    #patchwork::plot_annotation(tag_levels = "A") &
    font_plot_tag(16) &
    ggplot2::theme(
      strip.text = ggplot2::element_text(face = "plain")))
```

```{r}
geneDetectedPercent <- function(obj, features, idents) {
  counts <- SeuratObject::GetAssayData(obj, slot = "counts")[
      features, SeuratObject::WhichCells(obj, idents = idents), drop = FALSE]

  res_list <- apply(
    X = counts, MARGIN = 1, simplify = FALSE,
    FUN = function(genecounts) {
      percent <- sum(genecounts > 0) / length(genecounts)
      summa <- summary(genecounts[genecounts > 0])
      structure(c(Pct.Expr = percent*100, summa), class = "summaryDefault")
    }
  )

  res <- do.call(rbind, res_list)
  class(res) <- "summaryDefault"
  res
}

geneDetectedPercent(obj, rev(me_2_marker), c("Me_2"))
```

### ISH for Outersheath

```{r}
va_4_marker <- c(
  F43 = "TraesCS3B02G290300",
  F45 = "TraesCS5A02G386000",
  F46 = "TraesCS3D02G305300"
)

geneDetectedPercent(obj, va_4_marker, c("Va_4"))
```

```{r}
knitr::knit_exit()
```

### Lignin deposition

```{r}
PAL <- c("Aromatic amino acid", "General phenylpropanoid", "Lignin biosynthesis")
filelist <- sprintf("../data/pathways/%s.txt", PAL)
names(filelist) <- basename(filelist) |> stringr::str_remove(stringr::fixed(".txt"))

pathways <- lapply(filelist, function(filename) {
  genes <- stringr::str_trim(readLines(filename))
  genes[genes != ""]
})

pathways$Combined <- unlist(pathways, use.names = FALSE)
```

```{r}
expr_gsva <- GSVA::gsva(
  pseudobulk_expr, pathways, method = "gsva", kcdf = "Gaussian")

gsva_df <- as.data.frame(expr_gsva) |>
    tibble::rownames_to_column("Pathway") |>
    tidyr::pivot_longer(-Pathway, names_to = "bulk", values_to = "ES") |>
    tidyr::separate(bulk, sep = "\\.", into = c("treatment", "time", "cellType"))
```

```{r}
# Put inner sheath next to outer sheath
gsva_df$cellType <- factor(gsva_df$cellType, levels = c(
  "Gu", "Ep_1", "Ep_2",
  "Me_1", "Me_2", "Va_1", "Me_3", "Me_5", "Me_4",
  "Me_6", "Va_2", "Va_3", "Va_4", "BS", "CC",
  "MPV_1", "MPV_2"))
gsva_df$Pathway <- factor(gsva_df$Pathway, levels = c(PAL, "Combined"))
```

```{r}
gsva_df
```

```{r fig.height=11, fig.width=14}
cls_colors <- c(
        "Gu" = "black", "Ep_1" = "black", "Ep_2" = "black",
        "Me_1" = "#00B0F0", "Me_2" = "#00B0F0", "Va_1" = "#00B0F0", "Me_3" = "#00B0F0", "Me_5" = "#00B0F0",
        "Me_4" = "black", "Me_6" = "#FFC000", "Va_2" = "#FFC000", "Va_3" = "#FFC000", "Va_4" = "#FFC000",
        "BS"  = "black", "CC"  = "black", "MPV_1"  = "black", "MPV_2"  = "black")
cls_face <- ifelse(cls_colors == "black", "plain", "bold")

p <- gsva_df |>
  dplyr::filter(Pathway != "Combined") |>
  dplyr::arrange(cellType) |>
  ggplot2::ggplot(
    mapping = ggplot2::aes(
      x = cellType, y = ES,
      group = treatment,
      color = treatment,
      fill = treatment)) +
  ggplot2::geom_polygon(size = 1, alpha = .1) +
  ggplot2::geom_point() +
  ggplot2::scale_fill_brewer(type = "qual", palette = "Dark2") +
  ggplot2::scale_color_brewer(type = "qual", palette = "Dark2") +
  ggplot2::scale_y_continuous(position = "right") +
  ggplot2::facet_grid(Pathway ~ time, switch = "y") +
  ggplot2::guides(
    color = ggplot2::guide_legend(title = "Treatment"),
    fill = ggplot2::guide_legend(title = "Treatment")) +
  see::coord_radar() +
  see::theme_radar(
    base_size = 16,
    legend.position = "bottom",
    legend.title.size = 13,
    legend.text.size = 12,
    axis.text.size = 10) +
  empty_strip() +
  ggplot2::theme(
    axis.text.x = ggplot2::element_text(
      size = 10, colour = cls_colors, face = cls_face),
    axis.title.x = ggplot2::element_blank(),
    legend.spacing.x = grid::unit(10, 'pt')) +
  NULL

p
```


```{r fig.height=11, fig.width=16}
(p <- p + ggplot2::theme(legend.position = "right"))
```

```{r fig.height=5, fig.width=14}
cls_colors <- c(
        "Gu" = "black", "Ep_1" = "black", "Ep_2" = "black",
        "Me_1" = "black", "Me_2" = "black", "Va_1" = "black", "Me_3" = "black", "Me_5" = "black",
        "Me_4" = "black", "Me_6" = "red", "Va_2" = "red", "Va_3" = "red", "Va_4" = "red",
        "BS"  = "black", "CC"  = "black", "MPV_1"  = "black", "MPV_2"  = "black")
cls_face <- ifelse(cls_colors == "black", "plain", "bold")

p <- gsva_df |>
  dplyr::filter(Pathway == "Combined") |>
  dplyr::arrange(cellType) |>
  ggplot2::ggplot(
    mapping = ggplot2::aes(
      x = cellType, y = ES,
      group = treatment,
      color = treatment,
      fill = treatment)) +
  ggplot2::ylab("Enrichment Score") +
  ggplot2::geom_polygon(size = 1, alpha = .1) +
  ggplot2::geom_point() +
  ggplot2::scale_fill_brewer(type = "qual", palette = "Dark2") +
  ggplot2::scale_color_brewer(type = "qual", palette = "Dark2") +
  ggplot2::facet_wrap(~ time, nrow = 1) +
  ggplot2::guides(
    color = ggplot2::guide_legend(title = "Treatment"),
    fill = ggplot2::guide_legend(title = "Treatment")) +
  see::coord_radar() +
  see::theme_radar(
    base_size = 16,
    legend.position = "bottom",
    legend.title.size = 13,
    legend.text.size = 12,
    axis.text.size = 10) +
  empty_strip() +
  ggplot2::theme(
    axis.text.x = ggplot2::element_text(
      size = 10, colour = cls_colors, face = cls_face),
    legend.justification = 'left',
    axis.title.x = ggplot2::element_blank(),
    legend.spacing.x = grid::unit(10, 'pt')) +
  NULL

p
```

```{r fig.height=5, fig.width=14}
p <- gsva_df |>
  dplyr::filter(Pathway == "Combined") |>
  dplyr::arrange(cellType) |>
  ggplot2::ggplot(
    mapping = ggplot2::aes(
      x = cellType, y = ES,
      group = time,
      color = time,
      fill = time)) +
  ggplot2::ylab("Enrichment Score") +
  ggplot2::geom_polygon(size = 1, alpha = .1) +
  ggplot2::geom_point() +
  ggplot2::scale_fill_brewer(type = "qual", palette = "BrBG") +
  ggplot2::scale_color_brewer(type = "qual", palette = "BrBG") +
  ggplot2::facet_wrap(~ treatment, nrow = 1) +
  ggplot2::guides(
    color = ggplot2::guide_legend(title = "Treatment"),
    fill = ggplot2::guide_legend(title = "Treatment")) +
  see::coord_radar() +
  see::theme_radar(
    base_size = 16,
    legend.position = "bottom",
    legend.title.size = 13,
    legend.text.size = 12,
    axis.text.size = 10) +
  empty_strip() +
  ggplot2::theme(
    axis.title.x = ggplot2::element_blank(),
    legend.spacing.x = grid::unit(10, 'pt')) +
  NULL

p
```

```{r}
sce_os_3dpi <- readRDS(Sys.glob(
  "../results/ObjectCache/TrajectoryInference/sce_all_slingshot_os_3dpi_*.rds",))

lignin_expr <- calculateModuleScore(logcounts(sce_os_3dpi), features = pathways["Combined"])
```

```{r}
pts <- slingshot::slingPseudotime(sce_os_3dpi$slingshot)[, "Lineage3"] |> na.omit() |> sort()
act <- lignin_expr[, names(pts)]

ggplot2::qplot(x = pts, y = act) +
  ggplot2::geom_point()
```

